// igniteui-react-inputs/esm5/lib/ButtonGroupView_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, Delegate_$type, fromEnum, runOn, markType, INotifyPropertyChanged_$type, PropertyChangedEventArgs, delegateCombine, enumGetBox, delegateRemove, fromEn } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { ButtonGroupOrientation_$type } from "./ButtonGroupOrientation";
import { ButtonGroupDisplayStyle_$type } from "./ButtonGroupDisplayStyle";
import { BaseControlTheme_$type } from "igniteui-react-core";
import { ControlDisplayDensity_$type } from "igniteui-react-core";
import { ButtonGroupButtonCollection } from "./ButtonGroupButtonCollection";
import { ThemeResolver } from "./ThemeResolver";
import { ButtonGroupSelectionChangedEventArgs } from "./ButtonGroupSelectionChangedEventArgs";
import { CornerRadius } from "igniteui-react-core";
import { ButtonGroupVisualModelExport } from "./ButtonGroupVisualModelExport";
import { AppearanceHelper } from "igniteui-react-core";
import { ButtonVisualModelExport } from "./ButtonVisualModelExport";
import { BrushUtil } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
var ButtonGroupView = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(ButtonGroupView, _super);
    function ButtonGroupView() {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.g = null;
        _this.b = null;
        _this.f = new List$1(Delegate_$type, 0);
        _this.m = null;
        _this.c = true;
        _this.h = null;
        _this.i = null;
        _this.j = null;
        _this.k = null;
        return _this;
    }
    ButtonGroupView.prototype.s = function () {
    };
    ButtonGroupView.prototype.d = function () {
        return true;
    };
    Object.defineProperty(ButtonGroupView.prototype, "l", {
        get: function () {
            return this.h;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupView.prototype, "y", {
        get: function () {
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupView.prototype, "e", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroupView.prototype.q = function (a) {
        var e_1, _a;
        var _this = this;
        if (a == null) {
            try {
                for (var _b = tslib_1.__values(fromEnum(this.f)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var b = _c.value;
                    b();
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
            this.f.clear();
        }
        this.g = a;
        a.rootWrapper.setStyleProperty("position", "relative");
        var c = this.g.createElement("div");
        c.addClass("ig-button-content");
        c.setStyleProperty("position", "relative");
        c.setStyleProperty("display", "flex");
        c.setStyleProperty("flex-direction", "row");
        c.setStyleProperty("flex-wrap", "no-wrap");
        c.setStyleProperty("align-items", "center");
        this.j = c;
        this.f.add(c.listen("click", runOn(this, this.n)));
        this.g.append(c);
        this.g.getPortal(c, "ChildContent", function (d) {
            var comp_ = d.componentRef;
            _this.k = comp_;
            if (_this.a.r == 0) {
                _this.j.setStyleProperty("flex-direction", "row");
                comp_.display = 'flex';
                comp_.flexDirection = 'row';
            }
            else {
                _this.j.setStyleProperty("flex-direction", "column");
                comp_.display = 'flex';
                comp_.flexDirection = 'column';
            }
        }, true);
        this.a.db();
    };
    ButtonGroupView.prototype.w = function (a) {
        if (this.m != a) {
            this.m = a;
            if (this.i != null) {
                this.i.setAttribute("d", this.m);
            }
        }
    };
    ButtonGroupView.prototype.v = function () {
        var comp_ = this.k;
        if (comp_ != null) {
            if (this.a.r == 0) {
                this.j.setStyleProperty("flex-direction", "row");
                comp_.display = 'flex';
                comp_.flexDirection = 'row';
            }
            else {
                this.j.setStyleProperty("flex-direction", "column");
                comp_.display = 'flex';
                comp_.flexDirection = 'column';
            }
        }
    };
    ButtonGroupView.prototype.n = function (a) {
    };
    ButtonGroupView.prototype.t = function () {
        if (this.d()) {
        }
    };
    ButtonGroupView.prototype.p = function () {
        this.c = true;
    };
    ButtonGroupView.prototype.r = function () {
        var e_2, _a;
        this.c = false;
        try {
            for (var _b = tslib_1.__values(fromEnum(this.f)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var a = _c.value;
                a();
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        this.f.clear();
    };
    ButtonGroupView.prototype.x = function (a) {
        if (a != null) {
            this.g.rootWrapper.setStyleProperty("font", a.fontString);
        }
    };
    ButtonGroupView.prototype.u = function (a) {
        if (this.a.r == 0) {
            a.gb = "flex";
            a.he = "row";
            a.fv = "center";
        }
        else {
            a.gb = "flex";
            a.he = "column";
            a.fv = "stretch";
        }
    };
    ButtonGroupView.prototype.o = function (a) {
        this.j.append(a);
    };
    ButtonGroupView.$t = markType(ButtonGroupView, 'ButtonGroupView');
    return ButtonGroupView;
}(Base));
export { ButtonGroupView };
/**
 * @hidden
 */
var ButtonGroup = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(ButtonGroup, _super);
    function ButtonGroup() {
        var _this = _super.call(this) || this;
        _this.o = 1;
        _this.e = 0;
        _this.i = 1;
        _this.h = 1;
        _this.a = new Array(0);
        _this.u = false;
        _this._view = null;
        _this.ax = ThemeResolver.b(ButtonGroup.f, ButtonGroup.av, ButtonGroup.ap);
        _this.gz = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gn, ButtonGroup.fn);
        _this.g1 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gp, ButtonGroup.fp);
        _this.g0 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.go, ButtonGroup.fo);
        _this.g2 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gq, ButtonGroup.fq);
        _this.aw = ThemeResolver.b(ButtonGroup.f, ButtonGroup.au, ButtonGroup.ao);
        _this.g6 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gu, ButtonGroup.fu);
        _this.g3 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gr, ButtonGroup.fr);
        _this.g7 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gv, ButtonGroup.fv);
        _this.ha = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gy, ButtonGroup.fy);
        _this.g4 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gs, ButtonGroup.fs);
        _this.g5 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gt, ButtonGroup.ft);
        _this.g8 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gw, ButtonGroup.fw);
        _this.g9 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gx, ButtonGroup.fx);
        _this.ar = ThemeResolver.b(ButtonGroup.f, ButtonGroup.at, ButtonGroup.an);
        _this.fz = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gb, ButtonGroup.fb);
        _this.f1 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gd, ButtonGroup.fd);
        _this.f0 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gc, ButtonGroup.fc);
        _this.f2 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.ge, ButtonGroup.fe);
        _this.aq = ThemeResolver.b(ButtonGroup.f, ButtonGroup.as, ButtonGroup.am);
        _this.f6 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gi, ButtonGroup.fi);
        _this.f3 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gf, ButtonGroup.ff);
        _this.f7 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gj, ButtonGroup.fj);
        _this.ga = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gm, ButtonGroup.fm);
        _this.f4 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gg, ButtonGroup.fg);
        _this.f5 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gh, ButtonGroup.fh);
        _this.f8 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gk, ButtonGroup.fk);
        _this.f9 = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gl, ButtonGroup.fl);
        _this.ad = ThemeResolver.b(ButtonGroup.f, ButtonGroup.av, ButtonGroup.ap);
        _this.dm = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gn, ButtonGroup.fn);
        _this.dp = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gp, ButtonGroup.fp);
        _this.du = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gu, ButtonGroup.fu);
        _this.dr = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gr, ButtonGroup.fr);
        _this.dn = ThemeResolver.e(ButtonGroup.f, ButtonGroup.go, ButtonGroup.fo);
        _this.dq = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gq, ButtonGroup.fq);
        _this.ac = ThemeResolver.b(ButtonGroup.f, ButtonGroup.au, ButtonGroup.ao);
        _this.dv = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gv, ButtonGroup.fv);
        _this.dy = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gy, ButtonGroup.fy);
        _this.ds = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gs, ButtonGroup.fs);
        _this.dt = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gt, ButtonGroup.ft);
        _this.dw = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gw, ButtonGroup.fw);
        _this.dx = ThemeResolver.e(ButtonGroup.f, ButtonGroup.gx, ButtonGroup.fx);
        _this.ah = NaN;
        _this.d7 = null;
        _this.d9 = null;
        _this.ee = null;
        _this.eb = null;
        _this.ec = null;
        _this.d8 = null;
        _this.ea = null;
        _this.ag = NaN;
        _this.ed = null;
        _this.ev = null;
        _this.ey = null;
        _this.ew = null;
        _this.ex = null;
        _this.aj = NaN;
        _this.ef = null;
        _this.eh = null;
        _this.em = null;
        _this.ej = null;
        _this.ek = null;
        _this.eg = null;
        _this.ei = null;
        _this.ai = NaN;
        _this.el = null;
        _this.er = null;
        _this.eu = null;
        _this.es = null;
        _this.et = null;
        _this.af = NaN;
        _this.dz = null;
        _this.d1 = null;
        _this.d6 = null;
        _this.d3 = null;
        _this.d4 = null;
        _this.d0 = null;
        _this.d2 = null;
        _this.ae = NaN;
        _this.d5 = null;
        _this.en = null;
        _this.eq = null;
        _this.eo = null;
        _this.ep = null;
        _this.q = 0;
        _this.a6 = "ig-buttongroup-" + ButtonGroup.a4++;
        _this.w = false;
        _this.v = false;
        _this.x = false;
        _this.propertyChanged = null;
        _this.m = new ButtonGroupButtonCollection();
        _this.selectionChanged = null;
        var a = new ButtonGroupView();
        a.a = _this;
        _this.view = a;
        _this.view.s();
        var b = _this.m;
        b.collectionChanged = delegateCombine(b.collectionChanged, runOn(_this, _this.c3));
        var c = _this.m;
        c.collectionResetting = delegateCombine(c.collectionResetting, runOn(_this, _this.c4));
        return _this;
    }
    Object.defineProperty(ButtonGroup.prototype, "p", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            var b = this.o;
            this.o = a;
            if (b != this.o) {
                this.da("DisplayType", enumGetBox(ButtonGroupDisplayStyle_$type, b), enumGetBox(ButtonGroupDisplayStyle_$type, a));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "g", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            var b = this.e;
            this.e = a;
            if (b != this.e) {
                this.da("BaseTheme", enumGetBox(BaseControlTheme_$type, b), enumGetBox(BaseControlTheme_$type, a));
            }
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroup.prototype.c4 = function (a, b) {
        var e_3, _a;
        try {
            for (var _b = tslib_1.__values(fromEnum(this.buttons)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                c.clicked = delegateRemove(c.clicked, runOn(this, this.c6));
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        this.df();
    };
    ButtonGroup.prototype.appendContentChild = function (a) {
        this.view.o(a);
    };
    ButtonGroup.prototype.c3 = function (a, b) {
        var e_4, _a, e_5, _b;
        if (b.oldItems != null) {
            try {
                for (var _c = tslib_1.__values(fromEn(b.oldItems)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var c = _d.value;
                    var d = c;
                    if (d != null) {
                        d.clicked = delegateRemove(d.clicked, runOn(this, this.c6));
                    }
                }
            }
            catch (e_4_1) {
                e_4 = { error: e_4_1 };
            }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return))
                        _a.call(_c);
                }
                finally {
                    if (e_4)
                        throw e_4.error;
                }
            }
        }
        if (b.newItems != null) {
            try {
                for (var _e = tslib_1.__values(fromEn(b.newItems)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var e = _f.value;
                    var f = e;
                    if (f != null) {
                        f.clicked = delegateCombine(f.clicked, runOn(this, this.c6));
                    }
                }
            }
            catch (e_5_1) {
                e_5 = { error: e_5_1 };
            }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return))
                        _b.call(_e);
                }
                finally {
                    if (e_5)
                        throw e_5.error;
                }
            }
        }
        this.df();
    };
    ButtonGroup.prototype.c6 = function (a, b) {
        var c = a;
        var d = this.buttons.indexOf(c);
        this.de(c, d);
    };
    Object.defineProperty(ButtonGroup.prototype, "k", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            var b = this.i;
            this.i = a;
            if (b != this.i) {
                this.da("Density", enumGetBox(ControlDisplayDensity_$type, b), enumGetBox(ControlDisplayDensity_$type, a));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "j", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            var b = this.h;
            this.h = a;
            if (b != this.h) {
                this.da("ActualDensity", enumGetBox(ControlDisplayDensity_$type, b), enumGetBox(ControlDisplayDensity_$type, a));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "d", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            var b = this.a;
            this.a = a;
            if (b != this.a) {
                this.da("SelectedIndices", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroup.prototype.de = function (a, b) {
        this.u = true;
        if (this.aa(a)) {
            this.d = this.b(this.d, a, b);
        }
        else {
            if (!this.z) {
                this.d = new Array(0);
            }
            this.d = this.c(this.d, a, b);
        }
        this.df();
        if (this.selectionChanged != null) {
            this.selectionChanged(this, new ButtonGroupSelectionChangedEventArgs());
        }
        this.u = false;
    };
    ButtonGroup.prototype.c = function (a, b, c) {
        var d = new Array(Math.max(0, this.a.length + 1));
        var e = 0;
        for (var f = 0; f < this.a.length; f++) {
            if (this.a[f] != c) {
                d[e] = this.a[f];
                e++;
            }
        }
        d[e] = c;
        return d;
    };
    ButtonGroup.prototype.b = function (a, b, c) {
        var d = new Array(Math.max(0, this.a.length - 1));
        var e = 0;
        for (var f = 0; f < this.a.length; f++) {
            if (this.a[f] != c) {
                d[e] = this.a[f];
                e++;
            }
        }
        return d;
    };
    Object.defineProperty(ButtonGroup.prototype, "view", {
        get: function () {
            return this._view;
        },
        set: function (a) {
            this._view = a;
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroup.prototype.destroy = function () {
        this.provideContainer(null);
    };
    ButtonGroup.prototype.da = function (a, b, c) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
        this.dc(a, b, c);
    };
    ButtonGroup.prototype.dc = function (a, b, c) {
        switch (a) {
            case "BaseTheme":
                this.dh(this.g);
                this.dg();
                break;
            case "DisplayType":
            case "ItemBorderColor":
            case "ItemDisabledBorderColor":
            case "ItemBackgroundColor":
            case "ItemDisabledBackgroundColor":
            case "SelectedItemBackgroundColor":
            case "ItemTextColor":
            case "ItemDisabledTextColor":
            case "SelectedItemTextColor":
            case "ItemHoverBackgroundColor":
            case "SelectedItemHoverBackgroundColor":
            case "ItemHoverTextColor":
            case "SelectedItemHoverTextColor":
            case "OutlinedItemBorderColor":
            case "OutlinedItemDisabledBorderColor":
            case "OutlinedItemBackgroundColor":
            case "OutlinedItemDisabledBackgroundColor":
            case "OutlinedSelectedItemBackgroundColor":
            case "OutlinedItemTextColor":
            case "OutlinedItemDisabledTextColor":
            case "OutlinedSelectedItemTextColor":
            case "OutlinedItemHoverBackgroundColor":
            case ButtonGroup.cv:
            case "OutlinedItemHoverTextColor":
            case "OutlinedSelectedItemHoverTextColor":
            case "FlatItemBorderColor":
            case "FlatItemDisabledBorderColor":
            case "FlatItemBackgroundColor":
            case "FlatItemDisabledBackgroundColor":
            case "FlatSelectedItemBackgroundColor":
            case "FlatItemTextColor":
            case "FlatItemDisabledTextColor":
            case "FlatSelectedItemTextColor":
            case "FlatItemHoverBackgroundColor":
            case "FlatSelectedItemHoverBackgroundColor":
            case "FlatItemHoverTextColor":
            case "FlatSelectedItemHoverTextColor":
                this.dg();
                break;
            case "ActualItemBorderColor":
            case "ActualItemDisabledBorderColor":
            case "ActualItemDisabledBackgroundColor":
            case "ActualItemBackgroundColor":
            case "ActualItemTextColor":
            case "ActualItemDisabledTextColor":
            case "ActualSelectedItemTextColor":
            case "ActualSelectedItemBackgroundColor":
                this.df();
                break;
            case "Orientation":
                this.view.v();
                this.df();
                break;
            case "Density":
                this.j = this.k == 0 ? 1 : this.k;
                break;
            case "ActualDensity":
                for (var d = 0; d < this.buttons.count; d++) {
                    this.buttons._inner[d].h = this.j;
                }
                break;
            case "IsDisabled":
                for (var e = 0; e < this.buttons.count; e++) {
                    this.buttons._inner[e].disabled = this.disabled;
                }
                break;
            case "SelectedIndices":
                if (!this.u) {
                    this.df();
                }
                break;
        }
    };
    ButtonGroup.prototype.dl = function (a, b) {
        return new CornerRadius(1, !isNaN_(a.c) ? a.c : b.c, !isNaN_(a.d) ? a.d : b.d, !isNaN_(a.b) ? a.b : b.b, !isNaN_(a.a) ? a.a : b.a);
    };
    ButtonGroup.prototype.dg = function () {
        switch (this.p) {
            case 1:
                this.al = !isNaN_(this.a1) ? this.a1 : this.ax;
                this.ez = this.hn != null ? this.hn : this.gz;
                this.e1 = this.hp != null ? this.hp : this.g1;
                this.e6 = this.hu != null ? this.hu : this.g6;
                this.e3 = this.hr != null ? this.hr : this.g3;
                this.e0 = this.ho != null ? this.ho : this.g0;
                this.e2 = this.hq != null ? this.hq : this.g2;
                this.ak = !isNaN_(this.a0) ? this.a0 : this.aw;
                this.e7 = this.h7 != null ? this.h7 : this.g7;
                this.fa = this.ia != null ? this.ia : this.ha;
                this.e4 = this.hs != null ? this.hs : this.g4;
                this.e5 = this.ht != null ? this.ht : this.g5;
                this.e8 = this.h8 != null ? this.h8 : this.g8;
                this.e9 = this.h9 != null ? this.h9 : this.g9;
                break;
            case 0:
                this.al = !isNaN_(this.a1) ? this.a1 : this.ar;
                this.ez = this.hn != null ? this.hn : this.fz;
                this.e1 = this.hp != null ? this.hp : this.f1;
                this.e6 = this.hu != null ? this.hu : this.f6;
                this.e3 = this.hr != null ? this.hr : this.f3;
                this.e0 = this.ho != null ? this.ho : this.f0;
                this.e2 = this.hq != null ? this.hq : this.f2;
                this.ak = !isNaN_(this.a0) ? this.a0 : this.aq;
                this.e7 = this.h7 != null ? this.h7 : this.f7;
                this.fa = this.ia != null ? this.ia : this.ga;
                this.e4 = this.hs != null ? this.hs : this.f4;
                this.e5 = this.ht != null ? this.ht : this.f5;
                this.e8 = this.h8 != null ? this.h8 : this.f8;
                this.e9 = this.h9 != null ? this.h9 : this.f9;
                break;
        }
    };
    ButtonGroup.prototype.df = function () {
        for (var a = 0; a < this.buttons.count; a++) {
            var b = this.buttons._inner[a];
            b.h = this.j;
            if (a == 0) {
                this.di(b);
            }
            else if (a == this.buttons.count - 1) {
                this.dj(b);
            }
            else {
                this.dk(b);
            }
            b.s = 1;
            b.ak = false;
            if (this.v) {
                b.disabled = this.disabled;
            }
            b.nl = this.e0;
            b.ax = this.ak;
            b.sh = this.e2;
            b.sg = this.e1;
            b.si = this.e3;
            this.view.u(b);
            if (this.aa(b)) {
                b.nk = this.e7;
                b.t7 = this.fa;
                b.s7 = this.e8;
                b.s8 = this.e9;
            }
            else {
                b.nk = this.ez;
                b.t7 = this.e6;
                b.s7 = this.e4;
                b.s8 = this.e5;
            }
        }
    };
    ButtonGroup.prototype.aa = function (a) {
        var b = this.buttons.indexOf(a);
        for (var c = 0; c < this.d.length; c++) {
            if (this.d[c] == b) {
                return true;
            }
        }
        return false;
    };
    ButtonGroup.prototype.dk = function (a) {
        a.kd = new CornerRadius(1, 0, 0, 0, 0);
    };
    ButtonGroup.prototype.dj = function (a) {
        if (this.r == 1) {
            a.kd = new CornerRadius(1, 0, 0, this.al, this.al);
        }
        else {
            a.kd = new CornerRadius(1, 0, this.al, this.al, 0);
        }
    };
    ButtonGroup.prototype.di = function (a) {
        if (this.r == 1) {
            a.kd = new CornerRadius(1, this.al, this.al, 0, 0);
        }
        else {
            a.kd = new CornerRadius(1, this.al, 0, 0, this.al);
        }
    };
    ButtonGroup.prototype.dh = function (a) {
        this.ax = ThemeResolver.b(a, ButtonGroup.av, ButtonGroup.ap);
        this.gz = ThemeResolver.e(a, ButtonGroup.gn, ButtonGroup.fn);
        this.g1 = ThemeResolver.e(a, ButtonGroup.gp, ButtonGroup.fp);
        this.g0 = ThemeResolver.e(a, ButtonGroup.go, ButtonGroup.fo);
        this.g2 = ThemeResolver.e(a, ButtonGroup.gq, ButtonGroup.fq);
        this.aw = ThemeResolver.b(a, ButtonGroup.au, ButtonGroup.ao);
        this.g6 = ThemeResolver.e(a, ButtonGroup.gu, ButtonGroup.fu);
        this.g3 = ThemeResolver.e(a, ButtonGroup.gr, ButtonGroup.fr);
        this.g7 = ThemeResolver.e(a, ButtonGroup.gv, ButtonGroup.fv);
        this.ha = ThemeResolver.e(a, ButtonGroup.gy, ButtonGroup.fy);
        this.g4 = ThemeResolver.e(a, ButtonGroup.gs, ButtonGroup.fs);
        this.g5 = ThemeResolver.e(a, ButtonGroup.gt, ButtonGroup.ft);
        this.g8 = ThemeResolver.e(a, ButtonGroup.gw, ButtonGroup.fw);
        this.g9 = ThemeResolver.e(a, ButtonGroup.gx, ButtonGroup.fx);
        this.ar = ThemeResolver.b(a, ButtonGroup.at, ButtonGroup.an);
        this.fz = ThemeResolver.e(a, ButtonGroup.gb, ButtonGroup.fb);
        this.f1 = ThemeResolver.e(a, ButtonGroup.gd, ButtonGroup.fd);
        this.f0 = ThemeResolver.e(a, ButtonGroup.gc, ButtonGroup.fc);
        this.f2 = ThemeResolver.e(a, ButtonGroup.ge, ButtonGroup.fe);
        this.aq = ThemeResolver.b(a, ButtonGroup.as, ButtonGroup.am);
        this.f6 = ThemeResolver.e(a, ButtonGroup.gi, ButtonGroup.fi);
        this.f3 = ThemeResolver.e(a, ButtonGroup.gf, ButtonGroup.ff);
        this.f7 = ThemeResolver.e(a, ButtonGroup.gj, ButtonGroup.fj);
        this.ga = ThemeResolver.e(a, ButtonGroup.gm, ButtonGroup.fm);
        this.f4 = ThemeResolver.e(a, ButtonGroup.gg, ButtonGroup.fg);
        this.f5 = ThemeResolver.e(a, ButtonGroup.gh, ButtonGroup.fh);
        this.f8 = ThemeResolver.e(a, ButtonGroup.gk, ButtonGroup.fk);
        this.f9 = ThemeResolver.e(a, ButtonGroup.gl, ButtonGroup.fl);
    };
    Object.defineProperty(ButtonGroup.prototype, "al", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            var b = this.ad;
            this.ad = a;
            if (b != this.ad) {
                this.da("ActualItemCornerRadius", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "ez", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            var b = this.dm;
            this.dm = a;
            if (b != this.dm) {
                this.da("ActualItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e1", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            var b = this.dp;
            this.dp = a;
            if (b != this.dp) {
                this.da("ActualItemDisabledBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e6", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            var b = this.du;
            this.du = a;
            if (b != this.du) {
                this.da("ActualItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e3", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            var b = this.dr;
            this.dr = a;
            if (b != this.dr) {
                this.da("ActualItemDisabledTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e0", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            var b = this.dn;
            this.dn = a;
            if (b != this.dn) {
                this.da("ActualItemBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e2", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            var b = this.dq;
            this.dq = a;
            if (b != this.dq) {
                this.da("ActualItemDisabledBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "ak", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            var b = this.ac;
            this.ac = a;
            if (b != this.ac) {
                this.da("ActualItemBorderWidth", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e7", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            var b = this.dv;
            this.dv = a;
            if (b != this.dv) {
                this.da("ActualSelectedItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "fa", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            var b = this.dy;
            this.dy = a;
            if (b != this.dy) {
                this.da("ActualSelectedItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e4", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            var b = this.ds;
            this.ds = a;
            if (b != this.ds) {
                this.da("ActualItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e5", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            var b = this.dt;
            this.dt = a;
            if (b != this.dt) {
                this.da("ActualItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e8", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            var b = this.dw;
            this.dw = a;
            if (b != this.dw) {
                this.da("ActualSelectedItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "e9", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            var b = this.dx;
            this.dx = a;
            if (b != this.dx) {
                this.da("ActualSelectedItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "a1", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            var b = this.ah;
            this.ah = a;
            if (b != this.ah) {
                this.da("ItemCornerRadius", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hn", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            var b = this.d7;
            this.d7 = a;
            if (b != this.d7) {
                this.da("ItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hp", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            var b = this.d9;
            this.d9 = a;
            if (b != this.d9) {
                this.da("ItemDisabledBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hu", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            var b = this.ee;
            this.ee = a;
            if (b != this.ee) {
                this.da("ItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hr", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            var b = this.eb;
            this.eb = a;
            if (b != this.eb) {
                this.da("ItemDisabledTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hs", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            var b = this.ec;
            this.ec = a;
            if (b != this.ec) {
                this.da("ItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "ho", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            var b = this.d8;
            this.d8 = a;
            if (b != this.d8) {
                this.da("ItemBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hq", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            var b = this.ea;
            this.ea = a;
            if (b != this.ea) {
                this.da("ItemDisabledBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "a0", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            var b = this.ag;
            this.ag = a;
            if (b != this.ag) {
                this.da("ItemBorderWidth", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "ht", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            var b = this.ed;
            this.ed = a;
            if (b != this.ed) {
                this.da("ItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h7", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            var b = this.ev;
            this.ev = a;
            if (b != this.ev) {
                this.da("SelectedItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "ia", {
        get: function () {
            return this.ey;
        },
        set: function (a) {
            var b = this.ey;
            this.ey = a;
            if (b != this.ey) {
                this.da("SelectedItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h8", {
        get: function () {
            return this.ew;
        },
        set: function (a) {
            var b = this.ew;
            this.ew = a;
            if (b != this.ew) {
                this.da("SelectedItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h9", {
        get: function () {
            return this.ex;
        },
        set: function (a) {
            var b = this.ex;
            this.ex = a;
            if (b != this.ex) {
                this.da("SelectedItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "a3", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            var b = this.aj;
            this.aj = a;
            if (b != this.aj) {
                this.da("OutlinedItemCornerRadius", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hv", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            var b = this.ef;
            this.ef = a;
            if (b != this.ef) {
                this.da("OutlinedItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hx", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            var b = this.eh;
            this.eh = a;
            if (b != this.eh) {
                this.da("OutlinedItemDisabledBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h2", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            var b = this.em;
            this.em = a;
            if (b != this.em) {
                this.da("OutlinedItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hz", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            var b = this.ej;
            this.ej = a;
            if (b != this.ej) {
                this.da("OutlinedItemDisabledTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h0", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            var b = this.ek;
            this.ek = a;
            if (b != this.ek) {
                this.da("OutlinedItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hw", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            var b = this.eg;
            this.eg = a;
            if (b != this.eg) {
                this.da("OutlinedItemBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hy", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            var b = this.ei;
            this.ei = a;
            if (b != this.ei) {
                this.da("OutlinedItemDisabledBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "a2", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            var b = this.ai;
            this.ai = a;
            if (b != this.ai) {
                this.da("OutlinedItemBorderWidth", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h1", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            var b = this.el;
            this.el = a;
            if (b != this.el) {
                this.da("OutlinedItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h3", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            var b = this.er;
            this.er = a;
            if (b != this.er) {
                this.da("OutlinedSelectedItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h6", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            var b = this.eu;
            this.eu = a;
            if (b != this.eu) {
                this.da("OutlinedSelectedItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h4", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            var b = this.es;
            this.es = a;
            if (b != this.es) {
                this.da(ButtonGroup.cv, b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "h5", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            var b = this.et;
            this.et = a;
            if (b != this.et) {
                this.da("OutlinedSelectedItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "az", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            var b = this.af;
            this.af = a;
            if (b != this.af) {
                this.da("FlatItemCornerRadius", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hb", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            var b = this.dz;
            this.dz = a;
            if (b != this.dz) {
                this.da("FlatItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hd", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            var b = this.d1;
            this.d1 = a;
            if (b != this.d1) {
                this.da("FlatItemDisabledBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hi", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            var b = this.d6;
            this.d6 = a;
            if (b != this.d6) {
                this.da("FlatItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hf", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            var b = this.d3;
            this.d3 = a;
            if (b != this.d3) {
                this.da("FlatItemDisabledTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hg", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            var b = this.d4;
            this.d4 = a;
            if (b != this.d4) {
                this.da("FlatItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hc", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            var b = this.d0;
            this.d0 = a;
            if (b != this.d0) {
                this.da("FlatItemBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "he", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            var b = this.d2;
            this.d2 = a;
            if (b != this.d2) {
                this.da("FlatItemDisabledBorderColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "ay", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            var b = this.ae;
            this.ae = a;
            if (b != this.ae) {
                this.da("FlatItemBorderWidth", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hh", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            var b = this.d5;
            this.d5 = a;
            if (b != this.d5) {
                this.da("FlatItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hj", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            var b = this.en;
            this.en = a;
            if (b != this.en) {
                this.da("FlatSelectedItemBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hm", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            var b = this.eq;
            this.eq = a;
            if (b != this.eq) {
                this.da("FlatSelectedItemTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hk", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            var b = this.eo;
            this.eo = a;
            if (b != this.eo) {
                this.da("FlatSelectedItemHoverBackgroundColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "hl", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            var b = this.ep;
            this.ep = a;
            if (b != this.ep) {
                this.da("FlatSelectedItemHoverTextColor", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "r", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            var b = this.q;
            this.q = a;
            if (b != this.q) {
                this.da("Orientation", enumGetBox(ButtonGroupOrientation_$type, b), enumGetBox(ButtonGroupOrientation_$type, a));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "b5", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            var b = this.a6;
            this.a6 = a;
            if (b != this.a6) {
                this.da("Id", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "disabled", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            var b = this.w;
            this.w = a;
            this.v = true;
            if (b != this.w) {
                this.da("IsDisabled", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroup.prototype, "z", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            var b = this.x;
            this.x = a;
            if (b != this.x) {
                this.da("IsMultiSelect", b, a);
            }
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroup.prototype.db = function () {
    };
    ButtonGroup.prototype.onDetachedFromUI = function () {
        this.view.r();
    };
    ButtonGroup.prototype.onAttachedToUI = function () {
        this.view.p();
    };
    ButtonGroup.prototype.ab = function () {
        return this.view.d();
    };
    ButtonGroup.prototype.a5 = function () {
        return this.l();
    };
    ButtonGroup.prototype.bq = function () {
        var a = this.l();
        return a.l();
    };
    ButtonGroup.prototype.l = function () {
        var a = new ButtonGroupVisualModelExport();
        a.e = this.g;
        a.f = this.j;
        a.a = this.d;
        a.k = this.al;
        a.b = AppearanceHelper.a(this.ez);
        a.d = AppearanceHelper.a(this.e6);
        a.c = AppearanceHelper.a(this.e0);
        a.j = this.ak;
        a.g = this.disabled;
        a.h = this.z;
        a.i = new List$1(ButtonVisualModelExport.$, 0);
        for (var b = 0; b < this.buttons.count; b++) {
            a.i.add1(this.buttons._inner[b].ey());
        }
        return a;
    };
    ButtonGroup.prototype.provideContainer = function (a) {
        this.view.q(a);
    };
    Object.defineProperty(ButtonGroup.prototype, "buttons", {
        get: function () {
            return this.m;
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroup.$t = markType(ButtonGroup, 'ButtonGroup', Base.$, [INotifyPropertyChanged_$type]);
    ButtonGroup.f = 2;
    ButtonGroup.av = 4;
    ButtonGroup.gn = BrushUtil.g(255, 255, 255, 255);
    ButtonGroup.gp = BrushUtil.g(10, 0, 0, 0);
    ButtonGroup.go = BrushUtil.g(255, 199, 199, 199);
    ButtonGroup.gq = BrushUtil.g(255, 199, 199, 199);
    ButtonGroup.au = 1;
    ButtonGroup.gu = BrushUtil.g(255, 0, 0, 0);
    ButtonGroup.gr = BrushUtil.g(66, 0, 0, 0);
    ButtonGroup.gv = BrushUtil.g(60, 0, 0, 0);
    ButtonGroup.gy = BrushUtil.g(255, 0, 0, 0);
    ButtonGroup.gs = BrushUtil.g(30, 0, 0, 0);
    ButtonGroup.gt = BrushUtil.g(255, 0, 0, 0);
    ButtonGroup.gw = BrushUtil.g(96, 0, 0, 0);
    ButtonGroup.gx = BrushUtil.g(255, 0, 0, 0);
    ButtonGroup.ap = 3;
    ButtonGroup.fn = BrushUtil.g(0, 255, 255, 255);
    ButtonGroup.fp = BrushUtil.g(10, 0, 0, 0);
    ButtonGroup.fo = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.fq = BrushUtil.g(255, 199, 199, 199);
    ButtonGroup.ao = 1;
    ButtonGroup.fu = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.fr = BrushUtil.g(66, 0, 0, 0);
    ButtonGroup.fv = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.fy = BrushUtil.g(255, 255, 255, 255);
    ButtonGroup.fs = BrushUtil.g(26, 33, 150, 243);
    ButtonGroup.ft = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.fw = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.fx = BrushUtil.g(255, 255, 255, 255);
    ButtonGroup.at = 4;
    ButtonGroup.gb = BrushUtil.g(0, 255, 255, 255);
    ButtonGroup.gd = BrushUtil.g(10, 0, 0, 0);
    ButtonGroup.gc = BrushUtil.g(0, 0, 0, 0);
    ButtonGroup.ge = BrushUtil.g(255, 199, 199, 199);
    ButtonGroup.as = 0;
    ButtonGroup.gi = BrushUtil.g(255, 228, 28, 119);
    ButtonGroup.gf = BrushUtil.g(66, 0, 0, 0);
    ButtonGroup.gj = BrushUtil.g(30, 228, 28, 119);
    ButtonGroup.gm = BrushUtil.g(255, 228, 28, 119);
    ButtonGroup.gg = BrushUtil.g(13, 228, 28, 119);
    ButtonGroup.gh = BrushUtil.g(255, 228, 28, 119);
    ButtonGroup.gk = BrushUtil.g(30, 228, 28, 119);
    ButtonGroup.gl = BrushUtil.g(255, 228, 28, 119);
    ButtonGroup.an = 0;
    ButtonGroup.fb = BrushUtil.g(0, 255, 255, 255);
    ButtonGroup.fd = BrushUtil.g(10, 0, 0, 0);
    ButtonGroup.fc = BrushUtil.g(0, 0, 0, 0);
    ButtonGroup.fe = BrushUtil.g(0, 0, 0, 0);
    ButtonGroup.am = 0;
    ButtonGroup.fi = BrushUtil.g(255, 119, 119, 119);
    ButtonGroup.ff = BrushUtil.g(66, 0, 0, 0);
    ButtonGroup.fj = BrushUtil.g(26, 33, 150, 243);
    ButtonGroup.fm = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.fg = BrushUtil.g(10, 33, 150, 243);
    ButtonGroup.fh = BrushUtil.g(255, 119, 119, 119);
    ButtonGroup.fk = BrushUtil.g(26, 33, 150, 243);
    ButtonGroup.fl = BrushUtil.g(255, 33, 150, 243);
    ButtonGroup.cv = "OutlinedSelectedItemHoverBackgroundColor";
    ButtonGroup.a4 = 0;
    return ButtonGroup;
}(Base));
export { ButtonGroup };
