// igniteui-react-inputs/esm2015/lib/ButtonView_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Delegate_$type, fromEnum, runOn, markType, INotifyPropertyChanged_$type, PropertyChangedEventArgs, enumGetBox } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { Ripple } from "./Ripple";
import { NativeUI } from "igniteui-react-core";
import { ControlDisplayDensity_$type } from "igniteui-react-core";
import { KeyFrameAnimationFactory } from "igniteui-react-core";
import { ElevationMode_$type } from "igniteui-react-core";
import { ElevationHelper } from "./ElevationHelper";
import { ButtonDisplayStyle_$type } from "./ButtonDisplayStyle";
import { isNaN_ } from "igniteui-react-core";
import { BaseControlTheme_$type } from "igniteui-react-core";
import { CornerRadius } from "igniteui-react-core";
import { AnimationKeyFrameEffectTiming } from "igniteui-react-core";
import { ThemeResolver } from "./ThemeResolver";
import { ButtonClickEventArgs } from "./ButtonClickEventArgs";
import { Thickness } from "igniteui-react-core";
import { BrushUtil } from "igniteui-react-core";
import { ButtonVisualModelExport } from "./ButtonVisualModelExport";
import { AppearanceHelper } from "igniteui-react-core";
import { FontUtil } from "igniteui-react-core";
/**
 * @hidden
 */
export let ButtonView = /*@__PURE__*/ (() => {
    class ButtonView extends Base {
        constructor() {
            super();
            this.a = null;
            this.j = null;
            this.c = null;
            this.g = new List$1(Delegate_$type, 0);
            this.d = true;
            this.m = null;
            this.i = NaN;
            this.h = NaN;
            this.k = null;
            this.b = null;
            this.l = null;
            this.o = null;
            this.n = null;
        }
        ai() {
        }
        f() {
            return true;
        }
        get p() {
            return this.m;
        }
        get q() {
            return this.l;
        }
        ac(a) {
            if (a == null) {
                for (let b of fromEnum(this.g)) {
                    b();
                }
                this.g.clear();
                return;
            }
            this.j = a;
            this.j.rootWrapper.setStyleProperty("display", "flex");
            this.j.rootWrapper.setStyleProperty("flex-direction", "row");
            this.j.rootWrapper.setStyleProperty("flex-wrap", "no-wrap");
            this.j.rootWrapper.setStyleProperty("align-items", "center");
            let c = this.j.createElement("button");
            this.m = c;
            this.m.setStyleProperty("position", "relative");
            this.m.setStyleProperty("display", "inline-flex");
            this.m.setStyleProperty("align-items", "center");
            this.m.setStyleProperty("justify-content", "center");
            this.ar();
            this.m.setStyleProperty("border", "none");
            this.m.setStyleProperty("cursor", "pointer");
            this.m.setStyleProperty("user-select", "none");
            this.m.setStyleProperty("outline-style", "none");
            this.m.setStyleProperty("-webkit-tap-highlight-color", "transparent");
            this.m.setStyleProperty("overflow", "hidden");
            this.m.setStyleProperty("white-space", "no-wrap");
            this.m.setStyleProperty("color", "#000");
            this.m.setStyleProperty("background", this.a.m7._fill);
            NativeUI.l(this.m, this.a.kb);
            this.m.setAttribute("id", this.a.h7);
            this.ao();
            this.g.add(c.listen("change", runOn(this, this.v)));
            this.g.add(c.listen("click", runOn(this, this.w)));
            this.g.add(c.listen("focus", runOn(this, this.x)));
            this.g.add(c.listen("blur", runOn(this, this.u)));
            this.g.add(this.j.rootWrapper.listen("mouseenter", runOn(this, this.s)));
            this.g.add(this.j.rootWrapper.listen("mouseleave", runOn(this, this.t)));
            this.j.append(c);
            let d = this.j.createElement("div");
            let e = this.j.getSubRenderer(d);
            d.setStyleProperty("position", "absolute");
            d.setStyleProperty("width", "100%");
            d.setStyleProperty("height", "100%");
            d.setStyleProperty("left", "0px");
            d.setStyleProperty("top", "0px");
            d.setStyleProperty("overflow", "hidden");
            let f = new Ripple();
            f.provideContainer(e);
            f.q = c;
            let g = this.j.getExternal(f, this.j.rootWrapper, this.j.getExternal(this.a, null, null));
            f.h = false;
            f.k = true;
            this.b = f;
            this.m.append(d);
            let h = this.j.createElement("div");
            h.addClass("ig-button-content");
            h.setStyleProperty("font-family", "'Segoe UI', Verdana");
            h.setStyleProperty("font-weight", "400");
            h.setStyleProperty("font-size", ".875rem");
            h.setStyleProperty("cursor", "pointer");
            this.l = h;
            this.g.add(h.listen("click", runOn(this, this.r)));
            this.m.append(h);
            this.n = h;
            this.j.getPortal(h, "ChildContent", (i) => {
                let comp_ = i.componentRef;
                this.o = comp_;
                {
                    this.n.setStyleProperty("flex-direction", this.a.he);
                    this.n.setStyleProperty("align-items", this.a.fv);
                    let direction_ = this.a.he;
                    let display_ = this.a.gb;
                    let alignItems_ = this.a.fv;
                    comp_.display = display_;
                    comp_.flexDirection = direction_;
                    comp_.alignItems = alignItems_;
                }
            }, true);
            this.a.jy();
        }
        ae() {
            this.b.i = this.a.ae;
        }
        ar() {
            let a = "2.25rem";
            let b = "2.25rem";
            let c = "2.25rem";
            let d = "2.25rem";
            let e = "3.5rem";
            let f = "3.5rem";
            let g = "3.5rem";
            let h = "3.5rem";
            let i = "5.5rem";
            let j = "2.25rem";
            let k = "0px";
            let l = ".9375rem";
            let m = ".5625rem 1rem";
            switch (this.a.g) {
                case 2:
                    m = ".375rem .75rem";
                    j = "1.875rem";
                    e = "3rem";
                    f = "3rem";
                    g = "3rem";
                    h = "3rem";
                    l = ".6875rem";
                    break;
                case 3:
                    m = ".1875rem 1rem";
                    j = "1.5rem";
                    f = "2.5rem";
                    e = "2.5rem";
                    h = "2.5rem";
                    g = "2.5rem";
                    l = ".4375rem";
                    break;
            }
            if (this.a.s == 4) {
                this.m.setStyleProperty("min-width", a);
                this.m.setStyleProperty("min-height", b);
                this.m.setStyleProperty("padding", k);
                this.m.setStyleProperty("width", c);
                this.m.setStyleProperty("height", d);
            }
            else if (this.a.s == 3) {
                this.m.setStyleProperty("min-width", e);
                this.m.setStyleProperty("min-height", f);
                this.m.setStyleProperty("padding", l);
                this.m.setStyleProperty("width", g);
                this.m.setStyleProperty("height", h);
            }
            else {
                this.m.setStyleProperty("min-width", i);
                this.m.setStyleProperty("min-height", j);
                this.m.setStyleProperty("padding", m);
            }
            if (!isNaN_(this.a.aw) && this.a.aw > 0) {
                this.m.setStyleProperty("border-style", "solid");
            }
            else {
                this.m.setStyleProperty("border-style", "none");
            }
            if (!isNaN_(this.i)) {
                this.m.setStyleProperty("min-width", this.i.toString() + "px");
            }
            if (!isNaN_(this.h)) {
                this.m.setStyleProperty("min-height", this.h.toString() + "px");
            }
        }
        an(a) {
            this.i = a;
        }
        am(a) {
            this.h = a;
        }
        ak() {
            this.m.setAttribute("disabled", this.a.disabled ? "true" : "false");
            this.m.setProperty("disabled", this.a.disabled);
        }
        af() {
            if (this.j != null) {
                this.j.rootWrapper.setStyleProperty("display", this.a.gb);
                let comp_ = this.o;
                if (this.o != null) {
                    this.n.setStyleProperty("display", this.a.gb);
                    let display_ = this.a.gb;
                    comp_.display = display_;
                }
                let ext_ = this.j.getExternal(this.a, null, null);
                if (ext_ && ext_._onDisplayChanged) {
                    ext_._onDisplayChanged();
                }
                ;
            }
        }
        aa() {
            if (this.j != null) {
                this.j.rootWrapper.setStyleProperty("align-items", this.a.fv);
                this.m.setStyleProperty("align-items", this.a.fv);
                let comp_ = this.o;
                if (this.o != null) {
                    this.n.setStyleProperty("align-items", this.a.fv);
                    let alignItems_ = this.a.fv;
                    comp_.alignItems = alignItems_;
                }
                let ext_ = this.j.getExternal(this.a, null, null);
                if (ext_ && ext_._onAlignItemsChanged) {
                    ext_._onAlignItemsChanged();
                }
                ;
            }
        }
        ag() {
            if (this.j != null) {
                this.j.rootWrapper.setStyleProperty("flex-direction", this.a.he);
                this.m.setStyleProperty("flex-direction", this.a.he);
                let ext_ = this.j.getExternal(this.a, null, null);
                let comp_ = this.o;
                if (this.o != null) {
                    this.n.setStyleProperty("flex-direction", this.a.he);
                    let direction_ = this.a.he;
                    comp_.flexDirection = direction_;
                }
                if (ext_ && ext_._onFlexDirectionChanged) {
                    ext_._onFlexDirectionChanged();
                }
                ;
            }
        }
        ah() {
            if (this.j != null) {
                this.j.rootWrapper.setStyleProperty("flex-grow", this.a.hg);
                this.m.setStyleProperty("flex-grow", this.a.hg);
                let ext_ = this.j.getExternal(this.a, null, null);
                if (ext_ && ext_._onFlexGrowChanged) {
                    ext_._onFlexGrowChanged();
                }
                ;
            }
        }
        r(a) {
        }
        ao() {
            if (this.f()) {
                let a = this.a.h7;
                this.m.setAttribute("name", this.a.ig);
                this.m.setProperty("value", this.a.ez);
                if (this.a.ah) {
                    this.m.setAttribute("tabindex", this.a.eu.toString());
                }
                this.m.setProperty("disabled", this.a.disabled);
                {
                }
                this.m.setAttribute("aria-label", this.a.fx);
            }
        }
        aj() {
            if (this.f()) {
                let a = this.a.h7;
                this.m.setAttribute("id", a);
            }
        }
        s(a) {
            this.a.jv();
        }
        t(a) {
            this.a.jw();
        }
        u(a) {
            this.a.jr();
        }
        x(a) {
            this.a.ju();
        }
        w(a) {
            this.a.js();
        }
        e() {
            return false;
        }
        v(a) {
        }
        z() {
            if (KeyFrameAnimationFactory.h == null) {
                KeyFrameAnimationFactory.h = this.j;
            }
        }
        ab() {
            this.d = true;
            this.g.add(this.m.listen("change", runOn(this, this.v)));
            this.g.add(this.m.listen("click", runOn(this, this.w)));
            this.g.add(this.m.listen("focus", runOn(this, this.x)));
            this.g.add(this.m.listen("blur", runOn(this, this.u)));
            this.g.add(this.l.listen("click", runOn(this, this.r)));
            this.g.add(this.m.listen("mouseenter", runOn(this, this.s)));
            this.g.add(this.m.listen("mouseleave", runOn(this, this.t)));
        }
        ad() {
            this.d = false;
            for (let a of fromEnum(this.g)) {
                a();
            }
            this.g.clear();
        }
        al() {
            let model_ = this.a;
            if (model_.externalObject && model_.externalObject.refreshDefaults) {
                model_.externalObject.refreshDefaults();
            }
            ;
        }
        ap(a, b) {
            this.b.ax = a;
            this.b.aw = b;
        }
        aq(a, b) {
            let c = ElevationHelper.e(a, this.a.nj, this.a.ng, this.a.m6, b);
            this.m.setStyleProperty("box-shadow", c);
        }
        as(a) {
            if (a != null) {
                this.l.setStyleProperty("font", a.fontString);
            }
        }
        y(a) {
            this.l.append(a);
        }
    }
    ButtonView.$t = /*@__PURE__*/ markType(ButtonView, 'ButtonView');
    return ButtonView;
})();
/**
 * @hidden
 */
export let Button = /*@__PURE__*/ (() => {
    class Button extends Base {
        constructor() {
            super();
            this._view = null;
            this.sf = ThemeResolver.e(Button.c, Button.rr, Button.o8);
            this.r3 = ThemeResolver.e(Button.c, Button.rf, Button.ow);
            this.nm = ThemeResolver.e(Button.c, Button.p6, Button.nn);
            this.m = ThemeResolver.a(Button.c, Button.n, Button.l);
            this.r6 = ThemeResolver.e(Button.c, Button.ri, Button.oz);
            this.r7 = ThemeResolver.e(Button.c, Button.rj, Button.o0);
            this.r8 = ThemeResolver.e(Button.c, Button.rk, Button.o1);
            this.d2 = ThemeResolver.c(Button.c, Button.du, Button.cx);
            this.d5 = ThemeResolver.c(Button.c, Button.dx, Button.c0);
            this.d4 = ThemeResolver.c(Button.c, Button.dw, Button.cz);
            this.d3 = ThemeResolver.c(Button.c, Button.dv, Button.cy);
            this.r5 = ThemeResolver.e(Button.c, Button.rh, Button.oy);
            this.sd = ThemeResolver.e(Button.c, Button.rp, Button.o6);
            this.ks = ThemeResolver.d(Button.c, Button.kq, Button.ki);
            this.bc = ThemeResolver.b(Button.c, Button.ba, Button.a2);
            this.r4 = ThemeResolver.e(Button.c, Button.rg, Button.ox);
            this.se = ThemeResolver.e(Button.c, Button.rq, Button.o7);
            this.sc = ThemeResolver.e(Button.c, Button.ro, Button.o5);
            this.sb = ThemeResolver.e(Button.c, Button.rn, Button.o4);
            this.r9 = ThemeResolver.e(Button.c, Button.rl, Button.o2);
            this.sa = ThemeResolver.e(Button.c, Button.rm, Button.o3);
            this.pm = ThemeResolver.e(Button.c, Button.qk, Button.n1);
            this.pn = ThemeResolver.e(Button.c, Button.ql, Button.n2);
            this.po = ThemeResolver.e(Button.c, Button.qm, Button.n3);
            this.c5 = ThemeResolver.c(Button.c, Button.dh, Button.cl);
            this.c8 = ThemeResolver.c(Button.c, Button.dk, Button.co);
            this.c7 = ThemeResolver.c(Button.c, Button.dj, Button.cn);
            this.c6 = ThemeResolver.c(Button.c, Button.di, Button.cm);
            this.pl = ThemeResolver.e(Button.c, Button.qj, Button.n0);
            this.pt = ThemeResolver.e(Button.c, Button.qr, Button.n8);
            this.kk = ThemeResolver.d(Button.c, Button.kn, Button.kf);
            this.a4 = ThemeResolver.b(Button.c, Button.a7, Button.az);
            this.pk = ThemeResolver.e(Button.c, Button.qi, Button.nz);
            this.pu = ThemeResolver.e(Button.c, Button.qs, Button.n9);
            this.ps = ThemeResolver.e(Button.c, Button.qq, Button.n7);
            this.pr = ThemeResolver.e(Button.c, Button.qp, Button.n6);
            this.pp = ThemeResolver.e(Button.c, Button.qn, Button.n4);
            this.pq = ThemeResolver.e(Button.c, Button.qo, Button.n5);
            this.ru = ThemeResolver.e(Button.c, Button.q6, Button.on);
            this.rv = ThemeResolver.e(Button.c, Button.q7, Button.oo);
            this.rw = ThemeResolver.e(Button.c, Button.q8, Button.op);
            this.dy = ThemeResolver.c(Button.c, Button.dq, Button.ct);
            this.d1 = ThemeResolver.c(Button.c, Button.dt, Button.cw);
            this.d0 = ThemeResolver.c(Button.c, Button.ds, Button.cv);
            this.dz = ThemeResolver.c(Button.c, Button.dr, Button.cu);
            this.rt = ThemeResolver.e(Button.c, Button.q5, Button.om);
            this.r1 = ThemeResolver.e(Button.c, Button.rd, Button.ou);
            this.kr = ThemeResolver.d(Button.c, Button.kp, Button.kh);
            this.bb = ThemeResolver.b(Button.c, Button.a9, Button.a1);
            this.rs = ThemeResolver.e(Button.c, Button.q4, Button.ol);
            this.r2 = ThemeResolver.e(Button.c, Button.re, Button.ov);
            this.r0 = ThemeResolver.e(Button.c, Button.rc, Button.ot);
            this.rz = ThemeResolver.e(Button.c, Button.rb, Button.os);
            this.rx = ThemeResolver.e(Button.c, Button.q9, Button.oq);
            this.ry = ThemeResolver.e(Button.c, Button.ra, Button.or);
            this.px = ThemeResolver.e(Button.c, Button.qv, Button.oc);
            this.py = ThemeResolver.e(Button.c, Button.qw, Button.od);
            this.pz = ThemeResolver.e(Button.c, Button.qx, Button.oe);
            this.c9 = ThemeResolver.c(Button.c, Button.dl, Button.cp);
            this.dc = ThemeResolver.c(Button.c, Button.dp, Button.cs);
            this.db = ThemeResolver.c(Button.c, Button.dn, Button.cr);
            this.da = ThemeResolver.c(Button.c, Button.dm, Button.cq);
            this.pw = ThemeResolver.e(Button.c, Button.qu, Button.ob);
            this.p4 = ThemeResolver.e(Button.c, Button.q2, Button.oj);
            this.kl = ThemeResolver.d(Button.c, Button.ko, Button.kg);
            this.a5 = ThemeResolver.b(Button.c, Button.a8, Button.a0);
            this.pv = ThemeResolver.e(Button.c, Button.qt, Button.oa);
            this.p5 = ThemeResolver.e(Button.c, Button.q3, Button.ok);
            this.p3 = ThemeResolver.e(Button.c, Button.q1, Button.oi);
            this.p2 = ThemeResolver.e(Button.c, Button.q0, Button.oh);
            this.p0 = ThemeResolver.e(Button.c, Button.qy, Button.of);
            this.p1 = ThemeResolver.e(Button.c, Button.qz, Button.og);
            this.pb = ThemeResolver.e(Button.c, Button.p9, Button.nq);
            this.pc = ThemeResolver.e(Button.c, Button.qa, Button.nr);
            this.pd = ThemeResolver.e(Button.c, Button.qb, Button.ns);
            this.c1 = ThemeResolver.c(Button.c, Button.dd, Button.ch);
            this.c4 = ThemeResolver.c(Button.c, Button.dg, Button.ck);
            this.c3 = ThemeResolver.c(Button.c, Button.df, Button.cj);
            this.c2 = ThemeResolver.c(Button.c, Button.de, Button.ci);
            this.pa = ThemeResolver.e(Button.c, Button.p8, Button.np);
            this.pi = ThemeResolver.e(Button.c, Button.qg, Button.nx);
            this.kj = ThemeResolver.d(Button.c, Button.km, Button.ke);
            this.a3 = ThemeResolver.b(Button.c, Button.a6, Button.ay);
            this.o9 = ThemeResolver.e(Button.c, Button.p7, Button.no);
            this.pj = ThemeResolver.e(Button.c, Button.qh, Button.ny);
            this.ph = ThemeResolver.e(Button.c, Button.qf, Button.nw);
            this.pg = ThemeResolver.e(Button.c, Button.qe, Button.nv);
            this.pe = ThemeResolver.e(Button.c, Button.qc, Button.nt);
            this.pf = ThemeResolver.e(Button.c, Button.qd, Button.nu);
            this.at = NaN;
            this.as = NaN;
            this.r = 0;
            this.f = 0;
            this.b = 0;
            this.e = 1;
            this.k0 = ThemeResolver.e(Button.c, Button.rh, Button.oy);
            this.lb = ThemeResolver.e(Button.c, Button.rr, Button.o8);
            this.k8 = ThemeResolver.e(Button.c, Button.rf, Button.ow);
            this.ky = ThemeResolver.e(Button.c, Button.p6, Button.nn);
            this.bn = ThemeResolver.c(Button.c, Button.dx, Button.c0);
            this.bl = ThemeResolver.c(Button.c, Button.dv, Button.cy);
            this.bm = ThemeResolver.c(Button.c, Button.dw, Button.cz);
            this.k9 = ThemeResolver.e(Button.c, Button.rp, Button.o6);
            this.j4 = ThemeResolver.d(Button.c, Button.kq, Button.ki);
            this.kz = ThemeResolver.e(Button.c, Button.rg, Button.ox);
            this.k1 = ThemeResolver.e(Button.c, Button.ri, Button.oz);
            this.k2 = ThemeResolver.e(Button.c, Button.rj, Button.o0);
            this.k3 = ThemeResolver.e(Button.c, Button.rk, Button.o1);
            this.bk = ThemeResolver.c(Button.c, Button.du, Button.cx);
            this.i = ThemeResolver.a(Button.c, Button.n, Button.l);
            this.la = ThemeResolver.e(Button.c, Button.rq, Button.o7);
            this.k7 = ThemeResolver.e(Button.c, Button.ro, Button.o5);
            this.k4 = ThemeResolver.e(Button.c, Button.rl, Button.o2);
            this.k5 = ThemeResolver.e(Button.c, Button.rm, Button.o3);
            this.k6 = ThemeResolver.e(Button.c, Button.rn, Button.o4);
            this.an = ThemeResolver.b(Button.c, Button.ba, Button.a2);
            this.mt = null;
            this.mv = null;
            this.mw = null;
            this.mx = null;
            this.b7 = -1;
            this.m0 = null;
            this.mu = null;
            this.ca = -1;
            this.b9 = -1;
            this.b8 = -1;
            this.av = NaN;
            this.ka = new CornerRadius(1, NaN, NaN, NaN, NaN);
            this.m2 = null;
            this.lr = null;
            this.lt = null;
            this.lu = null;
            this.lv = null;
            this.bt = -1;
            this.ly = null;
            this.ls = null;
            this.bw = -1;
            this.bv = -1;
            this.bu = -1;
            this.l0 = null;
            this.l1 = null;
            this.lz = null;
            this.lw = null;
            this.lx = null;
            this.aq = NaN;
            this.j7 = new CornerRadius(1, NaN, NaN, NaN, NaN);
            this.lc = null;
            this.le = null;
            this.lf = null;
            this.lg = null;
            this.bo = -1;
            this.j = 0;
            this.l4 = null;
            this.ld = null;
            this.cb = -1;
            this.by = -1;
            this.bx = -1;
            this.u = null;
            this.m5 = null;
            this.l6 = null;
            this.m4 = null;
            this.m3 = null;
            this.m1 = null;
            this.my = null;
            this.mz = null;
            this.l2 = null;
            this.l3 = null;
            this.ao = NaN;
            this.j5 = new CornerRadius(1, NaN, NaN, NaN, NaN);
            this.mi = null;
            this.mk = null;
            this.ml = null;
            this.mm = null;
            this.b3 = -1;
            this.mp = null;
            this.mj = null;
            this.b6 = -1;
            this.b5 = -1;
            this.b4 = -1;
            this.au = NaN;
            this.j9 = new CornerRadius(1, NaN, NaN, NaN, NaN);
            this.mr = null;
            this.ms = null;
            this.mq = null;
            this.mn = null;
            this.mo = null;
            this.lh = null;
            this.lj = null;
            this.lk = null;
            this.ll = null;
            this.bp = -1;
            this.l5 = null;
            this.li = null;
            this.bs = -1;
            this.br = -1;
            this.bq = -1;
            this.ap = NaN;
            this.j6 = new CornerRadius(1, NaN, NaN, NaN, NaN);
            this.lp = null;
            this.lq = null;
            this.lo = null;
            this.lm = null;
            this.ln = null;
            this.l7 = null;
            this.l9 = null;
            this.ma = null;
            this.mb = null;
            this.bz = -1;
            this.me = null;
            this.l8 = null;
            this.b2 = -1;
            this.b1 = -1;
            this.b0 = -1;
            this.ar = NaN;
            this.j8 = new CornerRadius(1, NaN, NaN, NaN, NaN);
            this.mg = null;
            this.mh = null;
            this.mf = null;
            this.mc = null;
            this.md = null;
            this.e6 = "ig-native-button-" + Button.ew++;
            this.e5 = "ig-button-" + Button.ev++;
            this.e2 = "inline-block";
            this.e3 = "inline-block";
            this.e4 = null;
            this.e0 = "center";
            this.ex = null;
            this.e7 = null;
            this.cc = 0;
            this.z = false;
            this.ab = false;
            this.ac = true;
            this.e1 = null;
            this.x = false;
            this.w = false;
            this.ad = false;
            this.aa = false;
            this.y = false;
            this.propertyChanged = null;
            this.clicked = null;
            let a = new ButtonView();
            a.a = this;
            this.view = a;
            this.view.ai();
        }
        get view() {
            return this._view;
        }
        set view(a) {
            this._view = a;
        }
        destroy() {
            this.provideContainer(null);
        }
        jx(a, b, c) {
            if (this.propertyChanged != null) {
                this.propertyChanged(this, new PropertyChangedEventArgs(a));
            }
            this.jz(a, b, c);
        }
        jz(a, b, c) {
            switch (a) {
                case "CornerRadius":
                case "BorderColor":
                case "RestingElevation":
                case "HoverElevation":
                case "FocusElevation":
                case "RippleColor":
                case "BorderWidth":
                case "BackgroundColor":
                case "DisabledBackgroundColor":
                case "DisabledBorderColor":
                case "DisabledTextColor":
                case "DisabledElevation":
                case "HoverBackgroundColor":
                case "TextStyle":
                case "FocusTextColor":
                case "TextColor":
                case "HoverTextColor":
                case "FocusBackgroundColor":
                case "RaisedBorderColor":
                case "RaisedRestingElevation":
                case "RaisedFocusElevation":
                case "RaisedHoverElevation":
                case "RaisedRippleColor":
                case "RaisedBorderWidth":
                case "RaisedBackgroundColor":
                case "RaisedDisabledBackgroundColor":
                case "RaisedDisabledBorderColor":
                case "RaisedDisabledTextColor":
                case "RaisedDisabledElevation":
                case "RaisedHoverBackgroundColor":
                case "RaisedCornerRadius":
                case "RaisedFocusBackgroundColor":
                case "RaisedFocusTextColor":
                case "RaisedTextColor":
                case "RaisedHoverTextColor":
                case "FlatBorderColor":
                case "FlatRestingElevation":
                case "FlatFocusElevation":
                case "FlatHoverElevation":
                case "FlatRippleColor":
                case "FlatBorderWidth":
                case "FlatBackgroundColor":
                case "FlatDisabledBackgroundColor":
                case "FlatDisabledBorderColor":
                case "FlatDisabledTextColor":
                case "FlatDisabledElevation":
                case "FlatHoverBackgroundColor":
                case "FlatCornerRadius":
                case "FlatFocusBackgroundColor":
                case "FlatFocusTextColor":
                case "FlatTextColor":
                case "FlatHoverTextColor":
                case "OutlinedRippleColor":
                case "OutlinedRestingElevation":
                case "OutlinedFocusElevation":
                case "OutlinedHoverElevation":
                case "OutlinedBorderColor":
                case "OutlinedBorderWidth":
                case "OutlinedBackgroundColor":
                case "OutlinedDisabledBackgroundColor":
                case "OutlinedDisabledBorderColor":
                case "OutlinedDisabledTextColor":
                case "OutlinedDisabledElevation":
                case "OutlinedHoverBackgroundColor":
                case "OutlinedCornerRadius":
                case "OutlinedFocusBackgroundColor":
                case "OutlinedFocusTextColor":
                case "OutlinedTextColor":
                case "OutlinedHoverTextColor":
                case "FabBorderColor":
                case "FabRippleColor":
                case "FabRestingElevation":
                case "FabFocusElevation":
                case "FabHoverElevation":
                case "FabBorderWidth":
                case "FabBackgroundColor":
                case "FabDisabledBackgroundColor":
                case "FabDisabledBorderColor":
                case "FabDisabledTextColor":
                case "FabDisabledElevation":
                case "FabHoverBackgroundColor":
                case "FabCornerRadius":
                case "FabFocusBackgroundColor":
                case "FabFocusTextColor":
                case "FabTextColor":
                case "FabHoverTextColor":
                case "IconRestingElevation":
                case "IconFocusElevation":
                case "IconHoverElevation":
                case "IconBorderColor":
                case "IconRippleColor":
                case "IconBorderWidth":
                case "IconBackgroundColor":
                case "IconDisabledBackgroundColor":
                case "IconDisabledBorderColor":
                case "IconDisabledTextColor":
                case "IconDisabledElevation":
                case "IconHoverBackgroundColor":
                case "IconCornerRadius":
                case "IconFocusBackgroundColor":
                case "IconFocusTextColor":
                case "IconTextColor":
                case "IconHoverTextColor":
                case "ElevationMode":
                    this.j2();
                    break;
                case "BaseTheme":
                    this.j3(this.d == 0 ? 2 : this.d);
                    this.j2();
                    break;
                case "IsDisabled":
                    this.ae = this.disabled || this.af;
                    this.j2();
                    this.view.ak();
                    break;
                case "DisableRipple":
                    this.ae = this.disabled || this.af;
                    this.j2();
                    break;
                case "ActualDisableRipple":
                    this.view.ae();
                    break;
                case "ActualBackgroundColor":
                case "ActualDisabledBackgroundColor":
                case "ActualDisabledBorderColor":
                case "ActualDisabledTextColor":
                case "ActualDisabledElevation":
                case "ActualHoverBackgroundColor":
                case "ActualBorderColor":
                case "ActualFocusTextColor":
                case "ActualHoverTextColor":
                case "ActualElevationMode":
                    this.j1();
                    break;
                case "Density":
                    this.g = this.h == 0 ? 1 : this.h;
                    break;
                case "ActualDensity":
                case "ActualBorderWidth":
                case "ActualCornerRadius":
                    this.view.ar();
                    this.j1();
                    break;
                case "DisplayType":
                    this.view.ar();
                    this.j2();
                    this.j1();
                    break;
                case "InputId":
                    this.view.aj();
                    break;
                case "Display":
                    this.view.af();
                    break;
                case "FlexDirection":
                    this.view.ag();
                    break;
                case "FlexGrow":
                    this.view.ah();
                    break;
                case "AlignItems":
                    this.view.aa();
                    break;
                case "MinWidth":
                    this.view.an(this.bh);
                    this.view.ar();
                    break;
                case "MinHeight":
                    this.view.am(this.bg);
                    this.view.ar();
                    break;
                case "TabIndex":
                    this.view.ao();
                    break;
            }
        }
        kc(a, b) {
            return new CornerRadius(1, !isNaN_(a.c) ? a.c : b.c, !isNaN_(a.d) ? a.d : b.d, !isNaN_(a.b) ? a.b : b.b, !isNaN_(a.a) ? a.a : b.a);
        }
        j2() {
            this.k = this.o != 0 ? this.o : this.m;
            this.nj = this.sf;
            this.ng = this.r3;
            this.m6 = this.nm;
            switch (this.s) {
                case 0:
                    this.m9 = this.sg != null ? this.sg : this.tx != null ? this.tx : this.r6;
                    this.na = this.sh != null ? this.sh : this.ty != null ? this.ty : this.r7;
                    this.nb = this.si != null ? this.si : this.tz != null ? this.tz : this.r8;
                    this.cd = this.d6 >= 0 ? this.d6 : this.ep >= 0 ? this.ep : this.d2;
                    this.cg = this.et >= 0 ? this.et : this.es >= 0 ? this.es : this.d5;
                    this.cf = this.eg >= 0 ? this.eg : this.er >= 0 ? this.er : this.d4;
                    this.ce = this.ef >= 0 ? this.ef : this.eq >= 0 ? this.eq : this.d3;
                    this.m8 = this.nl != null ? this.nl : this.tw != null ? this.tw : this.r5;
                    this.m7 = this.nk != null ? this.nk : this.tv != null ? this.tv : this.r4;
                    this.ne = this.s7 != null ? this.s7 : this.t2 != null ? this.t2 : this.sb;
                    this.kb = this.kc(this.kd, this.kc(this.kx, this.ks));
                    this.aw = !isNaN_(this.ax) ? this.ax : !isNaN_(this.bj) ? this.bj : this.bc;
                    this.nh = this.t6 != null ? this.t6 : this.t4 != null ? this.t4 : this.sd;
                    this.nc = this.s5 != null ? this.s5 : this.t0 != null ? this.t0 : this.r9;
                    this.nd = this.s6 != null ? this.s6 : this.t1 != null ? this.t1 : this.sa;
                    this.ni = this.t7 != null ? this.t7 : this.t5 != null ? this.t5 : this.se;
                    this.nf = this.s8 != null ? this.s8 : this.t3 != null ? this.t3 : this.sc;
                    break;
                case 2:
                    this.m9 = this.sg != null ? this.sg : this.tm != null ? this.tm : this.ru;
                    this.na = this.sh != null ? this.sh : this.tn != null ? this.tn : this.rv;
                    this.nb = this.si != null ? this.si : this.to != null ? this.to : this.rw;
                    this.cd = this.d6 >= 0 ? this.d6 : this.el >= 0 ? this.el : this.dy;
                    this.cg = this.et >= 0 ? this.et : this.eo >= 0 ? this.eo : this.d1;
                    this.cf = this.eg >= 0 ? this.eg : this.en >= 0 ? this.en : this.d0;
                    this.ce = this.ef >= 0 ? this.ef : this.em >= 0 ? this.em : this.dz;
                    this.m8 = this.nl != null ? this.nl : this.tl != null ? this.tl : this.rt;
                    this.m7 = this.s7 != null ? this.s7 : this.nk != null ? this.nk : this.tk != null ? this.tk : this.rs;
                    this.ne = this.tr != null ? this.tr : this.rz;
                    this.kb = this.kc(this.kd, this.kc(this.kw, this.kr));
                    this.aw = !isNaN_(this.ax) ? this.ax : !isNaN_(this.bi) ? this.bi : this.bb;
                    this.nh = this.t6 != null ? this.t6 : this.tt != null ? this.tt : this.r1;
                    this.nc = this.s5 != null ? this.s5 : this.tp != null ? this.tp : this.rx;
                    this.nd = this.s6 != null ? this.s6 : this.tq != null ? this.tq : this.ry;
                    this.ni = this.t7 != null ? this.t7 : this.tu != null ? this.tu : this.r2;
                    this.nf = this.s8 != null ? this.s8 : this.ts != null ? this.ts : this.r0;
                    break;
                case 1:
                    this.m9 = this.sg != null ? this.sg : this.sw != null ? this.sw : this.pm;
                    this.na = this.sh != null ? this.sh : this.sx != null ? this.sx : this.pn;
                    this.nb = this.si != null ? this.si : this.sy != null ? this.sy : this.po;
                    this.cd = this.d6 >= 0 ? this.d6 : this.eb >= 0 ? this.eb : this.c5;
                    this.cg = this.et >= 0 ? this.et : this.ee >= 0 ? this.ee : this.c8;
                    this.cf = this.eg >= 0 ? this.eg : this.ed >= 0 ? this.ed : this.c7;
                    this.ce = this.ef >= 0 ? this.ef : this.ec >= 0 ? this.ec : this.c6;
                    this.m8 = this.nl != null ? this.nl : this.sv != null ? this.sv : this.pl;
                    this.m7 = this.nk != null ? this.nk : this.su != null ? this.su : this.pk;
                    this.ne = this.s7 != null ? this.s7 : this.s1 != null ? this.s1 : this.pr;
                    this.kb = this.kc(this.kd, this.kc(this.ku, this.kk));
                    this.aw = !isNaN_(this.ax) ? this.ax : !isNaN_(this.be) ? this.be : this.a4;
                    this.nh = this.t6 != null ? this.t6 : this.s3 != null ? this.s3 : this.pt;
                    this.nc = this.s5 != null ? this.s5 : this.sz != null ? this.sz : this.pp;
                    this.nd = this.s6 != null ? this.s6 : this.s0 != null ? this.s0 : this.pq;
                    this.ni = this.t7 != null ? this.t7 : this.s4 != null ? this.s4 : this.pu;
                    this.nf = this.s8 != null ? this.s8 : this.s2 != null ? this.s2 : this.ps;
                    break;
                case 4:
                    this.m9 = this.sg != null ? this.sg : this.tb != null ? this.tb : this.px;
                    this.na = this.sh != null ? this.sh : this.tc != null ? this.tc : this.py;
                    this.nb = this.si != null ? this.si : this.td != null ? this.td : this.pz;
                    this.cd = this.d6 >= 0 ? this.d6 : this.eh >= 0 ? this.eh : this.c9;
                    this.cg = this.et >= 0 ? this.et : this.ek >= 0 ? this.ek : this.dc;
                    this.cf = this.eg >= 0 ? this.eg : this.ej >= 0 ? this.ej : this.db;
                    this.ce = this.ef >= 0 ? this.ef : this.ei >= 0 ? this.ei : this.da;
                    this.m8 = this.nl != null ? this.nl : this.ta != null ? this.ta : this.pw;
                    this.m7 = this.nk != null ? this.nk : this.s9 != null ? this.s9 : this.pv;
                    this.ne = this.s7 != null ? this.s7 : this.tg != null ? this.tg : this.p2;
                    this.kb = this.kc(this.kd, this.kc(this.kv, this.kl));
                    this.aw = !isNaN_(this.ax) ? this.ax : !isNaN_(this.bf) ? this.bf : this.a5;
                    this.nh = this.t6 != null ? this.t6 : this.ti != null ? this.ti : this.p4;
                    this.nc = this.s5 != null ? this.s5 : this.te != null ? this.te : this.p0;
                    this.nd = this.s6 != null ? this.s6 : this.tf != null ? this.tf : this.p1;
                    this.ni = this.t7 != null ? this.t7 : this.tj != null ? this.tj : this.p5;
                    this.nf = this.s8 != null ? this.s8 : this.th != null ? this.th : this.p3;
                    break;
                case 3:
                    this.m9 = this.sg != null ? this.sg : this.sl != null ? this.sl : this.pb;
                    this.na = this.sh != null ? this.sh : this.sm != null ? this.sm : this.pc;
                    this.nb = this.si != null ? this.si : this.sn != null ? this.sn : this.pd;
                    this.cd = this.d6 >= 0 ? this.d6 : this.d7 >= 0 ? this.d7 : this.c1;
                    this.cg = this.et >= 0 ? this.et : this.ea >= 0 ? this.ea : this.c4;
                    this.cf = this.eg >= 0 ? this.eg : this.d9 >= 0 ? this.d9 : this.c3;
                    this.ce = this.ef >= 0 ? this.ef : this.d8 >= 0 ? this.d8 : this.c2;
                    this.m8 = this.nl != null ? this.nl : this.sk != null ? this.sk : this.pa;
                    this.m7 = this.nk != null ? this.nk : this.sj != null ? this.sj : this.o9;
                    this.ne = this.s7 != null ? this.s7 : this.sq != null ? this.sq : this.pg;
                    this.kb = this.kc(this.kd, this.kc(this.kt, this.kj));
                    this.aw = !isNaN_(this.ax) ? this.ax : !isNaN_(this.bd) ? this.bd : this.a3;
                    this.nh = this.t6 != null ? this.t6 : this.ss != null ? this.ss : this.sd;
                    this.nc = this.s5 != null ? this.s5 : this.so != null ? this.so : this.pe;
                    this.nd = this.s6 != null ? this.s6 : this.sp != null ? this.sp : this.pf;
                    this.ni = this.t7 != null ? this.t7 : this.st != null ? this.st : this.pj;
                    this.nf = this.s8 != null ? this.s8 : this.sr != null ? this.sr : this.ph;
                    break;
            }
        }
        js() {
            if (this.disabled) {
                return;
            }
            if (this.clicked != null) {
                this.clicked(this, new ButtonClickEventArgs());
            }
        }
        jr() {
            this.focused = false;
            this.j1();
        }
        ju() {
            this.focused = true;
            this.j1();
        }
        jw() {
            this.al = false;
            this.j1();
        }
        jv() {
            this.al = true;
            this.j1();
        }
        j1() {
            let a = this.view.p;
            NativeUI.l(a, this.kb);
            NativeUI.k(a, this.m8);
            if (!isNaN_(this.aw)) {
                NativeUI.m(a, new Thickness(0, this.aw));
            }
            if (this.disabled) {
                this.view.aq(this.cd, this.k);
            }
            else if (this.focused && this.ak) {
                this.view.aq(this.ce, this.k);
            }
            else {
                if (this.al) {
                    this.view.aq(this.cf, this.k);
                }
                else {
                    this.view.aq(this.cg, this.k);
                }
            }
            if (this.disabled) {
                NativeUI.j(a, this.m9);
            }
            else if (this.focused && this.ak) {
                NativeUI.j(a, this.nc);
            }
            else {
                if (this.al) {
                    NativeUI.j(a, this.ne);
                }
                else {
                    NativeUI.j(a, this.m7);
                }
            }
            this.view.ap(this.nh, BrushUtil.g(0, 0, 0, 0));
            if (this.v != null) {
                this.view.as(this.v);
            }
            if (this.disabled) {
                NativeUI.n(this.view.q, this.nb);
                NativeUI.o(this.view.q, this.nb);
            }
            else if (this.focused && this.ak) {
                NativeUI.n(this.view.q, this.nd);
                NativeUI.o(this.view.q, this.nd);
            }
            else {
                if (this.al) {
                    NativeUI.n(this.view.q, this.nf);
                    NativeUI.o(this.view.q, this.nf);
                }
                else {
                    NativeUI.n(this.view.q, this.ni);
                    NativeUI.o(this.view.q, this.ni);
                }
            }
        }
        j3(a) {
            this.sf = ThemeResolver.e(a, Button.rr, Button.o8);
            this.r3 = ThemeResolver.e(a, Button.rf, Button.ow);
            this.nm = ThemeResolver.e(a, Button.p6, Button.nn);
            this.m = ThemeResolver.a(a, Button.n, Button.l);
            this.r6 = ThemeResolver.e(a, Button.ri, Button.oz);
            this.r7 = ThemeResolver.e(a, Button.rj, Button.o0);
            this.r8 = ThemeResolver.e(a, Button.rk, Button.o1);
            this.d2 = ThemeResolver.c(a, Button.du, Button.cx);
            this.d5 = ThemeResolver.c(a, Button.dx, Button.c0);
            this.d4 = ThemeResolver.c(a, Button.dw, Button.cz);
            this.d3 = ThemeResolver.c(a, Button.dv, Button.cy);
            this.r5 = ThemeResolver.e(a, Button.rh, Button.oy);
            this.sd = ThemeResolver.e(a, Button.rp, Button.o6);
            this.ks = ThemeResolver.d(a, Button.kq, Button.ki);
            this.bc = ThemeResolver.b(a, Button.ba, Button.a2);
            this.r4 = ThemeResolver.e(a, Button.rg, Button.ox);
            this.se = ThemeResolver.e(a, Button.rq, Button.o7);
            this.sc = ThemeResolver.e(a, Button.ro, Button.o5);
            this.sb = ThemeResolver.e(a, Button.rn, Button.o4);
            this.r9 = ThemeResolver.e(a, Button.rl, Button.o2);
            this.sa = ThemeResolver.e(a, Button.rm, Button.o3);
            this.pm = ThemeResolver.e(a, Button.qk, Button.n1);
            this.pn = ThemeResolver.e(a, Button.ql, Button.n2);
            this.po = ThemeResolver.e(a, Button.qm, Button.n3);
            this.c5 = ThemeResolver.c(a, Button.dh, Button.cl);
            this.c8 = ThemeResolver.c(a, Button.dk, Button.co);
            this.c7 = ThemeResolver.c(a, Button.dj, Button.cn);
            this.c6 = ThemeResolver.c(a, Button.di, Button.cm);
            this.pl = ThemeResolver.e(a, Button.qj, Button.n0);
            this.pt = ThemeResolver.e(a, Button.qr, Button.n8);
            this.kk = ThemeResolver.d(a, Button.kn, Button.kf);
            this.a4 = ThemeResolver.b(a, Button.a7, Button.az);
            this.pk = ThemeResolver.e(a, Button.qi, Button.nz);
            this.pu = ThemeResolver.e(a, Button.qs, Button.n9);
            this.ps = ThemeResolver.e(a, Button.qq, Button.n7);
            this.pr = ThemeResolver.e(a, Button.qp, Button.n6);
            this.pp = ThemeResolver.e(a, Button.qn, Button.n4);
            this.pq = ThemeResolver.e(a, Button.qo, Button.n5);
            this.ru = ThemeResolver.e(a, Button.q6, Button.on);
            this.rv = ThemeResolver.e(a, Button.q7, Button.oo);
            this.rw = ThemeResolver.e(a, Button.q8, Button.op);
            this.dy = ThemeResolver.c(a, Button.dq, Button.ct);
            this.d1 = ThemeResolver.c(a, Button.dt, Button.cw);
            this.d0 = ThemeResolver.c(a, Button.ds, Button.cv);
            this.dz = ThemeResolver.c(a, Button.dr, Button.cu);
            this.rt = ThemeResolver.e(a, Button.q5, Button.om);
            this.r1 = ThemeResolver.e(a, Button.rd, Button.ou);
            this.kr = ThemeResolver.d(a, Button.kp, Button.kh);
            this.bb = ThemeResolver.b(a, Button.a9, Button.a1);
            this.rs = ThemeResolver.e(a, Button.q4, Button.ol);
            this.r2 = ThemeResolver.e(a, Button.re, Button.ov);
            this.r0 = ThemeResolver.e(a, Button.rc, Button.ot);
            this.rz = ThemeResolver.e(a, Button.rb, Button.os);
            this.rx = ThemeResolver.e(a, Button.q9, Button.oq);
            this.ry = ThemeResolver.e(a, Button.ra, Button.or);
            this.px = ThemeResolver.e(a, Button.qv, Button.oc);
            this.py = ThemeResolver.e(a, Button.qw, Button.od);
            this.pz = ThemeResolver.e(a, Button.qx, Button.oe);
            this.c9 = ThemeResolver.c(a, Button.dl, Button.cp);
            this.dc = ThemeResolver.c(a, Button.dp, Button.cs);
            this.db = ThemeResolver.c(a, Button.dn, Button.cr);
            this.da = ThemeResolver.c(a, Button.dm, Button.cq);
            this.pw = ThemeResolver.e(a, Button.qu, Button.ob);
            this.p4 = ThemeResolver.e(a, Button.q2, Button.oj);
            this.kl = ThemeResolver.d(a, Button.ko, Button.kg);
            this.a5 = ThemeResolver.b(a, Button.a8, Button.a0);
            this.pv = ThemeResolver.e(a, Button.qt, Button.oa);
            this.p5 = ThemeResolver.e(a, Button.q3, Button.ok);
            this.p3 = ThemeResolver.e(a, Button.q1, Button.oi);
            this.p2 = ThemeResolver.e(a, Button.q0, Button.oh);
            this.p0 = ThemeResolver.e(a, Button.qy, Button.of);
            this.p1 = ThemeResolver.e(a, Button.qz, Button.og);
            this.pb = ThemeResolver.e(a, Button.p9, Button.nq);
            this.pc = ThemeResolver.e(a, Button.qa, Button.nr);
            this.pd = ThemeResolver.e(a, Button.qb, Button.ns);
            this.c1 = ThemeResolver.c(a, Button.dd, Button.ch);
            this.c4 = ThemeResolver.c(a, Button.dg, Button.ck);
            this.c3 = ThemeResolver.c(a, Button.df, Button.cj);
            this.c2 = ThemeResolver.c(a, Button.de, Button.ci);
            this.pa = ThemeResolver.e(a, Button.p8, Button.np);
            this.pi = ThemeResolver.e(a, Button.qg, Button.nx);
            this.kj = ThemeResolver.d(a, Button.km, Button.ke);
            this.a3 = ThemeResolver.b(a, Button.a6, Button.ay);
            this.o9 = ThemeResolver.e(a, Button.p7, Button.no);
            this.pj = ThemeResolver.e(a, Button.qh, Button.ny);
            this.ph = ThemeResolver.e(a, Button.qf, Button.nw);
            this.pg = ThemeResolver.e(a, Button.qe, Button.nv);
            this.pe = ThemeResolver.e(a, Button.qc, Button.nt);
            this.pf = ThemeResolver.e(a, Button.qd, Button.nu);
        }
        get bh() {
            return this.at;
        }
        set bh(a) {
            let b = this.at;
            this.at = a;
            if (b != this.at) {
                this.jx("MinWidth", b, a);
            }
        }
        get bg() {
            return this.as;
        }
        set bg(a) {
            let b = this.as;
            this.as = a;
            if (b != this.as) {
                this.jx("MinHeight", b, a);
            }
        }
        get s() {
            return this.r;
        }
        set s(a) {
            let b = this.r;
            this.r = a;
            if (b != this.r) {
                this.jx("DisplayType", enumGetBox(ButtonDisplayStyle_$type, b), enumGetBox(ButtonDisplayStyle_$type, a));
            }
        }
        get h() {
            return this.f;
        }
        set h(a) {
            let b = this.f;
            this.f = a;
            if (b != this.f) {
                this.jx("Density", enumGetBox(ControlDisplayDensity_$type, b), enumGetBox(ControlDisplayDensity_$type, a));
            }
        }
        get d() {
            return this.b;
        }
        set d(a) {
            let b = this.b;
            this.b = a;
            if (b != this.b) {
                this.jx("BaseTheme", enumGetBox(BaseControlTheme_$type, b), enumGetBox(BaseControlTheme_$type, a));
            }
        }
        get g() {
            return this.e;
        }
        set g(a) {
            let b = this.e;
            this.e = a;
            if (b != this.e) {
                this.jx("ActualDensity", enumGetBox(ControlDisplayDensity_$type, b), enumGetBox(ControlDisplayDensity_$type, a));
            }
        }
        get m8() {
            return this.k0;
        }
        set m8(a) {
            let b = this.k0;
            this.k0 = a;
            if (b != this.k0) {
                this.jx("ActualBorderColor", b, a);
            }
        }
        get nj() {
            return this.lb;
        }
        set nj(a) {
            let b = this.lb;
            this.lb = a;
            if (b != this.lb) {
                this.jx("ActualUmbraShadowColor", b, a);
            }
        }
        get ng() {
            return this.k8;
        }
        set ng(a) {
            let b = this.k8;
            this.k8 = a;
            if (b != this.k8) {
                this.jx("ActualPenumbraShadowColor", b, a);
            }
        }
        get m6() {
            return this.ky;
        }
        set m6(a) {
            let b = this.ky;
            this.ky = a;
            if (b != this.ky) {
                this.jx("ActualAmbientShadowColor", b, a);
            }
        }
        get cg() {
            return this.bn;
        }
        set cg(a) {
            let b = this.bn;
            this.bn = a;
            if (b != this.bn) {
                this.jx("ActualRestingElevation", b, a);
            }
        }
        get ce() {
            return this.bl;
        }
        set ce(a) {
            let b = this.bl;
            this.bl = a;
            if (b != this.bl) {
                this.jx("ActualFocusElevation", b, a);
            }
        }
        get cf() {
            return this.bm;
        }
        set cf(a) {
            let b = this.bm;
            this.bm = a;
            if (b != this.bm) {
                this.jx("ActualHoverElevation", b, a);
            }
        }
        get nh() {
            return this.k9;
        }
        set nh(a) {
            let b = this.k9;
            this.k9 = a;
            if (b != this.k9) {
                this.jx("ActualRippleColor", b, a);
            }
        }
        get kb() {
            return this.j4;
        }
        set kb(a) {
            let b = this.j4;
            this.j4 = a;
            if (b != this.j4) {
                this.jx("ActualCornerRadius", b, a);
            }
        }
        get m7() {
            return this.kz;
        }
        set m7(a) {
            let b = this.kz;
            this.kz = a;
            if (b != this.kz) {
                this.jx("ActualBackgroundColor", b, a);
            }
        }
        get m9() {
            return this.k1;
        }
        set m9(a) {
            let b = this.k1;
            this.k1 = a;
            if (b != this.k1) {
                this.jx("ActualDisabledBackgroundColor", b, a);
            }
        }
        get na() {
            return this.k2;
        }
        set na(a) {
            let b = this.k2;
            this.k2 = a;
            if (b != this.k2) {
                this.jx("ActualDisabledBorderColor", b, a);
            }
        }
        get nb() {
            return this.k3;
        }
        set nb(a) {
            let b = this.k3;
            this.k3 = a;
            if (b != this.k3) {
                this.jx("ActualDisabledTextColor", b, a);
            }
        }
        get cd() {
            return this.bk;
        }
        set cd(a) {
            let b = this.bk;
            this.bk = a;
            if (b != this.bk) {
                this.jx("ActualDisabledElevation", b, a);
            }
        }
        get k() {
            return this.i;
        }
        set k(a) {
            let b = this.i;
            this.i = a;
            if (b != this.i) {
                this.jx("ActualElevationMode", enumGetBox(ElevationMode_$type, b), enumGetBox(ElevationMode_$type, a));
            }
        }
        get ni() {
            return this.la;
        }
        set ni(a) {
            let b = this.la;
            this.la = a;
            if (b != this.la) {
                this.jx("ActualTextColor", b, a);
            }
        }
        get nf() {
            return this.k7;
        }
        set nf(a) {
            let b = this.k7;
            this.k7 = a;
            if (b != this.k7) {
                this.jx("ActualHoverTextColor", b, a);
            }
        }
        get nc() {
            return this.k4;
        }
        set nc(a) {
            let b = this.k4;
            this.k4 = a;
            if (b != this.k4) {
                this.jx("ActualFocusBackgroundColor", b, a);
            }
        }
        get nd() {
            return this.k5;
        }
        set nd(a) {
            let b = this.k5;
            this.k5 = a;
            if (b != this.k5) {
                this.jx("ActualFocusTextColor", b, a);
            }
        }
        get ne() {
            return this.k6;
        }
        set ne(a) {
            let b = this.k6;
            this.k6 = a;
            if (b != this.k6) {
                this.jx("ActualHoverBackgroundColor", b, a);
            }
        }
        get aw() {
            return this.an;
        }
        set aw(a) {
            let b = this.an;
            this.an = a;
            if (b != this.an) {
                this.jx("ActualBorderWidth", b, a);
            }
        }
        get tv() {
            return this.mt;
        }
        set tv(a) {
            let b = this.mt;
            this.mt = a;
            if (b != this.mt) {
                this.jx("RaisedBackgroundColor", b, a);
            }
        }
        get tx() {
            return this.mv;
        }
        set tx(a) {
            let b = this.mv;
            this.mv = a;
            if (b != this.mv) {
                this.jx("RaisedDisabledBackgroundColor", b, a);
            }
        }
        get ty() {
            return this.mw;
        }
        set ty(a) {
            let b = this.mw;
            this.mw = a;
            if (b != this.mw) {
                this.jx("RaisedDisabledBorderColor", b, a);
            }
        }
        get tz() {
            return this.mx;
        }
        set tz(a) {
            let b = this.mx;
            this.mx = a;
            if (b != this.mx) {
                this.jx("RaisedDisabledTextColor", b, a);
            }
        }
        get ep() {
            return this.b7;
        }
        set ep(a) {
            let b = this.b7;
            this.b7 = a;
            if (b != this.b7) {
                this.jx("RaisedDisabledElevation", b, a);
            }
        }
        get t2() {
            return this.m0;
        }
        set t2(a) {
            let b = this.m0;
            this.m0 = a;
            if (b != this.m0) {
                this.jx("RaisedHoverBackgroundColor", b, a);
            }
        }
        get tw() {
            return this.mu;
        }
        set tw(a) {
            let b = this.mu;
            this.mu = a;
            if (b != this.mu) {
                this.jx("RaisedBorderColor", b, a);
            }
        }
        get es() {
            return this.ca;
        }
        set es(a) {
            let b = this.ca;
            this.ca = a;
            if (b != this.ca) {
                this.jx("RaisedRestingElevation", b, a);
            }
        }
        get er() {
            return this.b9;
        }
        set er(a) {
            let b = this.b9;
            this.b9 = a;
            if (b != this.b9) {
                this.jx("RaisedHoverElevation", b, a);
            }
        }
        get eq() {
            return this.b8;
        }
        set eq(a) {
            let b = this.b8;
            this.b8 = a;
            if (b != this.b8) {
                this.jx("RaisedFocusElevation", b, a);
            }
        }
        get bj() {
            return this.av;
        }
        set bj(a) {
            let b = this.av;
            this.av = a;
            if (b != this.av) {
                this.jx("RaisedBorderWidth", b, a);
            }
        }
        get kx() {
            return this.ka;
        }
        set kx(a) {
            let b = this.ka;
            this.ka = a;
            {
                this.jx("RaisedCornerRadius", b, a);
            }
        }
        get t4() {
            return this.m2;
        }
        set t4(a) {
            let b = this.m2;
            this.m2 = a;
            if (b != this.m2) {
                this.jx("RaisedRippleColor", b, a);
            }
        }
        get su() {
            return this.lr;
        }
        set su(a) {
            let b = this.lr;
            this.lr = a;
            if (b != this.lr) {
                this.jx("FlatBackgroundColor", b, a);
            }
        }
        get sw() {
            return this.lt;
        }
        set sw(a) {
            let b = this.lt;
            this.lt = a;
            if (b != this.lt) {
                this.jx("FlatDisabledBackgroundColor", b, a);
            }
        }
        get sx() {
            return this.lu;
        }
        set sx(a) {
            let b = this.lu;
            this.lu = a;
            if (b != this.lu) {
                this.jx("FlatDisabledBorderColor", b, a);
            }
        }
        get sy() {
            return this.lv;
        }
        set sy(a) {
            let b = this.lv;
            this.lv = a;
            if (b != this.lv) {
                this.jx("FlatDisabledTextColor", b, a);
            }
        }
        get eb() {
            return this.bt;
        }
        set eb(a) {
            let b = this.bt;
            this.bt = a;
            if (b != this.bt) {
                this.jx("FlatDisabledElevation", b, a);
            }
        }
        get s1() {
            return this.ly;
        }
        set s1(a) {
            let b = this.ly;
            this.ly = a;
            if (b != this.ly) {
                this.jx("FlatHoverBackgroundColor", b, a);
            }
        }
        get sv() {
            return this.ls;
        }
        set sv(a) {
            let b = this.ls;
            this.ls = a;
            if (b != this.ls) {
                this.jx("FlatBorderColor", b, a);
            }
        }
        get ee() {
            return this.bw;
        }
        set ee(a) {
            let b = this.bw;
            this.bw = a;
            if (b != this.bw) {
                this.jx("FlatRestingElevation", b, a);
            }
        }
        get ed() {
            return this.bv;
        }
        set ed(a) {
            let b = this.bv;
            this.bv = a;
            if (b != this.bv) {
                this.jx("FlatHoverElevation", b, a);
            }
        }
        get ec() {
            return this.bu;
        }
        set ec(a) {
            let b = this.bu;
            this.bu = a;
            if (b != this.bu) {
                this.jx("FlatFocusElevation", b, a);
            }
        }
        get s3() {
            return this.l0;
        }
        set s3(a) {
            let b = this.l0;
            this.l0 = a;
            if (b != this.l0) {
                this.jx("FlatRippleColor", b, a);
            }
        }
        get s4() {
            return this.l1;
        }
        set s4(a) {
            let b = this.l1;
            this.l1 = a;
            if (b != this.l1) {
                this.jx("FlatTextColor", b, a);
            }
        }
        get s2() {
            return this.lz;
        }
        set s2(a) {
            let b = this.lz;
            this.lz = a;
            if (b != this.lz) {
                this.jx("FlatHoverTextColor", b, a);
            }
        }
        get sz() {
            return this.lw;
        }
        set sz(a) {
            let b = this.lw;
            this.lw = a;
            if (b != this.lw) {
                this.jx("FlatFocusBackgroundColor", b, a);
            }
        }
        get s0() {
            return this.lx;
        }
        set s0(a) {
            let b = this.lx;
            this.lx = a;
            if (b != this.lx) {
                this.jx("FlatFocusTextColor", b, a);
            }
        }
        get be() {
            return this.aq;
        }
        set be(a) {
            let b = this.aq;
            this.aq = a;
            if (b != this.aq) {
                this.jx("FlatBorderWidth", b, a);
            }
        }
        get ku() {
            return this.j7;
        }
        set ku(a) {
            let b = this.j7;
            this.j7 = a;
            {
                this.jx("FlatCornerRadius", b, a);
            }
        }
        get nk() {
            return this.lc;
        }
        set nk(a) {
            let b = this.lc;
            this.lc = a;
            if (b != this.lc) {
                this.jx("BackgroundColor", b, a);
            }
        }
        get sg() {
            return this.le;
        }
        set sg(a) {
            let b = this.le;
            this.le = a;
            if (b != this.le) {
                this.jx("DisabledBackgroundColor", b, a);
            }
        }
        get sh() {
            return this.lf;
        }
        set sh(a) {
            let b = this.lf;
            this.lf = a;
            if (b != this.lf) {
                this.jx("DisabledBorderColor", b, a);
            }
        }
        get si() {
            return this.lg;
        }
        set si(a) {
            let b = this.lg;
            this.lg = a;
            if (b != this.lg) {
                this.jx("DisabledTextColor", b, a);
            }
        }
        get d6() {
            return this.bo;
        }
        set d6(a) {
            let b = this.bo;
            this.bo = a;
            if (b != this.bo) {
                this.jx("DisabledElevation", b, a);
            }
        }
        get o() {
            return this.j;
        }
        set o(a) {
            let b = this.j;
            this.j = a;
            if (b != this.j) {
                this.jx("ElevationMode", enumGetBox(ElevationMode_$type, b), enumGetBox(ElevationMode_$type, a));
            }
        }
        get s7() {
            return this.l4;
        }
        set s7(a) {
            let b = this.l4;
            this.l4 = a;
            if (b != this.l4) {
                this.jx("HoverBackgroundColor", b, a);
            }
        }
        get nl() {
            return this.ld;
        }
        set nl(a) {
            let b = this.ld;
            this.ld = a;
            if (b != this.ld) {
                this.jx("BorderColor", b, a);
            }
        }
        get et() {
            return this.cb;
        }
        set et(a) {
            let b = this.cb;
            this.cb = a;
            if (b != this.cb) {
                this.jx("RestingElevation", b, a);
            }
        }
        get eg() {
            return this.by;
        }
        set eg(a) {
            let b = this.by;
            this.by = a;
            if (b != this.by) {
                this.jx("HoverElevation", b, a);
            }
        }
        get ef() {
            return this.bx;
        }
        set ef(a) {
            let b = this.bx;
            this.bx = a;
            if (b != this.bx) {
                this.jx("FocusElevation", b, a);
            }
        }
        get v() {
            return this.u;
        }
        set v(a) {
            let b = this.u;
            this.u = a;
            if (b != this.u) {
                this.jx("TextStyle", b, a);
            }
        }
        get t7() {
            return this.m5;
        }
        set t7(a) {
            let b = this.m5;
            this.m5 = a;
            if (b != this.m5) {
                this.jx("TextColor", b, a);
            }
        }
        get s8() {
            return this.l6;
        }
        set s8(a) {
            let b = this.l6;
            this.l6 = a;
            if (b != this.l6) {
                this.jx("HoverTextColor", b, a);
            }
        }
        get t6() {
            return this.m4;
        }
        set t6(a) {
            let b = this.m4;
            this.m4 = a;
            if (b != this.m4) {
                this.jx("RippleColor", b, a);
            }
        }
        get t5() {
            return this.m3;
        }
        set t5(a) {
            let b = this.m3;
            this.m3 = a;
            if (b != this.m3) {
                this.jx("RaisedTextColor", b, a);
            }
        }
        get t3() {
            return this.m1;
        }
        set t3(a) {
            let b = this.m1;
            this.m1 = a;
            if (b != this.m1) {
                this.jx("RaisedHoverTextColor", b, a);
            }
        }
        get t0() {
            return this.my;
        }
        set t0(a) {
            let b = this.my;
            this.my = a;
            if (b != this.my) {
                this.jx("RaisedFocusBackgroundColor", b, a);
            }
        }
        get t1() {
            return this.mz;
        }
        set t1(a) {
            let b = this.mz;
            this.mz = a;
            if (b != this.mz) {
                this.jx("RaisedFocusTextColor", b, a);
            }
        }
        get s5() {
            return this.l2;
        }
        set s5(a) {
            let b = this.l2;
            this.l2 = a;
            if (b != this.l2) {
                this.jx("FocusBackgroundColor", b, a);
            }
        }
        get s6() {
            return this.l3;
        }
        set s6(a) {
            let b = this.l3;
            this.l3 = a;
            if (b != this.l3) {
                this.jx("FocusTextColor", b, a);
            }
        }
        get ax() {
            return this.ao;
        }
        set ax(a) {
            let b = this.ao;
            this.ao = a;
            if (b != this.ao) {
                this.jx("BorderWidth", b, a);
            }
        }
        get kd() {
            return this.j5;
        }
        set kd(a) {
            let b = this.j5;
            this.j5 = a;
            {
                this.jx("CornerRadius", b, a);
            }
        }
        get tk() {
            return this.mi;
        }
        set tk(a) {
            let b = this.mi;
            this.mi = a;
            if (b != this.mi) {
                this.jx("OutlinedBackgroundColor", b, a);
            }
        }
        get tm() {
            return this.mk;
        }
        set tm(a) {
            let b = this.mk;
            this.mk = a;
            if (b != this.mk) {
                this.jx("OutlinedDisabledBackgroundColor", b, a);
            }
        }
        get tn() {
            return this.ml;
        }
        set tn(a) {
            let b = this.ml;
            this.ml = a;
            if (b != this.ml) {
                this.jx("OutlinedDisabledBorderColor", b, a);
            }
        }
        get to() {
            return this.mm;
        }
        set to(a) {
            let b = this.mm;
            this.mm = a;
            if (b != this.mm) {
                this.jx("OutlinedDisabledTextColor", b, a);
            }
        }
        get el() {
            return this.b3;
        }
        set el(a) {
            let b = this.b3;
            this.b3 = a;
            if (b != this.b3) {
                this.jx("OutlinedDisabledElevation", b, a);
            }
        }
        get tr() {
            return this.mp;
        }
        set tr(a) {
            let b = this.mp;
            this.mp = a;
            if (b != this.mp) {
                this.jx("OutlinedHoverBackgroundColor", b, a);
            }
        }
        get tl() {
            return this.mj;
        }
        set tl(a) {
            let b = this.mj;
            this.mj = a;
            if (b != this.mj) {
                this.jx("OutlinedBorderColor", b, a);
            }
        }
        get eo() {
            return this.b6;
        }
        set eo(a) {
            let b = this.b6;
            this.b6 = a;
            if (b != this.b6) {
                this.jx("OutlinedRestingElevation", b, a);
            }
        }
        get en() {
            return this.b5;
        }
        set en(a) {
            let b = this.b5;
            this.b5 = a;
            if (b != this.b5) {
                this.jx("OutlinedHoverElevation", b, a);
            }
        }
        get em() {
            return this.b4;
        }
        set em(a) {
            let b = this.b4;
            this.b4 = a;
            if (b != this.b4) {
                this.jx("OutlinedFocusElevation", b, a);
            }
        }
        get bi() {
            return this.au;
        }
        set bi(a) {
            let b = this.au;
            this.au = a;
            if (b != this.au) {
                this.jx("OutlinedBorderWidth", b, a);
            }
        }
        get kw() {
            return this.j9;
        }
        set kw(a) {
            let b = this.j9;
            this.j9 = a;
            {
                this.jx("OutlinedCornerRadius", b, a);
            }
        }
        get tt() {
            return this.mr;
        }
        set tt(a) {
            let b = this.mr;
            this.mr = a;
            if (b != this.mr) {
                this.jx("OutlinedRippleColor", b, a);
            }
        }
        get tu() {
            return this.ms;
        }
        set tu(a) {
            let b = this.ms;
            this.ms = a;
            if (b != this.ms) {
                this.jx("OutlinedTextColor", b, a);
            }
        }
        get ts() {
            return this.mq;
        }
        set ts(a) {
            let b = this.mq;
            this.mq = a;
            if (b != this.mq) {
                this.jx("OutlinedHoverTextColor", b, a);
            }
        }
        get tp() {
            return this.mn;
        }
        set tp(a) {
            let b = this.mn;
            this.mn = a;
            if (b != this.mn) {
                this.jx("OutlinedFocusBackgroundColor", b, a);
            }
        }
        get tq() {
            return this.mo;
        }
        set tq(a) {
            let b = this.mo;
            this.mo = a;
            if (b != this.mo) {
                this.jx("OutlinedFocusTextColor", b, a);
            }
        }
        get sj() {
            return this.lh;
        }
        set sj(a) {
            let b = this.lh;
            this.lh = a;
            if (b != this.lh) {
                this.jx("FabBackgroundColor", b, a);
            }
        }
        get sl() {
            return this.lj;
        }
        set sl(a) {
            let b = this.lj;
            this.lj = a;
            if (b != this.lj) {
                this.jx("FabDisabledBackgroundColor", b, a);
            }
        }
        get sm() {
            return this.lk;
        }
        set sm(a) {
            let b = this.lk;
            this.lk = a;
            if (b != this.lk) {
                this.jx("FabDisabledBorderColor", b, a);
            }
        }
        get sn() {
            return this.ll;
        }
        set sn(a) {
            let b = this.ll;
            this.ll = a;
            if (b != this.ll) {
                this.jx("FabDisabledTextColor", b, a);
            }
        }
        get d7() {
            return this.bp;
        }
        set d7(a) {
            let b = this.bp;
            this.bp = a;
            if (b != this.bp) {
                this.jx("FabDisabledElevation", b, a);
            }
        }
        get sq() {
            return this.l5;
        }
        set sq(a) {
            let b = this.l5;
            this.l5 = a;
            if (b != this.l5) {
                this.jx("FabHoverBackgroundColor", b, a);
            }
        }
        get sk() {
            return this.li;
        }
        set sk(a) {
            let b = this.li;
            this.li = a;
            if (b != this.li) {
                this.jx("FabBorderColor", b, a);
            }
        }
        get ea() {
            return this.bs;
        }
        set ea(a) {
            let b = this.bs;
            this.bs = a;
            if (b != this.bs) {
                this.jx("FabRestingElevation", b, a);
            }
        }
        get d9() {
            return this.br;
        }
        set d9(a) {
            let b = this.br;
            this.br = a;
            if (b != this.br) {
                this.jx("FabHoverElevation", b, a);
            }
        }
        get d8() {
            return this.bq;
        }
        set d8(a) {
            let b = this.bq;
            this.bq = a;
            if (b != this.bq) {
                this.jx("FabFocusElevation", b, a);
            }
        }
        get bd() {
            return this.ap;
        }
        set bd(a) {
            let b = this.ap;
            this.ap = a;
            if (b != this.ap) {
                this.jx("FabBorderWidth", b, a);
            }
        }
        get kt() {
            return this.j6;
        }
        set kt(a) {
            let b = this.j6;
            this.j6 = a;
            {
                this.jx("FabCornerRadius", b, a);
            }
        }
        get ss() {
            return this.lp;
        }
        set ss(a) {
            let b = this.lp;
            this.lp = a;
            if (b != this.lp) {
                this.jx("FabRippleColor", b, a);
            }
        }
        get st() {
            return this.lq;
        }
        set st(a) {
            let b = this.lq;
            this.lq = a;
            if (b != this.lq) {
                this.jx("FabTextColor", b, a);
            }
        }
        get sr() {
            return this.lo;
        }
        set sr(a) {
            let b = this.lo;
            this.lo = a;
            if (b != this.lo) {
                this.jx("FabHoverTextColor", b, a);
            }
        }
        get so() {
            return this.lm;
        }
        set so(a) {
            let b = this.lm;
            this.lm = a;
            if (b != this.lm) {
                this.jx("FabFocusBackgroundColor", b, a);
            }
        }
        get sp() {
            return this.ln;
        }
        set sp(a) {
            let b = this.ln;
            this.ln = a;
            if (b != this.ln) {
                this.jx("FabFocusTextColor", b, a);
            }
        }
        get s9() {
            return this.l7;
        }
        set s9(a) {
            let b = this.l7;
            this.l7 = a;
            if (b != this.l7) {
                this.jx("IconBackgroundColor", b, a);
            }
        }
        get tb() {
            return this.l9;
        }
        set tb(a) {
            let b = this.l9;
            this.l9 = a;
            if (b != this.l9) {
                this.jx("IconDisabledBackgroundColor", b, a);
            }
        }
        get tc() {
            return this.ma;
        }
        set tc(a) {
            let b = this.ma;
            this.ma = a;
            if (b != this.ma) {
                this.jx("IconDisabledBorderColor", b, a);
            }
        }
        get td() {
            return this.mb;
        }
        set td(a) {
            let b = this.mb;
            this.mb = a;
            if (b != this.mb) {
                this.jx("IconDisabledTextColor", b, a);
            }
        }
        get eh() {
            return this.bz;
        }
        set eh(a) {
            let b = this.bz;
            this.bz = a;
            if (b != this.bz) {
                this.jx("IconDisabledElevation", b, a);
            }
        }
        get tg() {
            return this.me;
        }
        set tg(a) {
            let b = this.me;
            this.me = a;
            if (b != this.me) {
                this.jx("IconHoverBackgroundColor", b, a);
            }
        }
        get ta() {
            return this.l8;
        }
        set ta(a) {
            let b = this.l8;
            this.l8 = a;
            if (b != this.l8) {
                this.jx("IconBorderColor", b, a);
            }
        }
        get ek() {
            return this.b2;
        }
        set ek(a) {
            let b = this.b2;
            this.b2 = a;
            if (b != this.b2) {
                this.jx("IconRestingElevation", b, a);
            }
        }
        get ej() {
            return this.b1;
        }
        set ej(a) {
            let b = this.b1;
            this.b1 = a;
            if (b != this.b1) {
                this.jx("IconHoverElevation", b, a);
            }
        }
        get ei() {
            return this.b0;
        }
        set ei(a) {
            let b = this.b0;
            this.b0 = a;
            if (b != this.b0) {
                this.jx("IconFocusElevation", b, a);
            }
        }
        get bf() {
            return this.ar;
        }
        set bf(a) {
            let b = this.ar;
            this.ar = a;
            if (b != this.ar) {
                this.jx("IconBorderWidth", b, a);
            }
        }
        get kv() {
            return this.j8;
        }
        set kv(a) {
            let b = this.j8;
            this.j8 = a;
            {
                this.jx("IconCornerRadius", b, a);
            }
        }
        get ti() {
            return this.mg;
        }
        set ti(a) {
            let b = this.mg;
            this.mg = a;
            if (b != this.mg) {
                this.jx("IconRippleColor", b, a);
            }
        }
        get tj() {
            return this.mh;
        }
        set tj(a) {
            let b = this.mh;
            this.mh = a;
            if (b != this.mh) {
                this.jx("IconTextColor", b, a);
            }
        }
        get th() {
            return this.mf;
        }
        set th(a) {
            let b = this.mf;
            this.mf = a;
            if (b != this.mf) {
                this.jx("IconHoverTextColor", b, a);
            }
        }
        get te() {
            return this.mc;
        }
        set te(a) {
            let b = this.mc;
            this.mc = a;
            if (b != this.mc) {
                this.jx("IconFocusBackgroundColor", b, a);
            }
        }
        get tf() {
            return this.md;
        }
        set tf(a) {
            let b = this.md;
            this.md = a;
            if (b != this.md) {
                this.jx("IconFocusTextColor", b, a);
            }
        }
        get h7() {
            return this.e6;
        }
        set h7(a) {
            let b = this.e6;
            this.e6 = a;
            if (b != this.e6) {
                this.jx("InputId", b, a);
            }
        }
        get h5() {
            return this.e5;
        }
        set h5(a) {
            let b = this.e5;
            this.e5 = a;
            if (b != this.e5) {
                this.jx("Id", b, a);
            }
        }
        get gb() {
            return this.e2;
        }
        set gb(a) {
            let b = this.e2;
            this.e2 = a;
            if (b != this.e2) {
                this.jx("Display", b, a);
            }
        }
        get he() {
            return this.e3;
        }
        set he(a) {
            let b = this.e3;
            this.e3 = a;
            if (b != this.e3) {
                this.jx("FlexDirection", b, a);
            }
        }
        get hg() {
            return this.e4;
        }
        set hg(a) {
            let b = this.e4;
            this.e4 = a;
            if (b != this.e4) {
                this.jx("FlexGrow", b, a);
            }
        }
        get fv() {
            return this.e0;
        }
        set fv(a) {
            let b = this.e0;
            this.e0 = a;
            if (b != this.e0) {
                this.jx("AlignItems", b, a);
            }
        }
        get ez() {
            return this.ex;
        }
        set ez(a) {
            let b = this.ex;
            this.ex = a;
            if (b != this.ex) {
                this.jx("Value", b, a);
            }
        }
        get ig() {
            return this.e7;
        }
        set ig(a) {
            let b = this.e7;
            this.e7 = a;
            if (b != this.e7) {
                this.jx("Name", b, a);
            }
        }
        get ah() {
            return this.z;
        }
        get eu() {
            return this.cc;
        }
        set eu(a) {
            let b = this.cc;
            this.cc = a;
            this.z = true;
            if (b != this.cc) {
                this.jx("TabIndex", b, a);
            }
        }
        get focused() {
            return this.ab;
        }
        set focused(a) {
            let b = this.ab;
            this.ab = a;
            if (b != this.ab) {
                this.jx("IsFocused", b, a);
            }
        }
        get ak() {
            return this.ac;
        }
        set ak(a) {
            let b = this.ac;
            this.ac = a;
            if (b != this.ac) {
                this.jx("IsFocusStyleEnabled", b, a);
            }
        }
        get fx() {
            return this.e1;
        }
        set fx(a) {
            let b = this.e1;
            this.e1 = a;
            if (b != this.e1) {
                this.jx("AriaLabel", b, a);
            }
        }
        get af() {
            return this.x;
        }
        set af(a) {
            let b = this.x;
            this.x = a;
            if (b != this.x) {
                this.jx("DisableRipple", b, a);
            }
        }
        get ae() {
            return this.w;
        }
        set ae(a) {
            let b = this.w;
            this.w = a;
            if (b != this.w) {
                this.jx("ActualDisableRipple", b, a);
            }
        }
        get al() {
            return this.ad;
        }
        set al(a) {
            let b = this.ad;
            this.ad = a;
            if (b != this.ad) {
                this.jx("IsHover", b, a);
            }
        }
        get disabled() {
            return this.aa;
        }
        set disabled(a) {
            let b = this.aa;
            this.aa = a;
            if (b != this.aa) {
                this.jx("IsDisabled", b, a);
            }
        }
        get ag() {
            return this.y;
        }
        set ag(a) {
            let b = this.y;
            this.y = a;
            if (b != this.y) {
                this.jx("DisableTransitions", b, a);
            }
        }
        jy() {
            this.jp();
            this.view.ap(this.nh, BrushUtil.g(0, 0, 0, 0));
            this.j1();
        }
        get a() {
            let a = new AnimationKeyFrameEffectTiming();
            a.g = 300;
            a.b = 0;
            return a;
        }
        jp() {
            this.view.z();
        }
        onDetachedFromUI() {
            this.view.ad();
        }
        onAttachedToUI() {
            this.view.ab();
        }
        appendButtonContent(a) {
            this.view.y(a);
        }
        am() {
            return this.view.f();
        }
        ey() {
            return this.p();
        }
        gf() {
            let a = this.p();
            return a.ab();
        }
        p() {
            let a = new ButtonVisualModelExport();
            a.s = this.view.j.rootWrapper.getNativeElement().offsetTop;
            a.q = this.view.j.rootWrapper.getNativeElement().offsetLeft;
            a.p = this.view.j.rootWrapper.getNativeElement().offsetHeight;
            a.t = this.view.j.rootWrapper.getNativeElement().offsetWidth;
            a.h = this.d;
            a.i = this.g;
            a.c = AppearanceHelper.a(this.m8);
            a.g = AppearanceHelper.a(this.nj);
            a.d = AppearanceHelper.a(this.ng);
            a.a = AppearanceHelper.a(this.m6);
            a.u = this.cg;
            a.e = AppearanceHelper.a(this.nh);
            a.ah = this.kb;
            a.b = AppearanceHelper.a(this.m7);
            a.j = this.k;
            a.f = AppearanceHelper.a(this.ni);
            a.o = this.aw;
            a.x = this.gb;
            a.y = this.he;
            a.z = this.hg;
            a.w = this.fv;
            a.v = this.view.q.getNativeElement().innerText;
            a.aa = this.ig;
            a.m = this.focused;
            a.k = this.ae;
            a.n = this.al;
            a.l = this.disabled;
            if (this.v != null) {
                let b = this.v;
                if (this.view != null && b.n == null) {
                    let c = this.view.j;
                    let d = FontUtil.getFontInfoFromString(c, b.fontString);
                    b.n = d.n;
                    b.d = d.d;
                    b.o = d.o;
                    b.q = d.q;
                    b.r = d.r;
                    b.s = d.s;
                }
                if (b.n != null) {
                    a.ac = b.n;
                }
                if (!isNaN_(b.d)) {
                    a.r = b.d;
                }
                if (b.s != null) {
                    a.ag = b.s;
                }
                if (b.q != null) {
                    a.ae = b.q;
                }
                if (b.o != null) {
                    a.ad = b.q;
                }
            }
            return a;
        }
        provideContainer(a) {
            this.view.ac(a);
        }
    }
    Button.$t = /*@__PURE__*/ markType(Button, 'Button', Base.$, [INotifyPropertyChanged_$type]);
    Button.rr = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.rf = /*@__PURE__*/ BrushUtil.g(30, 0, 0, 0);
    Button.p6 = /*@__PURE__*/ BrushUtil.g(20, 0, 0, 0);
    Button.n = 1;
    Button.ri = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.rj = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.rk = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.du = 0;
    Button.dx = 2;
    Button.dw = 4;
    Button.dv = 8;
    Button.rh = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.rp = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.kq = /*@__PURE__*/ new CornerRadius(0, 4);
    Button.ba = NaN;
    Button.rg = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.rq = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.ro = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.rn = /*@__PURE__*/ BrushUtil.g(255, 220, 83, 145);
    Button.rl = /*@__PURE__*/ BrushUtil.g(255, 220, 83, 145);
    Button.rm = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.qk = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.ql = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.qm = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.dh = 0;
    Button.dk = 0;
    Button.dj = 0;
    Button.di = 0;
    Button.qj = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.qr = null;
    Button.kn = /*@__PURE__*/ new CornerRadius(0, 4);
    Button.a7 = NaN;
    Button.qi = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.qs = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.qq = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.qp = /*@__PURE__*/ BrushUtil.g(13, 228, 28, 119);
    Button.qn = /*@__PURE__*/ BrushUtil.g(30, 228, 28, 119);
    Button.qo = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.q6 = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.q7 = /*@__PURE__*/ BrushUtil.g(30, 0, 0, 0);
    Button.q8 = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.dq = 0;
    Button.dt = 0;
    Button.ds = 0;
    Button.dr = 0;
    Button.q5 = /*@__PURE__*/ BrushUtil.g(30, 0, 0, 0);
    Button.rd = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.kp = /*@__PURE__*/ new CornerRadius(0, 4);
    Button.a9 = 1;
    Button.q4 = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.re = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.rc = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.rb = /*@__PURE__*/ BrushUtil.g(13, 228, 28, 119);
    Button.q9 = /*@__PURE__*/ BrushUtil.g(30, 228, 28, 119);
    Button.ra = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.qv = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.qw = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.qx = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.dl = 0;
    Button.dp = 0;
    Button.dn = 0;
    Button.dm = 0;
    Button.qu = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.q2 = null;
    Button.ko = /*@__PURE__*/ new CornerRadius(0, 28);
    Button.a8 = NaN;
    Button.qt = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.q3 = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.q1 = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.q0 = /*@__PURE__*/ BrushUtil.g(30, 0, 0, 0);
    Button.qy = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.qz = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.p9 = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.qa = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.qb = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.dd = 0;
    Button.dg = 6;
    Button.df = 12;
    Button.de = 12;
    Button.p8 = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.qg = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.km = /*@__PURE__*/ new CornerRadius(0, 28);
    Button.a6 = NaN;
    Button.p7 = /*@__PURE__*/ BrushUtil.g(255, 228, 28, 119);
    Button.qh = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.qf = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.qe = /*@__PURE__*/ BrushUtil.g(255, 220, 83, 145);
    Button.qc = /*@__PURE__*/ BrushUtil.g(255, 220, 83, 145);
    Button.qd = /*@__PURE__*/ BrushUtil.g(255, 0, 0, 0);
    Button.o8 = /*@__PURE__*/ BrushUtil.g(100, 33, 150, 243);
    Button.ow = /*@__PURE__*/ BrushUtil.g(70, 33, 150, 243);
    Button.nn = /*@__PURE__*/ BrushUtil.g(40, 33, 150, 243);
    Button.l = 2;
    Button.oz = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.o0 = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.o1 = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.cx = 0;
    Button.c0 = 0;
    Button.cz = 0;
    Button.cy = 2;
    Button.oy = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.o6 = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.ki = /*@__PURE__*/ new CornerRadius(0, 3);
    Button.a2 = NaN;
    Button.ox = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.o7 = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.o5 = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.o4 = /*@__PURE__*/ BrushUtil.g(255, 27, 135, 220);
    Button.o2 = /*@__PURE__*/ BrushUtil.g(255, 21, 112, 184);
    Button.o3 = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.n1 = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.n2 = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.n3 = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.cl = 0;
    Button.co = 0;
    Button.cn = 0;
    Button.cm = 2;
    Button.n0 = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.n8 = /*@__PURE__*/ BrushUtil.g(100, 33, 150, 243);
    Button.kf = /*@__PURE__*/ new CornerRadius(0, 3);
    Button.az = NaN;
    Button.nz = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.n9 = /*@__PURE__*/ BrushUtil.g(255, 119, 119, 119);
    Button.n7 = /*@__PURE__*/ BrushUtil.g(255, 119, 119, 119);
    Button.n6 = /*@__PURE__*/ BrushUtil.g(26, 33, 150, 243);
    Button.n4 = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.n5 = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.on = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.oo = /*@__PURE__*/ BrushUtil.g(30, 0, 0, 0);
    Button.op = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.ct = 0;
    Button.cw = 0;
    Button.cv = 0;
    Button.cu = 2;
    Button.om = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.ou = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.kh = /*@__PURE__*/ new CornerRadius(0, 3);
    Button.a1 = 1;
    Button.ol = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.ov = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.ot = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.os = /*@__PURE__*/ BrushUtil.g(26, 33, 150, 243);
    Button.oq = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.or = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.oc = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.od = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.oe = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.cp = 0;
    Button.cs = 0;
    Button.cr = 0;
    Button.cq = 2;
    Button.ob = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.oj = null;
    Button.kg = /*@__PURE__*/ new CornerRadius(0, 28);
    Button.a0 = NaN;
    Button.oa = /*@__PURE__*/ BrushUtil.g(0, 0, 0, 0);
    Button.ok = /*@__PURE__*/ BrushUtil.g(255, 119, 119, 119);
    Button.oi = /*@__PURE__*/ BrushUtil.g(255, 119, 119, 119);
    Button.oh = /*@__PURE__*/ BrushUtil.g(30, 0, 0, 0);
    Button.of = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.og = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.nq = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.nr = /*@__PURE__*/ BrushUtil.g(10, 0, 0, 0);
    Button.ns = /*@__PURE__*/ BrushUtil.g(66, 0, 0, 0);
    Button.ch = 0;
    Button.ck = 0;
    Button.cj = 0;
    Button.ci = 2;
    Button.np = /*@__PURE__*/ BrushUtil.g(138, 0, 0, 0);
    Button.nx = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.ke = /*@__PURE__*/ new CornerRadius(0, 28);
    Button.ay = NaN;
    Button.no = /*@__PURE__*/ BrushUtil.g(255, 33, 150, 243);
    Button.ny = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.nw = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.nv = /*@__PURE__*/ BrushUtil.g(255, 27, 135, 220);
    Button.nt = /*@__PURE__*/ BrushUtil.g(255, 21, 112, 184);
    Button.nu = /*@__PURE__*/ BrushUtil.g(255, 255, 255, 255);
    Button.c = 2;
    Button.ew = 0;
    Button.ev = 0;
    return Button;
})();
