// igniteui-react-inputs/esm2015/lib/CalendarVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let CalendarVisualModelExport = /*@__PURE__*/ (() => {
    class CalendarVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.h = 0;
            this.g = 0;
            this.d = null;
            this.f = null;
            this.b = null;
            this.e = null;
            this.c = null;
            this.a = null;
            this.o = null;
            this.p = null;
            this.q = null;
            this.j = 0;
            this.m = null;
            this.n = null;
            this.k = null;
            this.i = null;
        }
        l() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("Density : " + this.h + ", ");
            a.l("BaseTheme : " + this.g + ", ");
            a.l("SelectedDateBackgroundColor : " + this.d.serialize() + ", ");
            a.l("SelectedFocusDateBackgroundColor : " + this.f.serialize() + ", ");
            a.l("FocusDateBackgroundColor : " + this.b.serialize() + ", ");
            a.l("SelectedDateTextColor : " + this.e.serialize() + ", ");
            a.l("FocusDateTextColor : " + this.c.serialize() + ", ");
            a.l("CurrentDateTextColor : " + this.a.serialize() + ", ");
            if (!stringIsNullOrEmpty(this.o)) {
                a.l("textStyleFontStyle : \"" + this.o + "\", ");
            }
            if (!stringIsNullOrEmpty(this.p)) {
                a.l("textStyleFontVariant : \"" + this.p + "\", ");
            }
            if (!stringIsNullOrEmpty(this.q)) {
                a.l("textStyleFontWeight : \"" + this.q + "\", ");
            }
            if (!stringIsNullOrEmpty(this.m)) {
                a.l("textStyleFontFamily : \"" + this.m + "\", ");
            }
            if (!stringIsNullOrEmpty(this.n)) {
                a.l("textStyleFontStretch : \"" + this.n + "\", ");
            }
            if (!isNaN_(this.j)) {
                a.l("textStyleFontSize : " + this.j + ", ");
            }
            a.l("activePicker : \"" + this.k + "\", ");
            a.l("buttonsModels: [");
            for (let b = 0; b < this.i.count; b++) {
                if (b != 0) {
                    a.l(",");
                }
                let c = this.i._inner[b].aa.charAt(0).toString().toLowerCase() + this.i._inner[b].aa.substr(1);
                a.l(this.i._inner[b].ab());
            }
            a.l("]");
            a.l("}");
            return a.toString();
        }
    }
    CalendarVisualModelExport.$t = /*@__PURE__*/ markType(CalendarVisualModelExport, 'CalendarVisualModelExport');
    return CalendarVisualModelExport;
})();
