// igniteui-react-inputs/esm2015/lib/DatePickerVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let DatePickerVisualModelExport = /*@__PURE__*/ (() => {
    class DatePickerVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.b = 0;
            this.a = 0;
            this.j = null;
            this.k = null;
            this.l = null;
            this.f = 0;
            this.h = null;
            this.i = null;
            this.e = null;
            this.d = null;
            this.c = null;
        }
        g() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("density : " + this.b + ", ");
            a.l("baseTheme : " + this.a + ", ");
            if (!stringIsNullOrEmpty(this.j)) {
                a.l("textStyleFontStyle : \"" + this.j + "\", ");
            }
            if (!stringIsNullOrEmpty(this.k)) {
                a.l("textStyleFontVariant : \"" + this.k + "\", ");
            }
            if (!stringIsNullOrEmpty(this.l)) {
                a.l("textStyleFontWeight : \"" + this.l + "\", ");
            }
            if (!stringIsNullOrEmpty(this.h)) {
                a.l("textStyleFontFamily : \"" + this.h + "\", ");
            }
            if (!stringIsNullOrEmpty(this.i)) {
                a.l("textStyleFontStretch : \"" + this.i + "\", ");
            }
            if (!isNaN_(this.f)) {
                a.l("textStyleFontSize : " + this.f + ", ");
            }
            let b = "popupModel";
            a.l(b + ": " + this.e.f() + ", ");
            let c = "inputGroupModel";
            a.l(c + ": " + this.d.y() + ", ");
            let d = "calendarModel";
            a.l(d + ": " + this.c.l() + ", ");
            a.l("}");
            return a.toString();
        }
    }
    DatePickerVisualModelExport.$t = /*@__PURE__*/ markType(DatePickerVisualModelExport, 'DatePickerVisualModelExport');
    return DatePickerVisualModelExport;
})();
