// igniteui-react-inputs/esm2015/lib/igr-checkbox-change-event-args.js
import { CheckboxChangeEventArgs as CheckboxChangeEventArgs_internal } from "./CheckboxChangeEventArgs";
import { ensureBool } from "igniteui-react-core";
export class IgrCheckboxChangeEventArgs {
    createImplementation() {
        return new CheckboxChangeEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get isChecked() {
        return this.i.isChecked;
    }
    set isChecked(v) {
        this.i.isChecked = ensureBool(v);
    }
    get isIndeterminate() {
        return this.i.isIndeterminate;
    }
    set isIndeterminate(v) {
        this.i.isIndeterminate = ensureBool(v);
    }
}
