// igniteui-react-inputs/esm2015/lib/InputGroupVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let InputGroupVisualModelExport = /*@__PURE__*/ (() => {
    class InputGroupVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.f = 0;
            this.e = 0;
            this.b = null;
            this.o = 0;
            this.p = 0;
            this.z = null;
            this.aa = null;
            this.x = 0;
            this.w = 0;
            this.t = 0;
            this.c = null;
            this.d = null;
            this.u = 0;
            this.s = 0;
            this.q = 0;
            this.r = 0;
            this.v = 0;
            this.a = null;
            this.g = false;
            this.j = false;
            this.i = false;
            this.h = false;
            this.m = null;
            this.l = null;
            this.k = null;
            this.n = null;
        }
        y() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("actualDensity : " + this.f + ", ");
            a.l("baseTheme : " + this.e + ", ");
            a.l("actualBorderColor : " + this.b.serialize() + ", ");
            a.l("actualBorderWidth : " + this.o + ", ");
            a.l("actualFocusBorderWidth : " + this.p + ", ");
            a.l("actualContentPadding : { left: " + this.aa.c + ",top: " + this.aa.e + ",right: " + this.aa.d + ",bottom: " + this.aa.b + "},");
            a.l("actualUnderlineWidth : " + this.x + ", ");
            a.l("actualUnderlineRippleWidth : " + this.w + ", ");
            a.l("actualBackgroundColor : " + this.a.serialize() + ", ");
            a.l("actualHoverUnderlineWidth : " + this.t + ", ");
            a.l("actualUnderlineColor : " + this.c.serialize() + ", ");
            a.l("actualUnderlineRippleColor : " + this.d.serialize() + ", ");
            a.l("actualUnderlineOpacity : " + this.u + ", ");
            a.l("actualHoverUnderlineOpacity : " + this.s + ", ");
            a.l("actualFocusUnderlineOpacity : " + this.q + ", ");
            a.l("actualFocusUnderlineRippleOpacity : " + this.r + ", ");
            a.l("actualUnderlineRippleOpacity : " + this.v + ", ");
            a.l("actualBackgroundColor : " + this.a.serialize() + ", ");
            a.l("actualIsExpanded : " + this.g + ", ");
            a.l("isHovered : " + this.j + ", ");
            a.l("isFocused : " + this.i + ", ");
            a.l("isDisabled : " + this.h + ", ");
            a.l("prefixModels : [");
            for (let b = 0; b < this.m.count; b++) {
                if (b != 0) {
                    a.l(",");
                }
                a.l(this.m._inner[b].e());
            }
            a.l("],");
            a.l("labelModels: [");
            for (let c = 0; c < this.l.count; c++) {
                if (c != 0) {
                    a.l(",");
                }
                let d = this.l._inner[c].q.charAt(0).toString().toLowerCase() + this.l._inner[c].q.substr(1);
                a.l(this.l._inner[c].r());
            }
            a.l("],");
            a.l("inputModels : [");
            for (let e = 0; e < this.k.count; e++) {
                if (e != 0) {
                    a.l(",");
                }
                let f = this.k._inner[e].k.charAt(0).toString().toLowerCase() + this.k._inner[e].k.substr(1);
                a.l(this.k._inner[e].n());
            }
            a.l("],");
            a.l("suffixModels : [");
            for (let g = 0; g < this.n.count; g++) {
                if (g != 0) {
                    a.l(",");
                }
                let h = this.n._inner[g].d.charAt(0).toString().toLowerCase() + this.n._inner[g].d.substr(1);
                a.l(this.n._inner[g].e());
            }
            a.l("]");
            a.l("}");
            return a.toString();
        }
    }
    InputGroupVisualModelExport.$t = /*@__PURE__*/ markType(InputGroupVisualModelExport, 'InputGroupVisualModelExport');
    return InputGroupVisualModelExport;
})();
