// igniteui-react-inputs/esm2015/lib/LabelVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let LabelVisualModelExport = /*@__PURE__*/ (() => {
    class LabelVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.c = 0;
            this.e = 0;
            this.d = 0;
            this.b = null;
            this.a = null;
            this.v = null;
            this.w = null;
            this.x = null;
            this.l = 0;
            this.t = null;
            this.u = null;
            this.f = 0;
            this.j = false;
            this.h = false;
            this.n = null;
            this.o = null;
            this.p = null;
            this.m = null;
            this.s = null;
            this.i = false;
            this.k = false;
            this.g = false;
            this.q = null;
        }
        r() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("baseTheme : " + this.c + ", ");
            a.l("density : " + this.e + ", ");
            a.l("actualDensity : " + this.d + ", ");
            a.l("actualTextColor : " + this.b.serialize() + ", ");
            a.l("actualHighlightTextColor : " + this.a.serialize() + ", ");
            if (!stringIsNullOrEmpty(this.v)) {
                a.l("textStyleFontStyle : \"" + this.v + "\", ");
            }
            if (!stringIsNullOrEmpty(this.w)) {
                a.l("textStyleFontVariant : \"" + this.w + "\", ");
            }
            if (!stringIsNullOrEmpty(this.x)) {
                a.l("textStyleFontWeight : \"" + this.x + "\", ");
            }
            if (!stringIsNullOrEmpty(this.t)) {
                a.l("textStyleFontFamily : \"" + this.t + "\", ");
            }
            if (!stringIsNullOrEmpty(this.u)) {
                a.l("textStyleFontStretch : \"" + this.u + "\", ");
            }
            if (!isNaN_(this.l)) {
                a.l("textStyleFontSize : " + this.l + ", ");
            }
            a.l("shiftType : " + this.f + ", ");
            a.l("isShifted : " + this.j + ", ");
            a.l("isHighlighted : " + this.h + ", ");
            a.l("display : \"" + this.n + "\", ");
            a.l("flexDirection : \"" + this.o + "\", ");
            a.l("flexGrow : \"" + this.p + "\", ");
            a.l("alignItems :\"" + this.m + "\", ");
            a.l("text : \"" + this.s + "\", ");
            a.l("isHover : " + this.i + ", ");
            a.l("value : " + this.k + ", ");
            a.l("isDisabled : " + this.g + ", ");
            a.l("name : \"" + this.q + "\", ");
            a.l("}");
            return a.toString();
        }
    }
    LabelVisualModelExport.$t = /*@__PURE__*/ markType(LabelVisualModelExport, 'LabelVisualModelExport');
    return LabelVisualModelExport;
})();
