// igniteui-react-inputs/esm2015/lib/PrefixVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, enumGetBox, markType } from "igniteui-react-core";
import { PrefixShiftType_$type } from "./PrefixShiftType";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let PrefixVisualModelExport = /*@__PURE__*/ (() => {
    class PrefixVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.a = 0;
            this.c = false;
            this.b = false;
            this.d = null;
        }
        e() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("shiftType : " + enumGetBox(PrefixShiftType_$type, this.a) + ", ");
            a.l("isHover : " + this.c + ", ");
            a.l("isDisabled : " + this.b + ", ");
            a.l("name : \"" + this.d + "\", ");
            a.l("}");
            return a.toString();
        }
    }
    PrefixVisualModelExport.$t = /*@__PURE__*/ markType(PrefixVisualModelExport, 'PrefixVisualModelExport');
    return PrefixVisualModelExport;
})();
