// igniteui-react-inputs/esm2015/lib/RippleVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let RippleVisualModelExport = /*@__PURE__*/ (() => {
    class RippleVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.b = null;
            this.f = 0;
            this.d = false;
            this.c = false;
            this.e = false;
        }
        g() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("ActualHoverColor : " + this.a.serialize() + ", ");
            a.l("ActualRippleColor : " + this.b.serialize() + ", ");
            a.l("RippleDuration : " + this.f + ", ");
            a.l("IsDisabled : " + this.d + ", ");
            a.l("IsCentered : " + this.c + ", ");
            a.l("IsHoverEnabled : " + this.e + ", ");
            a.l("}");
            return a.toString();
        }
    }
    RippleVisualModelExport.$t = /*@__PURE__*/ markType(RippleVisualModelExport, 'RippleVisualModelExport');
    return RippleVisualModelExport;
})();
