// igniteui-react-inputs/esm2015/lib/ThemeResolver.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ThemeResolver = /*@__PURE__*/ (() => {
    class ThemeResolver extends Base {
        static e(a, b, c) {
            if (a == 1) {
                return c;
            }
            return b;
        }
        static c(a, b, c) {
            if (a == 1) {
                return c;
            }
            return b;
        }
        static b(a, b, c) {
            if (a == 1) {
                return c;
            }
            return b;
        }
        static d(a, b, c) {
            if (a == 1) {
                return c;
            }
            return b;
        }
        static f(a, b, c) {
            if (a == 1) {
                return c;
            }
            return b;
        }
        static a(a, b, c) {
            if (a == 1) {
                return c;
            }
            return b;
        }
    }
    ThemeResolver.$t = /*@__PURE__*/ markType(ThemeResolver, 'ThemeResolver');
    return ThemeResolver;
})();
