// igniteui-react-grids/esm2015/lib/CalculationParameters.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CalculationParameters = /*@__PURE__*/ (() => {
    class CalculationParameters extends Base {
        constructor() {
            super(...arguments);
            this.j = 0;
            this.g = 0;
            this.h = 0;
            this.i = 0;
            this.b = null;
            this.a = null;
            this.k = 0;
            this.m = 0;
            this.l = 0;
            this.n = 0;
            this.d = 0;
            this.f = 0;
            this.e = 0;
            this.c = false;
            this.o = null;
        }
    }
    CalculationParameters.$t = /*@__PURE__*/ markType(CalculationParameters, 'CalculationParameters');
    return CalculationParameters;
})();
