// igniteui-react-grids/esm2015/lib/CanvasPlaceholderView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
import { truncate } from "igniteui-react-core";
/**
 * @hidden
 */
export let CanvasPlaceholderView = /*@__PURE__*/ (() => {
    class CanvasPlaceholderView extends Base {
        constructor() {
            super();
            this.f = null;
            this.j = new Color();
            this.g = 1;
            this.a = true;
            this.c = 0;
            this.b = 0;
            this.k = new Color();
        }
        get l() {
            return this.j;
        }
        set l(a) {
            this.j = a;
            this.i();
        }
        get h() {
            return this.g;
        }
        set h(a) {
            this.a = true;
            this.g = a;
            this.i();
        }
        get e() {
            return this.c;
        }
        set e(a) {
            this.c = a;
            this.i();
        }
        get d() {
            return this.b;
        }
        set d(a) {
            this.b = a;
            this.i();
        }
        i() {
            let a = this.e;
            let b = this.d;
            this.k = Color.u(truncate((this.j.l * this.g)), this.j.o, this.j.n, this.j.m);
        }
    }
    CanvasPlaceholderView.$t = /*@__PURE__*/ markType(CanvasPlaceholderView, 'CanvasPlaceholderView');
    return CanvasPlaceholderView;
})();
