// igniteui-react-grids/esm2015/lib/CellRange.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CellRange = /*@__PURE__*/ (() => {
    class CellRange extends Base {
        constructor() {
            super(...arguments);
            this.g = -2147483648;
            this.e = -2147483648;
            this.h = -2147483648;
            this.f = -2147483648;
        }
        get k() {
            return this.g;
        }
        set k(a) {
            this.g = a;
        }
        get i() {
            return this.e;
        }
        set i(a) {
            this.e = a;
        }
        get l() {
            return this.h;
        }
        set l(a) {
            this.h = a;
        }
        get j() {
            return this.f;
        }
        set j(a) {
            this.f = a;
        }
        b(a, b) {
            return this.d(a) && this.c(b);
        }
        d(a) {
            return a >= this.l && a <= this.j;
        }
        c(a) {
            return a >= this.k && a <= this.i;
        }
        a() {
            let a = new CellRange();
            a.k = this.k;
            a.i = this.i;
            a.l = this.l;
            a.j = this.j;
            return a;
        }
    }
    CellRange.$t = /*@__PURE__*/ markType(CellRange, 'CellRange');
    return CellRange;
})();
