// igniteui-react-grids/esm2015/lib/CellModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
import { stringIsNullOrEmpty, stringEndsWith } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let CellModelExport = /*@__PURE__*/ (() => {
    class CellModelExport extends Base {
        constructor() {
            super(...arguments);
            this.r = 0;
            this.s = 0;
            this.n = 0;
            this.o = 0;
            this.k = 0;
            this.q = 0;
            this.m = 0;
            this.h = 0;
            this.l = 0;
            this.i = 0;
            this.a = null;
            this.b = null;
            this.d = 0;
            this.e = 0;
            this.aa = null;
            this.c = null;
            this.w = null;
            this.x = null;
            this.y = null;
            this.j = 0;
            this.u = null;
            this.v = null;
            this.p = 0;
            this.f = 0;
            this.t = 0;
            this.g = false;
            this.z = null;
        }
        ab() {
            let a = new StringBuilder(0);
            a.l("{ ");
            a.l("type : \"" + this.ac + "\", ");
            a.l("x : " + this.r + ", ");
            a.l("y : " + this.s + ", ");
            a.l("snappedX : " + this.n + ", ");
            a.l("snappedY : " + this.o + ", ");
            a.l("height : " + this.k + ", ");
            a.l("width : " + this.q + ", ");
            a.l("lineBreakMode: \"" + this.z + "\", ");
            a.l("selectedStatus : " + this.m + ", ");
            a.l("activationStatus : " + this.h + ", ");
            a.l("virtualizationPercentage : " + this.p + ", ");
            a.l("backgroundBrushData : " + this.a.serialize() + ", ");
            if (this.b != null) {
                a.l("borderBrushData : " + this.b.serialize() + ", ");
            }
            a.l("opacity : " + this.l + ", ");
            a.l("contentOpacity : " + this.i + ", ");
            if (this.d != null) {
                a.l("horizontalAlignment : " + this.d + ", ");
            }
            if (this.e != null) {
                a.l("verticalAlignment : " + this.e + ", ");
            }
            a.l("renderValue : \"" + this.aa + "\", ");
            if (this.c != null) {
                a.l("textColor : " + this.c.serialize() + ", ");
            }
            if (!stringIsNullOrEmpty(this.w)) {
                a.l("fontStyle : \"" + this.w + "\", ");
            }
            if (!stringIsNullOrEmpty(this.x)) {
                a.l("fontVariant : \"" + this.x + "\", ");
            }
            if (!stringIsNullOrEmpty(this.y)) {
                a.l("fontWeight : \"" + this.y + "\", ");
            }
            if (!stringIsNullOrEmpty(this.u)) {
                a.l("fontFamily : \"" + this.u + "\", ");
            }
            if (!stringIsNullOrEmpty(this.v)) {
                a.l("fontStretch : \"" + this.v + "\", ");
            }
            if (!isNaN_(this.j)) {
                a.l("fontSize : " + this.j + ", ");
            }
            a.l("sortDirection : " + this.f + ", ");
            a.l("indent : " + this.t + ", ");
            a.l("isCollapsable : " + this.g.toString().toLowerCase() + ", ");
            this.ad(a);
            let b = a.toString();
            if (stringEndsWith(b, ", ")) {
                b = b.substr(0, b.length - 2);
            }
            b += "} ";
            return b;
        }
        get_ac() {
            return "Cell";
        }
        get ac() {
            return this.get_ac();
        }
        ad(a) {
        }
    }
    CellModelExport.$t = /*@__PURE__*/ markType(CellModelExport, 'CellModelExport');
    return CellModelExport;
})();
