// igniteui-react-grids/esm2015/lib/ColumnPinningView_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType, INotifyPropertyChanged_$type, PropertyChangedEventArgs, runOn, delegateRemove, delegateCombine, enumGetBox, fromEnum } from "igniteui-react-core";
import { BaseControlTheme_$type } from "igniteui-react-core";
import { ControlDisplayDensity_$type } from "igniteui-react-core";
import { BrushUtil } from "igniteui-react-core";
import { ObservableCollection$1 } from "igniteui-react-core";
import { ColumnPinningInfo } from "./ColumnPinningInfo";
/**
 * @hidden
 */
export let ColumnPinningView = /*@__PURE__*/ (() => {
    class ColumnPinningView extends Base {
        constructor() {
            super();
            this.c = null;
            this.a = null;
            this.b = null;
        }
        f() {
        }
        e(a) {
            if (a == null) {
                this.b = null;
                return;
            }
            this.b = a;
            let b = Math.round(a.rootWrapper.width());
            let c = Math.round(a.rootWrapper.height());
            this.b.rootWrapper.setStyleProperty("position", "relative");
            this.g();
            this.c = this.b.rootWrapper.getChildAt(0);
            this.c.setStyleProperty("margin", "4px 0px 4px 0px");
            this.i();
            this.h();
        }
        g() {
            if (this.b != null) {
                this.b.rootWrapper.setStyleProperty("background-color", this.a.al._fill);
            }
        }
        i() {
            if (this.c != null && this.a.t != null) {
                this.c.setText(this.a.t);
                this.c.setStyleProperty("font", "16px \"Segoe UI\", Verdana");
            }
            this.j();
            this.k();
            this.h();
        }
        j() {
            if (this.c != null && this.a.am != null) {
                this.c.setStyleProperty("color", this.a.am._fill);
            }
        }
        k() {
            if (this.c != null && this.a.j != null) {
                this.c.setStyleProperty("font", this.a.j.fontString);
            }
        }
        h() {
            if (this.b != null) {
                let a = this.b.rootWrapper.getChildAt(1);
                a.setStyleProperty("width", "100%");
                if (this.a.t != null) {
                    a.setStyleProperty("height", "calc(100% - 25px");
                }
                else {
                    a.setStyleProperty("height", "calc(100% - 5px");
                }
            }
        }
        d() {
        }
    }
    ColumnPinningView.$t = /*@__PURE__*/ markType(ColumnPinningView, 'ColumnPinningView');
    return ColumnPinningView;
})();
/**
 * @hidden
 */
export let ColumnPinning = /*@__PURE__*/ (() => {
    class ColumnPinning extends Base {
        constructor() {
            super();
            this.c = null;
            this.h = null;
            this.propertyChanged = null;
            this.k = false;
            this.a = null;
            this.m = null;
            this.i = null;
            this.ak = BrushUtil.g(255, 0, 153, 255);
            this.l = "Filter columns list ...";
            this.d = 1;
            this.f = 2;
            this.aj = BrushUtil.g(255, 248, 248, 248);
            this.k = false;
            let a = new ColumnPinningView();
            a.a = this;
            this.c = a;
            this.c.f();
        }
        ai(a) {
            if (this.h != null) {
                this.h.updateHrMargin(a);
            }
        }
        notifySizeChanged() {
            this.c.d();
        }
        destroy() {
            if (this.h != null) {
                let a = this.h;
                a.checkedChanged = delegateRemove(a.checkedChanged, runOn(this, this.ac));
            }
            this.provideContainer(null, null);
            if (this.a != null) {
                let b = this.a.actualColumns;
                b.collectionChanged = delegateRemove(b.collectionChanged, runOn(this, this.x));
                let c = this.a;
                c.columnPinnedChanged = delegateRemove(c.columnPinnedChanged, runOn(this, this.ah));
            }
        }
        provideContainer(a, b) {
            this.h = b;
            if (this.h != null) {
                let c = this.h;
                c.checkedChanged = delegateCombine(c.checkedChanged, runOn(this, this.ac));
            }
            this.y();
            this.c.e(a);
        }
        get b() {
            return this.a;
        }
        set b(a) {
            let b = this.a;
            this.a = a;
            if (b != this.a) {
                this.ab("TargetGrid", b, this.a);
            }
        }
        get t() {
            return this.m;
        }
        set t(a) {
            let b = this.m;
            this.m = a;
            if (b != this.m) {
                this.ab("Title", b, this.m);
            }
        }
        get j() {
            return this.i;
        }
        set j(a) {
            let b = this.i;
            this.i = a;
            if (b != this.i) {
                this.ab("TitleStyle", b, this.i);
            }
        }
        get am() {
            return this.ak;
        }
        set am(a) {
            let b = this.ak;
            this.ak = a;
            if (b != this.ak) {
                this.ab("TitleColor", b, this.ak);
            }
        }
        get q() {
            return this.l;
        }
        set q(a) {
            let b = this.l;
            this.l = a;
            if (b != this.l) {
                this.ab("FilterPlaceholderText", b, this.l);
            }
        }
        get e() {
            return this.d;
        }
        set e(a) {
            let b = this.d;
            this.d = a;
            if (b != this.d) {
                this.ab("BaseTheme", enumGetBox(BaseControlTheme_$type, b), enumGetBox(BaseControlTheme_$type, this.d));
            }
        }
        get g() {
            return this.f;
        }
        set g(a) {
            let b = this.f;
            this.f = a;
            if (b != this.f) {
                this.ab("Density", enumGetBox(ControlDisplayDensity_$type, b), enumGetBox(ControlDisplayDensity_$type, this.f));
            }
        }
        get al() {
            return this.aj;
        }
        set al(a) {
            let b = this.aj;
            this.aj = a;
            if (b != this.aj) {
                this.ab("BackgroundColor", b, this.aj);
            }
        }
        ab(a, b, c) {
            if (this.propertyChanged != null) {
                this.propertyChanged(this, new PropertyChangedEventArgs(a));
            }
            this.ad(a, b, c);
        }
        ad(a, b, c) {
            switch (a) {
                case "TargetGrid":
                    if (b != null) {
                        let d = b.actualColumns;
                        d.collectionChanged = delegateRemove(d.collectionChanged, runOn(this, this.x));
                        let e = b;
                        e.columnPinnedChanged = delegateRemove(e.columnPinnedChanged, runOn(this, this.ah));
                    }
                    if (this.a != null) {
                        let f = this.a.actualColumns;
                        f.collectionChanged = delegateCombine(f.collectionChanged, runOn(this, this.x));
                        let g = this.a;
                        g.columnPinnedChanged = delegateCombine(g.columnPinnedChanged, runOn(this, this.ah));
                    }
                    this.y();
                    break;
                case "Title":
                    this.c.i();
                    break;
                case "TitleStyle":
                    this.c.k();
                    break;
                case "TitleColor":
                    this.c.j();
                    break;
                case "FilterPlaceholderText":
                    if (this.h != null) {
                        this.h.ax = this.q;
                    }
                    break;
                case "BaseTheme":
                    this.ag();
                    break;
                case "Density":
                    this.af();
                    break;
                case "BackgroundColor":
                    if (this.h != null) {
                        this.h.cg = this.al;
                    }
                    this.c.g();
                    break;
            }
        }
        ag() {
            if (this.h != null) {
                this.h.l = this.e;
            }
        }
        af() {
            if (this.h != null) {
                this.h.n = this.g;
            }
        }
        ah(a, b) {
            if (this.k == false) {
                this.y();
            }
            this.k = false;
        }
        x(a, b) {
            this.y();
        }
        y() {
            if (this.a != null && this.h != null) {
                this.h.l = this.e;
                this.h.n = this.g;
                this.h.cg = this.al;
                let a;
                let b = new ObservableCollection$1(ColumnPinningInfo.$, 0);
                for (let c of fromEnum(this.a.actualColumns)) {
                    a = new ColumnPinningInfo();
                    let d = c.hy;
                    if (c.h2 != null) {
                        d = c.h2;
                    }
                    a.columnName = d;
                    if (c.fk == 1) {
                        a.isPinned = true;
                    }
                    else {
                        a.isPinned = false;
                    }
                    b.add(a);
                }
                this.h.ag = false;
                if (this.q != null) {
                    this.h.ax = this.q;
                }
                this.h.itemsSource = b;
                let e = "";
                let f = "";
                e = "columnName";
                f = "isPinned";
                this.h.a9 = f;
                this.h.at = e;
            }
        }
        ac(a, b) {
            this.k = true;
            let c = this.h.h.actualDataSource.getItemAtIndex(b.b);
            let d = this.a.actualColumns.count;
            for (let e = 0; e < d; e++) {
                if (this.a.actualColumns._inner[e].hy == c.columnName || this.a.actualColumns._inner[e].id == c.columnName || this.a.actualColumns._inner[e].h2 == c.columnName) {
                    if (this.a.actualColumns._inner[e].fk == 0) {
                        this.a.actualColumns._inner[e].fk = 1;
                    }
                    else if (this.a.actualColumns._inner[e].fk == 2) {
                        this.a.actualColumns._inner[e].fk = 1;
                    }
                    else {
                        this.a.actualColumns._inner[e].fk = 0;
                    }
                    break;
                }
            }
        }
    }
    ColumnPinning.$t = /*@__PURE__*/ markType(ColumnPinning, 'ColumnPinning', Base.$, [INotifyPropertyChanged_$type]);
    return ColumnPinning;
})();
