// igniteui-react-grids/esm2015/lib/ColumnSummaryDescriptionCollection.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { SyncableObservableCollection$2 } from "igniteui-react-core";
import { ColumnSummaryDescription } from "./ColumnSummaryDescription";
import { SummaryDescription } from "igniteui-react-core";
/**
 * @hidden
 */
export let ColumnSummaryDescriptionCollection = /*@__PURE__*/ (() => {
    class ColumnSummaryDescriptionCollection extends Base {
        constructor() {
            super();
            this.f = new SyncableObservableCollection$2(ColumnSummaryDescription.$, SummaryDescription.$, 0);
            this.a = null;
            let a = this.f;
            a.compare = (b, c) => b.af == c.propertyName && b.n == c.operand && b.j == c.calculator && b.l == c;
            a.createFrom = (b) => {
                let c = new ColumnSummaryDescription(1, b.propertyName);
                c.n = b.operand;
                c.j = b.calculator;
                c.l = b;
                return c;
            };
            a.createTo = (b) => {
                let c = new SummaryDescription(1, b.af);
                c.operand = b.n;
                c.calculator = b.j;
                b.l = c;
                return c;
            };
        }
        get b() {
            return this.a;
        }
        set b(a) {
            if (a == null) {
                this.f.syncTarget = null;
                this.a = null;
                return;
            }
            this.a = a;
            this.f.syncTarget = this.a.g();
        }
        g(a) {
            return this.f.add(a);
        }
        n(a, b) {
            this.f.add1(a, b);
        }
        m() {
            this.f.clear();
        }
        get(a) {
            return this.f.get(a);
        }
        j(a) {
            return this.f.indexOf(a);
        }
        h(a) {
            return this.f.remove1(a);
        }
        d(a) {
            return this.f.remove(a);
        }
        set(a, b) {
            return this.f.set(a, b);
        }
        size() {
            return this.f.size();
        }
        l(a) {
            this.f.addListener(a);
        }
        o(a) {
            this.f.removeListener(a);
        }
        get i() {
            return this.f.all;
        }
    }
    ColumnSummaryDescriptionCollection.$t = /*@__PURE__*/ markType(ColumnSummaryDescriptionCollection, 'ColumnSummaryDescriptionCollection');
    return ColumnSummaryDescriptionCollection;
})();
