// igniteui-react-grids/esm2015/lib/DataGridScrollerKeyboardListener.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { IScrollerKeyboardListener_$type } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataGridScrollerKeyboardListener = /*@__PURE__*/ (() => {
    class DataGridScrollerKeyboardListener extends Base {
        constructor(a) {
            super();
            this.a = null;
            this.a = a;
        }
        onDownArrow(a, b) {
            return this.a.h8(a, b);
        }
        onEnd(a, b) {
            return this.a.h9(a, b);
        }
        onHome(a, b) {
            return this.a.id(a, b);
        }
        onLeftArrow(a, b) {
            return this.a.ie(a, b);
        }
        onPageDown(a, b) {
            return this.a.ig(a, b);
        }
        onPageUp(a, b) {
            return this.a.ih(a, b);
        }
        onRightArrow(a, b) {
            return this.a.ii(a, b);
        }
        onTab(a, b) {
            return this.a.ij(a, b);
        }
        onUpArrow(a, b) {
            return this.a.ik(a, b);
        }
        onWheel() {
            this.a.sc();
        }
        onEnter(a, b) {
            return this.a.ia(a, b);
        }
        onEscape() {
            return this.a.ib();
        }
        onF2() {
            return this.a.ic();
        }
    }
    DataGridScrollerKeyboardListener.$t = /*@__PURE__*/ markType(DataGridScrollerKeyboardListener, 'DataGridScrollerKeyboardListener', Base.$, [IScrollerKeyboardListener_$type]);
    return DataGridScrollerKeyboardListener;
})();
