// igniteui-react-grids/esm2015/lib/DateTimeCellModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellModelExport } from "./CellModelExport";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DateTimeCellModelExport = /*@__PURE__*/ (() => {
    class DateTimeCellModelExport extends CellModelExport {
        constructor() {
            super(...arguments);
            this.af = new Date();
            this.ag = 0;
            this.ae = false;
        }
        get_ac() {
            return "DateTimeCell";
        }
        ad(a) {
            super.ad(a);
            a.l("dateTimeFormat : " + this.ag + ", ");
        }
    }
    DateTimeCellModelExport.$t = /*@__PURE__*/ markType(DateTimeCellModelExport, 'DateTimeCellModelExport', CellModelExport.$);
    return DateTimeCellModelExport;
})();
