// igniteui-react-grids/esm2015/lib/DragSelectEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DragSelectEventArgs = /*@__PURE__*/ (() => {
    class DragSelectEventArgs extends EventArgs {
        constructor() {
            super(...arguments);
            this._startX = 0;
            this._startY = 0;
            this._currentX = 0;
            this._currentY = 0;
            this._controlPressed = false;
        }
        get startX() {
            return this._startX;
        }
        set startX(a) {
            this._startX = a;
        }
        get startY() {
            return this._startY;
        }
        set startY(a) {
            this._startY = a;
        }
        get currentX() {
            return this._currentX;
        }
        set currentX(a) {
            this._currentX = a;
        }
        get currentY() {
            return this._currentY;
        }
        set currentY(a) {
            this._currentY = a;
        }
        get controlPressed() {
            return this._controlPressed;
        }
        set controlPressed(a) {
            this._controlPressed = a;
        }
    }
    DragSelectEventArgs.$t = /*@__PURE__*/ markType(DragSelectEventArgs, 'DragSelectEventArgs', EventArgs.$);
    return DragSelectEventArgs;
})();
