// igniteui-react-grids/esm2015/lib/EditorCellModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellModelExport } from "./CellModelExport";
import { DataSourceSchemaPropertyType_$type } from "igniteui-react-core";
import { enumGetBox, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let EditorCellModelExport = /*@__PURE__*/ (() => {
    class EditorCellModelExport extends CellModelExport {
        constructor() {
            super(...arguments);
            this.af = false;
            this.ag = null;
            this.ae = 0;
            this.ah = null;
        }
        get_ac() {
            return "EditorCell";
        }
        ad(a) {
            super.ad(a);
            a.l("isHidden: " + this.af.toString().toLowerCase() + ", ");
            a.l("editValue: '" + this.ag + "', ");
            a.l("dataType: '" + enumGetBox(DataSourceSchemaPropertyType_$type, this.ae) + "', ");
            a.l("errorMessage: '" + this.ah + "', ");
        }
    }
    EditorCellModelExport.$t = /*@__PURE__*/ markType(EditorCellModelExport, 'EditorCellModelExport', CellModelExport.$);
    return EditorCellModelExport;
})();
