// igniteui-react-grids/esm2015/lib/FormatSummaryTextEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FormatSummaryTextEventArgs = /*@__PURE__*/ (() => {
    class FormatSummaryTextEventArgs extends Base {
        constructor() {
            super(...arguments);
            this.c = null;
            this.b = null;
            this.a = null;
            this.d = null;
            this.e = null;
        }
    }
    FormatSummaryTextEventArgs.$t = /*@__PURE__*/ markType(FormatSummaryTextEventArgs, 'FormatSummaryTextEventArgs');
    return FormatSummaryTextEventArgs;
})();
