// igniteui-react-grids/esm2015/lib/GridLayerExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { CellModelExport } from "./CellModelExport";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let GridLayerExport = /*@__PURE__*/ (() => {
    class GridLayerExport extends Base {
        constructor(a) {
            super();
            this.a = null;
            this.c = null;
            this.c = a;
            this.a = new List$1(CellModelExport.$, 0);
        }
        get b() {
            return this.a;
        }
        e(a) {
            this.a.add(a);
        }
        f() {
            this.a.clear();
        }
        d() {
            let a = new StringBuilder(0);
            a.u("{");
            a.u("name: \"" + this.c + "\", cells: [");
            for (let b = 0; b < this.a.count; b++) {
                a.l(this.a._inner[b].ab());
                if (b < this.a.count - 1) {
                    a.l(", ");
                }
            }
            a.l("]");
            a.l("}");
            return a.toString();
        }
    }
    GridLayerExport.$t = /*@__PURE__*/ markType(GridLayerExport, 'GridLayerExport');
    return GridLayerExport;
})();
