// igniteui-react-grids/esm2015/lib/GridColumnSpacer.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let GridColumnSpacer = /*@__PURE__*/ (() => {
    class GridColumnSpacer extends Base {
        constructor() {
            super();
            this.g = 0;
            this.h = 0;
            this.m = 0;
            this.c = false;
            this.l = 0;
            this.b = false;
            this.a = false;
            this.m = GridColumnSpacer.k;
            GridColumnSpacer.k++;
        }
        get i() {
            return this.g;
        }
        set i(a) {
            this.g = a;
        }
        get j() {
            return this.h;
        }
        set j(a) {
            this.h = a;
        }
        get o() {
            return this.m;
        }
        get f() {
            return this.c;
        }
        set f(a) {
            this.c = a;
        }
        get n() {
            return this.l;
        }
        set n(a) {
            this.l = a;
        }
        get e() {
            return this.b;
        }
        set e(a) {
            this.b = a;
        }
        get d() {
            return this.a;
        }
        set d(a) {
            this.a = a;
        }
    }
    GridColumnSpacer.$t = /*@__PURE__*/ markType(GridColumnSpacer, 'GridColumnSpacer');
    GridColumnSpacer.k = 0;
    return GridColumnSpacer;
})();
