// igniteui-react-grids/esm2015/lib/GridColumnOptionsSimpleSectionBase.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSectionBase } from "./GridColumnOptionsSectionBase";
import { StringComparer } from "igniteui-react-core";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let GridColumnOptionsSimpleSectionBase = /*@__PURE__*/ (() => {
    class GridColumnOptionsSimpleSectionBase extends GridColumnOptionsSectionBase {
        constructor() {
            super();
            this.br = null;
            this.bp = true;
        }
        get bs() {
            if (null == this.br) {
                return this.bu;
            }
            return this.br;
        }
        set bs(a) {
            let b = this.bs;
            if (0 == StringComparer.b.compare(a, b)) {
                return;
            }
            this.br = a;
            this.a8("Caption", b, this.br);
        }
        get bq() {
            return this.bp;
        }
        set bq(a) {
            let b = this.bq;
            if (a == b) {
                return;
            }
            this.bp = a;
            this.a8("ShowCaption", b, this.bp);
        }
        ba(a, b, c) {
            super.ba(a, b, c);
            switch (a) {
                case "ShowCaption":
                case "Caption":
                    this.e.m();
                    break;
            }
        }
    }
    GridColumnOptionsSimpleSectionBase.$t = /*@__PURE__*/ markType(GridColumnOptionsSimpleSectionBase, 'GridColumnOptionsSimpleSectionBase', GridColumnOptionsSectionBase.$);
    return GridColumnOptionsSimpleSectionBase;
})();
