// igniteui-react-grids/esm2015/lib/GridStyleRepository.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
import { FontInfo } from "igniteui-react-core";
import { DeviceUtils } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { CornerRadius } from "igniteui-react-core";
import { truncate } from "igniteui-react-core";
/**
 * @hidden
 */
export let GridStyleRepository = /*@__PURE__*/ (() => {
    class GridStyleRepository extends Base {
        static get bi() {
            if (GridStyleRepository.ap == null) {
                let a = "#babfc7";
                GridStyleRepository.ap = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ap;
        }
        static get bk() {
            if (GridStyleRepository.ar == null) {
                let a = "#dde2eb";
                GridStyleRepository.ar = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ar;
        }
        static get bc() {
            if (GridStyleRepository.aj == null) {
                GridStyleRepository.aj = ((() => {
                    let $ret = new Brush();
                    $ret.color = Color.u(255, 224, 224, 224);
                    return $ret;
                })());
            }
            return GridStyleRepository.aj;
        }
        static get a2() {
            if (GridStyleRepository.z == null) {
                let a = "#babfc7";
                GridStyleRepository.z = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.z;
        }
        static get a0() {
            if (GridStyleRepository.x == null) {
                let a = "#babfc7";
                GridStyleRepository.x = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.x;
        }
        static get bm() {
            if (GridStyleRepository.at == null) {
                let a = "#ff80ab";
                GridStyleRepository.at = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.at;
        }
        static get bn() {
            if (GridStyleRepository.au == null) {
                let a = "#181d1f";
                GridStyleRepository.au = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.au;
        }
        static get av() {
            if (GridStyleRepository.s == null) {
                let a = "#2196f3";
                GridStyleRepository.s = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.s;
        }
        static get l() {
            if (GridStyleRepository.g == null) {
                GridStyleRepository.g = new FontInfo();
                GridStyleRepository.g.d = DeviceUtils.f(13);
                GridStyleRepository.g.n = "'Segoe UI', Verdana";
            }
            return GridStyleRepository.g;
        }
        static get k() {
            if (GridStyleRepository.f == null) {
                GridStyleRepository.f = new FontInfo();
                GridStyleRepository.f.d = DeviceUtils.f(13);
                GridStyleRepository.f.q = "italic";
                GridStyleRepository.f.n = "'Segoe UI', Verdana";
            }
            return GridStyleRepository.f;
        }
        static get h() {
            if (GridStyleRepository.c == null) {
                GridStyleRepository.c = new FontInfo();
                GridStyleRepository.c.d = DeviceUtils.f(14);
                GridStyleRepository.c.n = "'Segoe UI', Verdana";
            }
            return GridStyleRepository.c;
        }
        static get j() {
            if (GridStyleRepository.d == null) {
                GridStyleRepository.d = new FontInfo();
                GridStyleRepository.d.d = DeviceUtils.f(12);
                GridStyleRepository.d.s = "600";
                GridStyleRepository.d.n = "'Segoe UI', Verdana";
            }
            return GridStyleRepository.d;
        }
        static get i() {
            if (GridStyleRepository.e == null) {
                GridStyleRepository.e = new FontInfo();
                GridStyleRepository.e.d = DeviceUtils.f(12);
                GridStyleRepository.e.n = "'Segoe UI', Verdana";
            }
            return GridStyleRepository.e;
        }
        static get ba() {
            if (GridStyleRepository.ah == null) {
                GridStyleRepository.ah = new Brush();
                GridStyleRepository.ah.color = Color.u(255, 210, 234, 253);
            }
            return GridStyleRepository.ah;
        }
        static get aw() {
            if (GridStyleRepository.t == null) {
                GridStyleRepository.t = ((() => {
                    let $ret = new Brush();
                    $ret.fill = "White";
                    return $ret;
                })());
            }
            return GridStyleRepository.t;
        }
        static get a1() {
            if (GridStyleRepository.y == null) {
                let a = "#f3f9fe";
                GridStyleRepository.y = ((() => {
                    let $ret = new Brush();
                    $ret.color = Color.u(255, 232, 244, 254);
                    return $ret;
                })());
            }
            return GridStyleRepository.y;
        }
        static get bh() {
            if (GridStyleRepository.ao == null) {
                let a = "#F8F8F8";
                GridStyleRepository.ao = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ao;
        }
        static get bl() {
            if (GridStyleRepository.as == null) {
                let a = "#181d1f";
                GridStyleRepository.as = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.as;
        }
        static get bj() {
            if (GridStyleRepository.aq == null) {
                let a = "#181d1f";
                GridStyleRepository.aq = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.aq;
        }
        static get a8() {
            if (GridStyleRepository.af == null) {
                let a = "#fafafa";
                GridStyleRepository.af = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.af;
        }
        static get a7() {
            if (GridStyleRepository.ae == null) {
                GridStyleRepository.ae = ((() => {
                    let $ret = new Brush();
                    $ret.fill = "#f5f5dc";
                    return $ret;
                })());
            }
            return GridStyleRepository.ae;
        }
        static get bb() {
            if (GridStyleRepository.ai == null) {
                let a = "#ffffff";
                GridStyleRepository.ai = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ai;
        }
        static get a4() {
            if (GridStyleRepository.ab == null) {
                let a = "#ffffff";
                GridStyleRepository.ab = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ab;
        }
        static get a6() {
            if (GridStyleRepository.ad == null) {
                let a = "#f8f8f8";
                GridStyleRepository.ad = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ad;
        }
        static get a9() {
            if (GridStyleRepository.ag == null) {
                let a = "#fafafa";
                GridStyleRepository.ag = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ag;
        }
        static get bd() {
            if (GridStyleRepository.ak == null) {
                let a = "#0099FF";
                GridStyleRepository.ak = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ak;
        }
        static get be() {
            if (GridStyleRepository.al == null) {
                let a = "#181d1f";
                GridStyleRepository.al = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.al;
        }
        static get ay() {
            if (GridStyleRepository.v == null) {
                let a = "#2196f3";
                GridStyleRepository.v = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.v;
        }
        static get a5() {
            if (GridStyleRepository.ac == null) {
                let a = "#2196f3";
                GridStyleRepository.ac = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.ac;
        }
        static get a3() {
            if (GridStyleRepository.aa == null) {
                let a = "#babfc7";
                GridStyleRepository.aa = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.aa;
        }
        static get az() {
            if (GridStyleRepository.w == null) {
                let a = "#181d1f";
                GridStyleRepository.w = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.w;
        }
        static get a() {
            return 2;
        }
        static get b() {
            return 3;
        }
        static get o() {
            return truncate(DeviceUtils.g(30));
        }
        static get p() {
            return truncate(DeviceUtils.g(36));
        }
        static get n() {
            return truncate(DeviceUtils.g(40));
        }
        static get ax() {
            if (GridStyleRepository.u == null) {
                let a = "#babfc7";
                GridStyleRepository.u = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.u;
        }
        static get bp() {
            return GridStyleRepository.bo;
        }
        static get r() {
            return GridStyleRepository.q;
        }
        static get bg() {
            if (GridStyleRepository.an == null) {
                let a = "#ff0000";
                GridStyleRepository.an = ((() => {
                    let $ret = new Brush();
                    $ret.fill = a;
                    return $ret;
                })());
            }
            return GridStyleRepository.an;
        }
        static get bf() {
            if (GridStyleRepository.am == null) {
                GridStyleRepository.am = ((() => {
                    let $ret = new Brush();
                    $ret.color = Color.u(0, 0, 0, 0);
                    return $ret;
                })());
            }
            return GridStyleRepository.am;
        }
    }
    GridStyleRepository.$t = /*@__PURE__*/ markType(GridStyleRepository, 'GridStyleRepository');
    GridStyleRepository.ap = null;
    GridStyleRepository.ar = null;
    GridStyleRepository.aj = null;
    GridStyleRepository.z = null;
    GridStyleRepository.x = null;
    GridStyleRepository.at = null;
    GridStyleRepository.au = null;
    GridStyleRepository.s = null;
    GridStyleRepository.g = null;
    GridStyleRepository.f = null;
    GridStyleRepository.m = 0.54;
    GridStyleRepository.c = null;
    GridStyleRepository.d = null;
    GridStyleRepository.e = null;
    GridStyleRepository.ah = null;
    GridStyleRepository.t = null;
    GridStyleRepository.y = null;
    GridStyleRepository.ao = null;
    GridStyleRepository.as = null;
    GridStyleRepository.aq = null;
    GridStyleRepository.af = null;
    GridStyleRepository.ae = null;
    GridStyleRepository.ai = null;
    GridStyleRepository.ab = null;
    GridStyleRepository.ad = null;
    GridStyleRepository.ag = null;
    GridStyleRepository.ak = null;
    GridStyleRepository.al = null;
    GridStyleRepository.v = null;
    GridStyleRepository.ac = null;
    GridStyleRepository.aa = null;
    GridStyleRepository.w = null;
    GridStyleRepository.u = null;
    GridStyleRepository.bo = /*@__PURE__*/ new Thickness(0, 1);
    GridStyleRepository.q = /*@__PURE__*/ new CornerRadius(0, 3);
    GridStyleRepository.an = null;
    GridStyleRepository.am = null;
    return GridStyleRepository;
})();
