// igniteui-react-grids/esm2015/lib/igr-image-column.js
import { ImageStretchOptions_$type } from "./ImageStretchOptions";
import { EditorType_$type } from "./EditorType";
import { IgrColumn } from "./igr-column";
import { ImageColumn } from "./ImageColumn";
import { ensureEnum } from "igniteui-react-core";
/**
 * Represents an Image grid column, used to allow the developer to display an image in a cell.
*/
export class IgrImageColumn extends IgrColumn {
    createImplementation() {
        return new ImageColumn();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the stretch option to use when rendering images
    */
    get imageStretchOption() {
        return this.i.jt;
    }
    set imageStretchOption(v) {
        this.i.jt = ensureEnum(ImageStretchOptions_$type, v);
    }
    /**
     * Gets or sets the editor type used for editing cells in this column.
    */
    get editorType() {
        return this.i.jp;
    }
    set editorType(v) {
        this.i.jp = ensureEnum(EditorType_$type, v);
    }
    /**
     * Gets or sets the ComboBox data source.
    */
    get editorDataSource() {
        return this.i.ju;
    }
    set editorDataSource(v) {
        this.i.ju = v;
    }
    /**
     * Gets or sets the ComboBox text field.
    */
    get editorTextField() {
        return this.i.jv;
    }
    set editorTextField(v) {
        this.i.jv = v;
    }
    /**
     * Gets or sets the ComboBox value field.
    */
    get editorValueField() {
        return this.i.jo;
    }
    set editorValueField(v) {
        this.i.jo = v;
    }
}
