// igniteui-react-grids/esm2015/lib/igr-grid-column-width-changed-event-args.js
import { IgrColumn } from "./igr-column";
import { IgrColumnWidth } from "./igr-column-width";
import { GridColumnWidthChangedEventArgs as GridColumnWidthChangedEventArgs_internal } from "./GridColumnWidthChangedEventArgs";
/**
 * Event arguments for the GridColumnWidthChanged event.
*/
export class IgrGridColumnWidthChangedEventArgs {
    createImplementation() {
        return new GridColumnWidthChangedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The column whose Width has changed
    */
    get column() {
        if (this.i.a == null) {
            return null;
        }
        if (!this.i.a.externalObject) {
            let e = IgrColumn._createFromInternal(this.i.a);
            if (e) {
                e._implementation = this.i.a;
            }
            this.i.a.externalObject = e;
        }
        return this.i.a.externalObject;
    }
    set column(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    /**
     * The old column width.
    */
    get oldWidth() {
        if (this.i.b == null) {
            return null;
        }
        if (!this.i.b.externalObject) {
            let e = new IgrColumnWidth();
            e._implementation = this.i.b;
            this.i.b.externalObject = e;
        }
        return this.i.b.externalObject;
    }
    set oldWidth(v) {
        v == null ? this.i.b = null : this.i.b = v.i;
    }
}
