// igniteui-react-grids/esm2015/lib/igr-format-group-text-event-args.js
import { FormatGroupTextEventArgs as FormatGroupTextEventArgs_internal } from "./FormatGroupTextEventArgs";
export class IgrFormatGroupTextEventArgs {
    createImplementation() {
        return new FormatGroupTextEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets the group name.
    */
    get groupName() {
        return this.i.d;
    }
    set groupName(v) {
        this.i.d = v;
    }
    /**
     * Gets the original value before formatting.
    */
    get groupValue() {
        return this.i.a;
    }
    set groupValue(v) {
        this.i.a = v;
    }
    /**
     * Gets the group value as text, formatted based on the GroupDescription's ValueFormat.
    */
    get formattedValue() {
        return this.i.c;
    }
    set formattedValue(v) {
        this.i.c = v;
    }
    /**
     * Gets or sets the final text that will be displayed in the group header.
    */
    get formattedText() {
        return this.i.b;
    }
    set formattedText(v) {
        this.i.b = v;
    }
}
