// igniteui-react-grids/esm2015/lib/igr-section-header.js
import { GroupSummaryDisplayMode_$type } from "./GroupSummaryDisplayMode";
import { IgrDefinitionBase } from "./igr-definition-base";
import { SectionHeader } from "./SectionHeader";
import { brushToString, stringToBrush, ensureBool, ensureEnum } from "igniteui-react-core";
/**
 * Used to configure the appearance of the section header cells.
*/
export class IgrSectionHeader extends IgrDefinitionBase {
    constructor(props) {
        super(props);
    }
    createImplementation() {
        return new SectionHeader();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    get selectedBackground() {
        return brushToString(this.i.fe);
    }
    set selectedBackground(v) {
        this.i.fe = stringToBrush(v);
    }
    get actualSelectedBackground() {
        return brushToString(this.i.fd);
    }
    set actualSelectedBackground(v) {
        this.i.fd = stringToBrush(v);
    }
    /**
     * Gets or sets the amount of left padding to use for the cell content for this column.
    */
    get paddingLeft() {
        return this.i.ex;
    }
    set paddingLeft(v) {
        this.i.ex = +v;
    }
    /**
     * Gets or sets the amount of top padding to use for the cell content for this column.
    */
    get paddingTop() {
        return this.i.ez;
    }
    set paddingTop(v) {
        this.i.ez = +v;
    }
    /**
     * Gets or sets the amount of right padding to use for the cell content of this column.
    */
    get paddingRight() {
        return this.i.ey;
    }
    set paddingRight(v) {
        this.i.ey = +v;
    }
    /**
     * Gets or sets the amount of bottom padding to use for the cell content of this column.
    */
    get paddingBottom() {
        return this.i.ew;
    }
    set paddingBottom(v) {
        this.i.ew = +v;
    }
    get isCollapsable() {
        return this.i.eq;
    }
    set isCollapsable(v) {
        this.i.eq = ensureBool(v);
    }
    get summaryDisplayMode() {
        return this.i.en;
    }
    set summaryDisplayMode(v) {
        this.i.en = ensureEnum(GroupSummaryDisplayMode_$type, v);
    }
}
