// igniteui-react-grids/esm2015/lib/igr-text-header-cell-info.js
import { SortIndicatorStyle_$type } from "./SortIndicatorStyle";
import { ColumnOptionsIconAlignment_$type } from "./ColumnOptionsIconAlignment";
import { ColumnOptionsIconBehavior_$type } from "./ColumnOptionsIconBehavior";
import { IgrTextCellInfo } from "./igr-text-cell-info";
import { TextHeaderCellModel as TextHeaderCellModel_internal } from "./TextHeaderCellModel";
import { ensureBool, ensureEnum, brushToString, stringToBrush } from "igniteui-react-core";
/**
 * Backing information for a text header cell.
*/
export class IgrTextHeaderCellInfo extends IgrTextCellInfo {
    createImplementation() {
        return new TextHeaderCellModel_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Sets or gets whether the filter UI should be visible in the header.
    */
    get isColumnOptionsEnabled() {
        return this.i.k1;
    }
    set isColumnOptionsEnabled(v) {
        this.i.k1 = ensureBool(v);
    }
    get sortIndicatorStyle() {
        return this.i.kz;
    }
    set sortIndicatorStyle(v) {
        this.i.kz = ensureEnum(SortIndicatorStyle_$type, v);
    }
    /**
     * Gets or sets the alignment of the column options icon within the header cell.
    */
    get columnOptionsIconAlignment() {
        return this.i.kv;
    }
    set columnOptionsIconAlignment(v) {
        this.i.kv = ensureEnum(ColumnOptionsIconAlignment_$type, v);
    }
    /**
     * Gets or sets the color for the column options icon in the header.
    */
    get columnOptionsIconColor() {
        return brushToString(this.i.ld);
    }
    set columnOptionsIconColor(v) {
        this.i.ld = stringToBrush(v);
    }
    /**
     * Gets or sets how the column option icon will behave in the header.
    */
    get columnOptionsIconBehavior() {
        return this.i.kx;
    }
    set columnOptionsIconBehavior(v) {
        this.i.kx = ensureEnum(ColumnOptionsIconBehavior_$type, v);
    }
}
