// igniteui-react-grids/esm2015/lib/igr-combo-box-column.js
import { IgrColumn } from "./igr-column";
import { ComboBoxColumn } from "./ComboBoxColumn";
/**
 * A combobox column.
*/
export class IgrComboBoxColumn extends IgrColumn {
    createImplementation() {
        return new ComboBoxColumn();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets the actual data source that is being used by the combo. If a collection was provided for the combobox data, an implicit LocalDataSource is created, and this is available via this property.
    */
    get actualDataSource() {
        return this.i.jq;
    }
    set actualDataSource(v) {
        this.i.jq = v;
    }
    /**
     * Gets or sets the ComboBox data source.
    */
    get dataSource() {
        return this.i.jr;
    }
    set dataSource(v) {
        this.i.jr = v;
    }
    /**
     * Gets or sets the ComboBox text field.
    */
    get textField() {
        return this.i.jv;
    }
    set textField(v) {
        this.i.jv = v;
    }
    /**
     * Gets or sets the ComboBox value field.
    */
    get valueField() {
        return this.i.jo;
    }
    set valueField(v) {
        this.i.jo = v;
    }
}
