// igniteui-react-grids/esm2015/lib/igr-image-cell-info.js
import { ImageStretchOptions_$type } from "./ImageStretchOptions";
import { ImageResourceType_$type } from "./ImageResourceType";
import { IgrCellInfo } from "./igr-cell-info";
import { ImageCellModel as ImageCellModel_internal } from "./ImageCellModel";
import { ensureEnum } from "igniteui-react-core";
/**
 * Backing information for an image cell in the grid.
*/
export class IgrImageCellInfo extends IgrCellInfo {
    createImplementation() {
        return new ImageCellModel_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Sets or gets the resource path to use to get the image for the cell.
    */
    get imagePath() {
        return this.i.ky;
    }
    set imagePath(v) {
        this.i.ky = v;
    }
    /**
     * Sets or gets the image stretching behavior for the image.
    */
    get imageStretchOption() {
        return this.i.kt;
    }
    set imageStretchOption(v) {
        this.i.kt = ensureEnum(ImageStretchOptions_$type, v);
    }
    /**
     * Sets or gets the image resource type. Indicates which type of resource should be fetched and how ImagePath should be interpreted.
    */
    get imageResourceType() {
        return this.i.kr;
    }
    set imageResourceType(v) {
        this.i.kr = ensureEnum(ImageResourceType_$type, v);
    }
}
