// igniteui-react-grids/esm2015/lib/igr-grid-filter-dialog-view-model-row.js
import { GridFilterDialogViewModelRow as GridFilterDialogViewModelRow_internal } from "./GridFilterDialogViewModelRow";
export class IgrGridFilterDialogViewModelRow {
    createImplementation() {
        return new GridFilterDialogViewModelRow_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get operatorTypes() {
        return this.i.f;
    }
    set operatorTypes(v) {
        this.i.f = v;
    }
    get operators() {
        return this.i.e;
    }
    set operators(v) {
        this.i.e = v;
    }
    get currentOperator() {
        return this.i.l;
    }
    set currentOperator(v) {
        this.i.l = v;
    }
    get operandNumber() {
        return this.i.d;
    }
    set operandNumber(v) {
        this.i.d = v;
    }
    get operand1() {
        return this.i.i;
    }
    set operand1(v) {
        this.i.i = v;
    }
    get operand2() {
        return this.i.j;
    }
    set operand2(v) {
        this.i.j = v;
    }
    findByName(name) {
        return null;
    }
}
