// igniteui-react-grids/esm2015/lib/igr-text-column.js
import { EditorType_$type } from "./EditorType";
import { IgrColumn } from "./igr-column";
import { TextColumn } from "./TextColumn";
import { ensureEnum } from "igniteui-react-core";
/**
 * Represents a Text grid column, used to allow the developer to display a formatted text in a cell.
*/
export class IgrTextColumn extends IgrColumn {
    createImplementation() {
        return new TextColumn();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the editor type used for editing cells in this column.
    */
    get editorType() {
        return this.i.jp;
    }
    set editorType(v) {
        this.i.jp = ensureEnum(EditorType_$type, v);
    }
    /**
     * Gets or sets the ComboBox data source.
    */
    get editorDataSource() {
        return this.i.jq;
    }
    set editorDataSource(v) {
        this.i.jq = v;
    }
    /**
     * Gets or sets the ComboBox text field.
    */
    get editorTextField() {
        return this.i.js;
    }
    set editorTextField(v) {
        this.i.js = v;
    }
    /**
     * Gets or sets the ComboBox value field.
    */
    get editorValueField() {
        return this.i.jo;
    }
    set editorValueField(v) {
        this.i.jo = v;
    }
}
