// igniteui-react-grids/esm2015/lib/igr-column-pinned-changed-event-args.js
import { IgrColumn } from "./igr-column";
import { PinnedPositions_$type } from "./PinnedPositions";
import { ColumnPinnedChangedEventArgs as ColumnPinnedChangedEventArgs_internal } from "./ColumnPinnedChangedEventArgs";
import { ensureEnum } from "igniteui-react-core";
/**
 * Event arguments for the ColumnPinnedChanged event.
*/
export class IgrColumnPinnedChangedEventArgs {
    createImplementation() {
        return new ColumnPinnedChangedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The column whose pinned state has changed.
    */
    get column() {
        if (this.i.a == null) {
            return null;
        }
        if (!this.i.a.externalObject) {
            let e = IgrColumn._createFromInternal(this.i.a);
            if (e) {
                e._implementation = this.i.a;
            }
            this.i.a.externalObject = e;
        }
        return this.i.a.externalObject;
    }
    set column(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    /**
     * The new pinned state for the column.
    */
    get pinned() {
        return this.i.b;
    }
    set pinned(v) {
        this.i.b = ensureEnum(PinnedPositions_$type, v);
    }
}
