// igniteui-react-grids/esm2015/lib/igr-grid-size-changed-event-args.js
import { GridSizeChangedEventArgs as GridSizeChangedEventArgs_internal } from "./GridSizeChangedEventArgs";
/**
 * Event arguments for the SizeChanged event.
*/
export class IgrGridSizeChangedEventArgs {
    createImplementation() {
        return new GridSizeChangedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The new width.
    */
    get width() {
        return this.i.b;
    }
    set width(v) {
        this.i.b = +v;
    }
    /**
     * The new height.
    */
    get height() {
        return this.i.a;
    }
    set height(v) {
        this.i.a = +v;
    }
}
