// igniteui-react-grids/esm2015/lib/igr-column-width.js
import { ColumnWidth as ColumnWidth_internal } from "./ColumnWidth";
import { ensureBool } from "igniteui-react-core";
/**
 * Represents a width of a column.
*/
export class IgrColumnWidth {
    createImplementation() {
        return new ColumnWidth_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets or sets whether to use star sizing for the column's width
    */
    get isStarSized() {
        return this.i.c;
    }
    set isStarSized(v) {
        this.i.c = ensureBool(v);
    }
    /**
     * Gets or sets the fixed width or the proportional width of the column, depending on whether this width is a star width.
    */
    get value() {
        return this.i.f;
    }
    set value(v) {
        this.i.f = +v;
    }
    /**
     * Gets or sets a minimum width to use when star sizing.
    */
    get minimumWidth() {
        return this.i.e;
    }
    set minimumWidth(v) {
        this.i.e = +v;
    }
    findByName(name) {
        return null;
    }
}
