// igniteui-react-grids/esm2015/lib/igr-grid-row-edit-ended-event-args.js
import { GridRowEditEndedEventArgs as GridRowEditEndedEventArgs_internal } from "./GridRowEditEndedEventArgs";
export class IgrGridRowEditEndedEventArgs {
    createImplementation() {
        return new GridRowEditEndedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets the row that ended editing.
    */
    get row() {
        return this.i.a;
    }
    set row(v) {
        this.i.a = +v;
    }
    /**
     * Gets the row item.
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
}
