// igniteui-react-grids/esm2015/lib/igr-summary-row.js
import { IgrDefinitionBase } from "./igr-definition-base";
import { brushToString, stringToBrush } from "igniteui-react-core";
import { FontInfo } from "igniteui-react-core";
export class IgrSummaryRow extends IgrDefinitionBase {
    constructor(props) {
        super(props);
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    /**
     * Gets or sets the background color for summary cells when selected.
    */
    get selectedBackground() {
        return brushToString(this.i.fi);
    }
    set selectedBackground(v) {
        this.i.fi = stringToBrush(v);
    }
    /**
     * Gets the actual background color for summary cells when selected.
    */
    get actualSelectedBackground() {
        return brushToString(this.i.ff);
    }
    set actualSelectedBackground(v) {
        this.i.ff = stringToBrush(v);
    }
    /**
     * Gets or sets the text color for the summary labels.
    */
    get summaryLabelTextColor() {
        return brushToString(this.i.fj);
    }
    set summaryLabelTextColor(v) {
        this.i.fj = stringToBrush(v);
    }
    /**
     * Gets the actual text color for the summary labels.
    */
    get actualSummaryLabelTextColor() {
        return brushToString(this.i.fg);
    }
    set actualSummaryLabelTextColor(v) {
        this.i.fg = stringToBrush(v);
    }
    /**
     * Gets or sets the text style for the summary labels.
    */
    get summaryNameTextStyle() {
        if (this.i.summaryNameTextStyle == null) {
            return null;
        }
        return this.i.summaryNameTextStyle.fontString;
    }
    set summaryNameTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.summaryNameTextStyle = fi;
    }
    /**
     * Gets or sets the amount of left padding to use for the cell content for this column.
    */
    get paddingLeft() {
        return this.i.ev;
    }
    set paddingLeft(v) {
        this.i.ev = +v;
    }
    /**
     * Gets or sets the amount of top padding to use for the cell content for this column.
    */
    get paddingTop() {
        return this.i.ex;
    }
    set paddingTop(v) {
        this.i.ex = +v;
    }
    /**
     * Gets or sets the amount of right padding to use for the cell content of this column.
    */
    get paddingRight() {
        return this.i.ew;
    }
    set paddingRight(v) {
        this.i.ew = +v;
    }
    /**
     * Gets or sets the amount of bottom padding to use for the cell content of this column.
    */
    get paddingBottom() {
        return this.i.eu;
    }
    set paddingBottom(v) {
        this.i.eu = +v;
    }
}
