// igniteui-react-grids/esm2015/lib/igr-grid-filter-dialog-view-model-grouping.js
import { GridFilterDialogViewModelGrouping as GridFilterDialogViewModelGrouping_internal } from "./GridFilterDialogViewModelGrouping";
import { ensureBool } from "igniteui-react-core";
export class IgrGridFilterDialogViewModelGrouping {
    createImplementation() {
        return new GridFilterDialogViewModelGrouping_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get startIndex() {
        return this.i.c;
    }
    set startIndex(v) {
        this.i.c = +v;
    }
    get endIndex() {
        return this.i.b;
    }
    set endIndex(v) {
        this.i.b = +v;
    }
    get isOrGrouping() {
        return this.i.a;
    }
    set isOrGrouping(v) {
        this.i.a = ensureBool(v);
    }
    findByName(name) {
        return null;
    }
}
