// igniteui-react-grids/esm2015/lib/igr-cell-style-requested-event-args.js
import { CellStyleRequestedEventArgs as CellStyleRequestedEventArgs_internal } from "./CellStyleRequestedEventArgs";
/**
 * Information about the style request for a cell.
*/
export class IgrCellStyleRequestedEventArgs {
    createImplementation() {
        return new CellStyleRequestedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The proposed style key for the cell. Used for recycling the cell.
    */
    get styleKey() {
        return this.i.styleKey;
    }
    set styleKey(v) {
        this.i.styleKey = v;
    }
    /**
     * The resolved value for the cell.
    */
    get resolvedValue() {
        return this.i.resolvedValue;
    }
    set resolvedValue(v) {
        this.i.resolvedValue = v;
    }
    /**
     * The row number for the cell.
    */
    get rowNumber() {
        return this.i.rowNumber;
    }
    set rowNumber(v) {
        this.i.rowNumber = +v;
    }
}
