// igniteui-react-grids/esm2015/lib/igr-date-time-cell-info.js
import { DateTimeFormats_$type } from "./DateTimeFormats";
import { IgrCellInfo } from "./igr-cell-info";
import { DateTimeCellModel as DateTimeCellModel_internal } from "./DateTimeCellModel";
import { ensureEnum, ensureBool, arrayFindByName } from "igniteui-react-core";
/**
 * Backing information about a date time cell for the grid.
*/
export class IgrDateTimeCellInfo extends IgrCellInfo {
    createImplementation() {
        return new DateTimeCellModel_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Sets or gets the DateTime value for the cell.
    */
    get dateTimeValue() {
        return this.i.k0;
    }
    set dateTimeValue(v) {
        this.i.k0 = v;
    }
    /**
     * The format string to apply to the value
    */
    get formatStringOverride() {
        return this.i.lh;
    }
    set formatStringOverride(v) {
        this.i.lh = v;
    }
    get formatSpecifiers() {
        return this.i.kr;
    }
    set formatSpecifiers(v) {
        this.i.kr = v;
    }
    /**
     * The format options to apply to the value
    */
    get formatOverride() {
        return this.i.la;
    }
    set formatOverride(v) {
        this.i.la = v;
    }
    /**
     * Sets or gets the simple DateTimeFormat to use for the cell.
    */
    get dateTimeFormat() {
        return this.i.kw;
    }
    set dateTimeFormat(v) {
        this.i.kw = ensureEnum(DateTimeFormats_$type, v);
    }
    /**
     * Sets or gets whether the value to use is a DateTimeOffset value.
    */
    get isOffsetValue() {
        return this.i.ky;
    }
    set isOffsetValue(v) {
        this.i.ky = ensureBool(v);
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.formatSpecifiers != null && arrayFindByName(this.formatSpecifiers, name)) {
            return arrayFindByName(this.formatSpecifiers, name);
        }
        return null;
    }
}
