// igniteui-react-grids/esm2015/lib/igr-header.js
import { SortIndicatorStyle_$type } from "./SortIndicatorStyle";
import { IgrDefinitionBase } from "./igr-definition-base";
import { brushToString, stringToBrush, ensureEnum } from "igniteui-react-core";
/**
 * Represents a base class used to configure the appearance of the column header cells.
*/
export class IgrHeader extends IgrDefinitionBase {
    constructor(props) {
        super(props);
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    /**
     * Gets or sets the color to use for the sort indicator icon
    */
    get sortIndicatorColor() {
        return brushToString(this.i.fb);
    }
    set sortIndicatorColor(v) {
        this.i.fb = stringToBrush(v);
    }
    /**
     * Gets the actual background color that is used for the cells when they are selected
    */
    get actualSortIndicatorColor() {
        return brushToString(this.i.e9);
    }
    set actualSortIndicatorColor(v) {
        this.i.e9 = stringToBrush(v);
    }
    get sortIndicatorStyle() {
        return this.i.eo;
    }
    set sortIndicatorStyle(v) {
        this.i.eo = ensureEnum(SortIndicatorStyle_$type, v);
    }
    get actualSortIndicatorStyle() {
        return this.i.en;
    }
    set actualSortIndicatorStyle(v) {
        this.i.en = ensureEnum(SortIndicatorStyle_$type, v);
    }
    /**
     * Gets or sets the amount of left padding to use for the cell content.
    */
    get paddingLeft() {
        return this.i.eu;
    }
    set paddingLeft(v) {
        this.i.eu = +v;
    }
    /**
     * Gets or sets the amount of top padding to use for the cell content.
    */
    get paddingTop() {
        return this.i.ew;
    }
    set paddingTop(v) {
        this.i.ew = +v;
    }
    /**
     * Gets or sets the amount of right padding to use for the cell content.
    */
    get paddingRight() {
        return this.i.ev;
    }
    set paddingRight(v) {
        this.i.ev = +v;
    }
    /**
     * Gets or sets the amount of bottom padding to use for the cell content.
    */
    get paddingBottom() {
        return this.i.et;
    }
    set paddingBottom(v) {
        this.i.et = +v;
    }
}
