// igniteui-react-grids/esm2015/lib/igr-group-data.js
import { GroupData as GroupData_internal } from "./GroupData";
/**
 * Object used to store group information for group headers.
*/
export class IgrGroupData {
    createImplementation() {
        return new GroupData_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The grouping key.
    */
    get groupName() {
        return this.i.c;
    }
    set groupName(v) {
        this.i.c = v;
    }
    /**
     * The group value.
    */
    get groupValue() {
        return this.i.a;
    }
    set groupValue(v) {
        this.i.a = v;
    }
    /**
     * The value as formatted text.
    */
    get formattedText() {
        return this.i.b;
    }
    set formattedText(v) {
        this.i.b = v;
    }
    findByName(name) {
        return null;
    }
}
