// igniteui-react-grids/esm2015/lib/igr-data-grid-sort-indicator.js
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { ColumnSortDirection_$type } from "./ColumnSortDirection";
import { SortIndicatorStyle_$type } from "./SortIndicatorStyle";
import { IgrSortIndicatorRenderCompletedEventArgs } from "./igr-sort-indicator-render-completed-event-args";
import { GridSortIndicator } from "./GridSortIndicator";
import { TypeRegistrar } from "igniteui-react-core";
import { NamePatcher, getModifiedProps, isValidProp, ensureEnum, brushToString, stringToBrush, ensureBool, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
/**
 * An advanced grid for displaying data.
*/
export class IgrDataGridSortIndicator extends React.Component {
    constructor(props) {
        super(props);
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._renderCompleted = null;
        this._renderCompleted_wrapped = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    createImplementation() {
        return new GridSortIndicator();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    } /**
                                 * @hidden
                                 */
    static _createFromInternal(internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        let name = internal.$type.name;
        let externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    }
    onImplementationCreated() {
    }
    componentDidMount() {
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                this[p] = this.props[p];
            }
        }
    }
    shouldComponentUpdate(nextProps, nextState) {
        const mod = getModifiedProps(this.props, nextProps);
        for (const p of Object.keys(mod)) {
            if (isValidProp(this, p)) {
                this[p] = mod[p];
            }
        }
        return true;
    }
    render() {
        return null;
    }
    get sortDirection() {
        return this.i.b;
    }
    set sortDirection(v) {
        this.i.b = ensureEnum(ColumnSortDirection_$type, v);
    }
    get iconColor() {
        return brushToString(this.i.a0);
    }
    set iconColor(v) {
        this.i.a0 = stringToBrush(v);
    }
    get isAnimationEnabled() {
        return this.i.l;
    }
    set isAnimationEnabled(v) {
        this.i.l = ensureBool(v);
    }
    /**
     * Gets or sets the scaling value used to affect the pixel density of the control.
     * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
     * to appear blurry.
    */
    get pixelScalingRatio() {
        return this.i.w;
    }
    set pixelScalingRatio(v) {
        this.i.w = +v;
    }
    /**
     * Gets the actual pixel scaling ratio used to affect the pixel density of the control.
     * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
     * to appear blurry.
    */
    get actualPixelScalingRatio() {
        return this.i.p;
    }
    set actualPixelScalingRatio(v) {
        this.i.p = +v;
    }
    get sortIndicatorStyle() {
        return this.i.f;
    }
    set sortIndicatorStyle(v) {
        this.i.f = ensureEnum(SortIndicatorStyle_$type, v);
    }
    findByName(name) {
        return null;
    }
    get hasUserValues() {
        return this._hasUserValues;
    }
    __m(propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    }
    _styling(container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        let genericPrefix = "";
        let typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("DataGridSortIndicator");
        let additionalPrefixes = [];
        let prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        let b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            let basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            let parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            let parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    }
    getDesiredScale() {
        let iv = this.i.v();
        return (iv);
    }
    notifySizeChanged(width, height) {
        this.i.ar(width, height);
    }
    notifyCellSizeChanged() {
        this.i.aq();
    }
    get renderCompleted() {
        return this._renderCompleted;
    }
    set renderCompleted(ev) {
        if (this._renderCompleted_wrapped !== null) {
            this.i.renderCompleted = delegateRemove(this.i.renderCompleted, this._renderCompleted_wrapped);
            this._renderCompleted_wrapped = null;
            this._renderCompleted = null;
        }
        this._renderCompleted = ev;
        this._renderCompleted_wrapped = (o, e) => {
            let outerArgs = new IgrSortIndicatorRenderCompletedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeRenderCompleted) {
                this.beforeRenderCompleted(this, outerArgs);
            }
            if (this._renderCompleted) {
                this._renderCompleted(this, outerArgs);
            }
        };
        this.i.renderCompleted = delegateCombine(this.i.renderCompleted, this._renderCompleted_wrapped);
    }
}
