// igniteui-react-grids/esm2015/lib/igr-grid-cell-edit-ended-event-args.js
import { IgrColumn } from "./igr-column";
import { GridCellEditEndedEventArgs as GridCellEditEndedEventArgs_internal } from "./GridCellEditEndedEventArgs";
export class IgrGridCellEditEndedEventArgs {
    createImplementation() {
        return new GridCellEditEndedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets the column of the cell that ended editing.
    */
    get column() {
        if (this.i.a == null) {
            return null;
        }
        if (!this.i.a.externalObject) {
            let e = IgrColumn._createFromInternal(this.i.a);
            if (e) {
                e._implementation = this.i.a;
            }
            this.i.a.externalObject = e;
        }
        return this.i.a.externalObject;
    }
    set column(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    /**
     * Gets the row index of the cell that ended editing.
    */
    get row() {
        return this.i.b;
    }
    set row(v) {
        this.i.b = +v;
    }
    /**
     * Gets the row item.
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
}
