// igniteui-react-grids/esm2015/lib/igr-column-hidden-changed-event-args.js
import { IgrColumn } from "./igr-column";
import { ColumnHiddenChangedEventArgs as ColumnHiddenChangedEventArgs_internal } from "./ColumnHiddenChangedEventArgs";
import { ensureBool } from "igniteui-react-core";
export class IgrColumnHiddenChangedEventArgs {
    createImplementation() {
        return new ColumnHiddenChangedEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get column() {
        if (this.i.a == null) {
            return null;
        }
        if (!this.i.a.externalObject) {
            let e = IgrColumn._createFromInternal(this.i.a);
            if (e) {
                e._implementation = this.i.a;
            }
            this.i.a.externalObject = e;
        }
        return this.i.a.externalObject;
    }
    set column(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    get isHidden() {
        return this.i.b;
    }
    set isHidden(v) {
        this.i.b = ensureBool(v);
    }
}
