// igniteui-react-grids/esm2015/lib/igr-grid-cell-position.js
import { GridCellPosition as GridCellPosition_internal } from "./GridCellPosition";
export class IgrGridCellPosition {
    createImplementation() {
        return new GridCellPosition_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets or sets the row position of the cell.
    */
    get rowIndex() {
        return this.i.b;
    }
    set rowIndex(v) {
        this.i.b = +v;
    }
    /**
     * Gets or sets the column position of the cell.  Can be null for spanning elements like section headers.
    */
    get columnUniqueKey() {
        return this.i.d;
    }
    set columnUniqueKey(v) {
        this.i.d = v;
    }
    findByName(name) {
        return null;
    }
}
