// igniteui-react-grids/esm2015/lib/igr-section-header-cell-info.js
import { GroupSummaryDisplayMode_$type } from "./GroupSummaryDisplayMode";
import { IgrCellInfo } from "./igr-cell-info";
import { SectionHeaderCellModel as SectionHeaderCellModel_internal } from "./SectionHeaderCellModel";
import { ensureEnum } from "igniteui-react-core";
/**
 * Backing information for a section header cell in the grid.
*/
export class IgrSectionHeaderCellInfo extends IgrCellInfo {
    createImplementation() {
        return new SectionHeaderCellModel_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Sets or gets the resolved text for the section header cell.
    */
    get resolvedText() {
        return this.i.k6;
    }
    set resolvedText(v) {
        this.i.k6 = v;
    }
    /**
     * Sets or gets the resolved summary text for the section header cell.
    */
    get resolvedSummaryText() {
        return this.i.k4;
    }
    set resolvedSummaryText(v) {
        this.i.k4 = v;
    }
    /**
     * Sets or gets how summaries are displayed in the section header cell.
    */
    get summaryDisplayMode() {
        return this.i.kv;
    }
    set summaryDisplayMode(v) {
        this.i.kv = ensureEnum(GroupSummaryDisplayMode_$type, v);
    }
}
