// igniteui-react-grids/esm2015/lib/igr-cell-range.js
import { CellRange as CellRange_internal } from "./CellRange";
/**
 * An indentifier for a specific cell in the grid.
*/
export class IgrCellRange {
    createImplementation() {
        return new CellRange_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets or sets the start column of the range.
    */
    get startColumn() {
        return this.i.k;
    }
    set startColumn(v) {
        this.i.k = +v;
    }
    /**
     * Gets or sets the end column of the range.
    */
    get endColumn() {
        return this.i.i;
    }
    set endColumn(v) {
        this.i.i = +v;
    }
    /**
     * Gets or sets the start row of the range.
    */
    get startRow() {
        return this.i.l;
    }
    set startRow(v) {
        this.i.l = +v;
    }
    /**
     * Gets or sets the end row of the range.
    */
    get endRow() {
        return this.i.j;
    }
    set endRow(v) {
        this.i.j = +v;
    }
    findByName(name) {
        return null;
    }
}
