// igniteui-react-grids/esm2015/lib/igr-format-summary-text-event-args.js
import { FormatSummaryTextEventArgs as FormatSummaryTextEventArgs_internal } from "./FormatSummaryTextEventArgs";
export class IgrFormatSummaryTextEventArgs {
    createImplementation() {
        return new FormatSummaryTextEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets the summary property name.
    */
    get field() {
        return this.i.c;
    }
    set field(v) {
        this.i.c = v;
    }
    /**
     * Gets the summary display name.
    */
    get displayName() {
        return this.i.b;
    }
    set displayName(v) {
        this.i.b = v;
    }
    /**
     * Gets the original value before formatting.
    */
    get summaryResult() {
        return this.i.a;
    }
    set summaryResult(v) {
        this.i.a = v;
    }
    /**
     * Gets or sets the summary value as formatted text.  Setting this property will only affect summary cells.
    */
    get formattedResult() {
        return this.i.d;
    }
    set formattedResult(v) {
        this.i.d = v;
    }
    /**
     * Gets or sets the final text that will be displayed in the group header.
    */
    get formattedText() {
        return this.i.e;
    }
    set formattedText(v) {
        this.i.e = v;
    }
}
