// igniteui-react-grids/esm2015/lib/igr-date-time-column.js
import { DateTimeFormats_$type } from "./DateTimeFormats";
import { EditorType_$type } from "./EditorType";
import { IgrColumn } from "./igr-column";
import { DateTimeColumn } from "./DateTimeColumn";
import { ensureEnum, arrayFindByName } from "igniteui-react-core";
/**
 * A column for displaying date/time values.
*/
export class IgrDateTimeColumn extends IgrColumn {
    createImplementation() {
        return new DateTimeColumn();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the date time format to use for this column. If FormatString is specificied this value is ignored.
    */
    get dateTimeFormat() {
        return this.i.jt;
    }
    set dateTimeFormat(v) {
        this.i.jt = ensureEnum(DateTimeFormats_$type, v);
    }
    /**
     * Gets or sets the format string to apply to the value. If set, the other value formatting properties on this column are ignored.
    */
    get formatString() {
        return this.i.j4;
    }
    set formatString(v) {
        this.i.j4 = v;
    }
    get formatSpecifiers() {
        return this.i.jq;
    }
    set formatSpecifiers(v) {
        this.i.jq = v;
    }
    /**
     * Gets or sets the INTL DateTimeFormat object to use for formatting the date values.
    */
    get formatOverride() {
        return this.i.jy;
    }
    set formatOverride(v) {
        this.i.jy = v;
    }
    /**
     * Gets or sets the editor type used for editing cells in this column.
    */
    get editorType() {
        return this.i.ju;
    }
    set editorType(v) {
        this.i.ju = ensureEnum(EditorType_$type, v);
    }
    /**
     * Gets or sets the ComboBox data source.
    */
    get editorDataSource() {
        return this.i.jx;
    }
    set editorDataSource(v) {
        this.i.jx = v;
    }
    /**
     * Gets or sets the ComboBox text field.
    */
    get editorTextField() {
        return this.i.j1;
    }
    set editorTextField(v) {
        this.i.j1 = v;
    }
    /**
     * Gets or sets the ComboBox value field.
    */
    get editorValueField() {
        return this.i.jp;
    }
    set editorValueField(v) {
        this.i.jp = v;
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.formatSpecifiers != null && arrayFindByName(this.formatSpecifiers, name)) {
            return arrayFindByName(this.formatSpecifiers, name);
        }
        return null;
    }
}
