// igniteui-react-grids/esm2015/lib/igr-column-property-setter.js
import { ColumnPropertySetter as ColumnPropertySetter_internal } from "./ColumnPropertySetter";
/**
 * Indicates that a specified column should have a specified property set to a value.
*/
export class IgrColumnPropertySetter {
    createImplementation() {
        return new ColumnPropertySetter_internal(0);
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The name or key of the column to change the property of
    */
    get columnName() {
        return this.i.c;
    }
    set columnName(v) {
        this.i.c = v;
    }
    /**
     * The name of the property to change
    */
    get propertyName() {
        return this.i.d;
    }
    set propertyName(v) {
        this.i.d = v;
    }
    /**
     * The new value of the property
    */
    get value() {
        return this.i.b;
    }
    set value(v) {
        this.i.b = v;
    }
    findByName(name) {
        return null;
    }
}
