// igniteui-react-grids/esm2015/lib/igr-summary-data.js
import { SummaryOperand_$type } from "igniteui-react-core";
import { SummaryData as SummaryData_internal } from "./SummaryData";
import { ensureEnum } from "igniteui-react-core";
/**
 * Object used to store summary information for group headers.
*/
export class IgrSummaryData {
    createImplementation() {
        return new SummaryData_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get summaryName() {
        return this.i.e;
    }
    set summaryName(v) {
        this.i.e = v;
    }
    get summaryValue() {
        return this.i.b;
    }
    set summaryValue(v) {
        this.i.b = v;
    }
    get summaryOperand() {
        return this.i.a;
    }
    set summaryOperand(v) {
        this.i.a = ensureEnum(SummaryOperand_$type, v);
    }
    get formattedText() {
        return this.i.c;
    }
    set formattedText(v) {
        this.i.c = v;
    }
    get formattedValue() {
        return this.i.d;
    }
    set formattedValue(v) {
        this.i.d = v;
    }
    findByName(name) {
        return null;
    }
}
