// igniteui-react-grids/esm2015/lib/igr-summary-cell-info.js
import { IgrCellInfo } from "./igr-cell-info";
import { SummaryCellModel as SummaryCellModel_internal } from "./SummaryCellModel";
import { brushToString, stringToBrush } from "igniteui-react-core";
import { FontInfo } from "igniteui-react-core";
/**
 * Backing information for a section header cell in the grid.
*/
export class IgrSummaryCellInfo extends IgrCellInfo {
    createImplementation() {
        return new SummaryCellModel_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Sets or gets the resolved summary label for the summary cell.
    */
    get resolvedSummaryLabel() {
        return this.i.ky;
    }
    set resolvedSummaryLabel(v) {
        this.i.ky = v;
    }
    /**
     * Sets or gets the resolved summary value for the sumamry cell.
    */
    get resolvedSummaryValue() {
        return this.i.k0;
    }
    set resolvedSummaryValue(v) {
        this.i.k0 = v;
    }
    /**
     * Gets or sets the color of the text for the summary label in the cell.
    */
    get summaryLabelTextColor() {
        return brushToString(this.i.k5);
    }
    set summaryLabelTextColor(v) {
        this.i.k5 = stringToBrush(v);
    }
    get font() {
        if (this.i.kr == null) {
            return null;
        }
        return this.i.kr.fontString;
    }
    set font(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.kr = fi;
    }
}
