// igniteui-react-grids/esm2015/lib/NumericCellModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellModelExport } from "./CellModelExport";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let NumericCellModelExport = /*@__PURE__*/ (() => {
    class NumericCellModelExport extends CellModelExport {
        constructor() {
            super(...arguments);
            this.ae = 0;
            this.af = 0;
            this.ag = 0;
        }
        get_ac() {
            return "NumericCell";
        }
        ad(a) {
            super.ad(a);
            a.l("numericValue : " + this.ae + ", ");
            a.l("maxDecimalPlaces : " + this.af + ", ");
            a.l("minDecimalPlaces : " + this.ag + ", ");
        }
    }
    NumericCellModelExport.$t = /*@__PURE__*/ markType(NumericCellModelExport, 'NumericCellModelExport', CellModelExport.$);
    return NumericCellModelExport;
})();
