// igniteui-react-grids/esm2015/lib/PlaceholderView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
import { truncate } from "igniteui-react-core";
/**
 * @hidden
 */
export let PlaceholderView = /*@__PURE__*/ (() => {
    class PlaceholderView extends Base {
        constructor(a) {
            super();
            this.f = null;
            this.k = new Color();
            this.h = 1;
            this.a = true;
            this.c = 0;
            this.b = 0;
            this.g = null;
            this.f = a;
            this.g = this.f.createElement("div");
            this.g.setRawStyleProperty("position", "absolute");
            this.g.setRawStyleProperty("pointer-events", "none");
        }
        get l() {
            return this.k;
        }
        set l(a) {
            this.k = a;
            this.j();
        }
        get i() {
            return this.h;
        }
        set i(a) {
            this.a = true;
            this.h = a;
            this.j();
        }
        get e() {
            return this.c;
        }
        set e(a) {
            this.c = a;
            this.j();
        }
        get d() {
            return this.b;
        }
        set d(a) {
            this.b = a;
            this.j();
        }
        j() {
            let a = this.e;
            let b = this.d;
            let c = Color.u(truncate((this.k.l * this.h)), this.k.o, this.k.n, this.k.m);
            this.g.setRawStyleProperty("background-color", c.colorString);
            this.g.setRawStyleProperty("width", a + "px");
            this.g.setRawStyleProperty("height", b + "px");
        }
    }
    PlaceholderView.$t = /*@__PURE__*/ markType(PlaceholderView, 'PlaceholderView');
    return PlaceholderView;
})();
