// igniteui-react-grids/esm2015/lib/TemplateHeader.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Header } from "./Header";
import { TemplateHeaderCellModel } from "./TemplateHeaderCellModel";
import { markType } from "igniteui-react-core";
import { TemplateHeaderCellUpdatingEventArgs } from "./TemplateHeaderCellUpdatingEventArgs";
/**
 * @hidden
 */
export let TemplateHeader = /*@__PURE__*/ (() => {
    class TemplateHeader extends Header {
        constructor() {
            super();
            this.cellUpdating = null;
        }
        l(a) {
            return new TemplateHeaderCellModel();
        }
        cm(a, b, c) {
            let d = "template-header-" + a.az.toString();
            return d;
        }
        cy(a, b, c, d) {
            super.cy(a, b, c, d);
            let e = a;
            let f = b.a0(a.k);
            e.ks = f.hh;
            e.m = f.e6;
            e.kv = f.f3;
        }
        fd(a, b) {
            if (this.cellUpdating != null) {
                this.cellUpdating(this, ((() => {
                    let $ret = new TemplateHeaderCellUpdatingEventArgs();
                    $ret.cellInfo = a;
                    $ret.content = b;
                    return $ret;
                })()));
            }
        }
    }
    TemplateHeader.$t = /*@__PURE__*/ markType(TemplateHeader, 'TemplateHeader', Header.$);
    return TemplateHeader;
})();
