// igniteui-react-grids/esm2015/lib/TextCell.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridCellBase } from "./GridCellBase";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TextCell = /*@__PURE__*/ (() => {
    class TextCell extends GridCellBase {
        constructor(a) {
            super(a);
            this.b7 = null;
            this.ca = -1;
            this.b9 = -1;
            this.b7 = a.createElement("span");
            this.z.setRawStyleProperty("text-align", "left");
            this.b7.setRawStyleProperty("display", "inline-block");
            this.b7.setRawStyleProperty("line-height", "normal");
            this.b7.setRawStyleProperty("vertical-align", "middle");
            this.b7.setRawStyleProperty("overflow", "hidden");
            this.b7.setRawStyleProperty("white-space", "nowrap");
            this.b7.setRawStyleProperty("text-overflow", "ellipsis");
            this.z.setRawStyleProperty("vertical-align", "middle");
            this.z.append(this.b7);
        }
        get b8() {
            return this.b7;
        }
        bw(a, b) {
            super.bw(a, b);
            let c = Math.max(0, a - b - (this.au + this.aw + this.av));
            if (this.ca != c) {
                this.ca = c;
                this.b7.setRawStyleProperty("max-width", c.toString() + "px");
            }
        }
        bq(a, b) {
            super.bq(a, b);
            let c = Math.max(0, a - b);
            if (this.b9 != c) {
                this.b9 = c;
                this.b7.setRawStyleProperty("max-height", c.toString() + "px");
            }
        }
        get_r() {
            return true;
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            if (a == "Indent" || a == "IsCollapsable") {
                this.b7.setRawStyleProperty("margin-left", this.au + this.aw + this.av + "px");
            }
        }
    }
    TextCell.$t = /*@__PURE__*/ markType(TextCell, 'TextCell', GridCellBase.$);
    return TextCell;
})();
