// igniteui-react-grids/esm2015/lib/VisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { RectData } from "igniteui-react-core";
import { GridLayerExport } from "./GridLayerExport";
import { List$1 } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let VisualModelExport = /*@__PURE__*/ (() => {
    class VisualModelExport extends Base {
        constructor() {
            super();
            this.ae = null;
            this.ah = 0;
            this.ag = 0;
            this.an = 0;
            this.am = 0;
            this.ak = 0;
            this.al = 0;
            this.ai = 0;
            this.aq = 0;
            this.ar = 0;
            this.as = 0;
            this.ap = 0;
            this.ao = 0;
            this.at = 0;
            this.x = null;
            this.a = null;
            this.aa = null;
            this.b = null;
            this.r = null;
            this.q = null;
            this.s = null;
            this.h = null;
            this.c = null;
            this.k = null;
            this.f = null;
            this.g = null;
            this.w = null;
            this.u = null;
            this.t = null;
            this.v = null;
            this.z = null;
            this.y = null;
            this.o = null;
            this.n = null;
            this.ac = null;
            this.ab = null;
            this.j = null;
            this.i = null;
            this.e = null;
            this.d = null;
            this.m = null;
            this.l = null;
            this.ad = null;
            this.p = null;
            this.aj = 0;
            this.af = null;
            this.au = 0;
            this.ae = RectData.empty;
            this.x = new GridLayerExport("SubContentLayer");
            this.a = new GridLayerExport("ContentLayer");
            this.aa = new GridLayerExport("SuperContentLayer");
            this.b = new GridLayerExport("DecoratorLayer");
            this.r = new GridLayerExport("PinnedRowsSubLayer");
            this.q = new GridLayerExport("PinnedRowsLayer");
            this.s = new GridLayerExport("PinnedRowsSuperLayer");
            this.h = new GridLayerExport("FixedSubContentLayer");
            this.c = new GridLayerExport("FixedContentLayer");
            this.k = new GridLayerExport("FixedSuperContentLayer");
            this.f = new GridLayerExport("FixedPinnedRowsLayer");
            this.g = new GridLayerExport("FixedSeparatorLayer");
            this.w = new GridLayerExport("SectionLayer");
            this.u = new GridLayerExport("PinnedSectionSubLayer");
            this.t = new GridLayerExport("PinnedSectionLayer");
            this.v = new GridLayerExport("PinnedSectionSuperLayer");
            this.z = new GridLayerExport("SubHeaderLayer");
            this.y = new GridLayerExport("SubHeaderDecoratorLayer");
            this.o = new GridLayerExport("HeaderLayer");
            this.n = new GridLayerExport("HeaderDecoratorLayer");
            this.ac = new GridLayerExport("SuperHeaderLayer");
            this.ab = new GridLayerExport("SuperHeaderDecoratorLayer");
            this.j = new GridLayerExport("FixedSubHeaderLayer");
            this.i = new GridLayerExport("FixedSubHeaderDecoratorLayer");
            this.e = new GridLayerExport("FixedHeaderLayer");
            this.d = new GridLayerExport("FixedHeaderDecoratorLayer");
            this.m = new GridLayerExport("FixedSuperHeaderLayer");
            this.l = new GridLayerExport("FixedSuperHeaderDecoratorLayer");
            this.ad = new GridLayerExport("VerticalSeparatorLayer");
            this.p = new GridLayerExport("HiddenLayer");
            this.aj = 1;
        }
        ax() {
            this.af = new List$1(GridLayerExport.$, 0);
            this.af.add(this.x);
            this.af.add(this.a);
            this.af.add(this.aa);
            this.af.add(this.b);
            this.af.add(this.r);
            this.af.add(this.q);
            this.af.add(this.s);
            this.af.add(this.h);
            this.af.add(this.c);
            this.af.add(this.k);
            this.af.add(this.f);
            this.af.add(this.g);
            this.af.add(this.w);
            this.af.add(this.u);
            this.af.add(this.t);
            this.af.add(this.v);
            this.af.add(this.z);
            this.af.add(this.y);
            this.af.add(this.o);
            this.af.add(this.n);
            this.af.add(this.ac);
            this.af.add(this.ab);
            this.af.add(this.j);
            this.af.add(this.i);
            this.af.add(this.e);
            this.af.add(this.d);
            this.af.add(this.m);
            this.af.add(this.l);
            this.af.add(this.ad);
        }
        aw() {
            this.x.f();
            this.a.f();
            this.aa.f();
            this.b.f();
            this.r.f();
            this.q.f();
            this.s.f();
            this.h.f();
            this.c.f();
            this.k.f();
            this.f.f();
            this.g.f();
            this.w.f();
            this.u.f();
            this.t.f();
            this.v.f();
            this.z.f();
            this.y.f();
            this.o.f();
            this.n.f();
            this.ac.f();
            this.ab.f();
            this.j.f();
            this.i.f();
            this.e.f();
            this.d.f();
            this.m.f();
            this.l.f();
            this.ad.f();
        }
        av() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("top : " + this.as + ", ");
            a.l("left : " + this.ap + ", ");
            a.l("height : " + this.ao + ", ");
            a.l("width : " + this.at + ", ");
            a.l("absoluteWidth : " + this.ah + ", ");
            a.l("absoluteHeight : " + this.ag + ", ");
            a.l("headerHeight : " + this.an + ", ");
            a.l("fixedRowHeight : " + this.am + ", ");
            a.l("fixedColumnWidthLeft : " + this.ak + ", ");
            a.l("fixedColumnWidthRight : " + this.al + ", ");
            a.l("dipScalingRatio: " + this.aj + ", ");
            a.l("columnSpacing : " + this.ai + ", ");
            a.l("separatorLeft : " + this.aq + ", ");
            a.l("separatorRight : " + this.ar + ", ");
            for (let b = 0; b < this.af.count; b++) {
                let c = this.af._inner[b].c.charAt(0).toString().toLowerCase() + this.af._inner[b].c.substr(1);
                a.l(c + ": " + this.af._inner[b].d() + ", ");
            }
            a.l("actualVisibleRegion: { left: " + this.ae.left + ", top: " + this.ae.top + ", width: " + this.ae.width + ", height: " + this.ae.height + "}");
            a.l("}");
            return a.toString();
        }
    }
    VisualModelExport.$t = /*@__PURE__*/ markType(VisualModelExport, 'VisualModelExport');
    return VisualModelExport;
})();
