// igniteui-react-grids/esm5/lib/DataGridScrollerKeyboardListener.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, markType } from "igniteui-react-core";
import { IScrollerKeyboardListener_$type } from "igniteui-react-core";
/**
 * @hidden
 */
var DataGridScrollerKeyboardListener = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(DataGridScrollerKeyboardListener, _super);
    function DataGridScrollerKeyboardListener(a) {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.a = a;
        return _this;
    }
    DataGridScrollerKeyboardListener.prototype.onDownArrow = function (a, b) {
        return this.a.h8(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onEnd = function (a, b) {
        return this.a.h9(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onHome = function (a, b) {
        return this.a.id(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onLeftArrow = function (a, b) {
        return this.a.ie(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onPageDown = function (a, b) {
        return this.a.ig(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onPageUp = function (a, b) {
        return this.a.ih(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onRightArrow = function (a, b) {
        return this.a.ii(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onTab = function (a, b) {
        return this.a.ij(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onUpArrow = function (a, b) {
        return this.a.ik(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onWheel = function () {
        this.a.sc();
    };
    DataGridScrollerKeyboardListener.prototype.onEnter = function (a, b) {
        return this.a.ia(a, b);
    };
    DataGridScrollerKeyboardListener.prototype.onEscape = function () {
        return this.a.ib();
    };
    DataGridScrollerKeyboardListener.prototype.onF2 = function () {
        return this.a.ic();
    };
    DataGridScrollerKeyboardListener.$t = markType(DataGridScrollerKeyboardListener, 'DataGridScrollerKeyboardListener', Base.$, [IScrollerKeyboardListener_$type]);
    return DataGridScrollerKeyboardListener;
}(Base));
export { DataGridScrollerKeyboardListener };
