// igniteui-react-grids/esm5/lib/igr-grid-animation-phase-settings.js
import { GridEasingFunctionType_$type } from "./GridEasingFunctionType";
import { GridAnimationPhaseSettings as GridAnimationPhaseSettings_internal } from "./GridAnimationPhaseSettings";
import { ensureBool, ensureEnum } from "igniteui-react-core";
/**
 * This class is used to group the settings to define an animation phase.
*/
var IgrGridAnimationPhaseSettings = /** @class */ /*@__PURE__*/ (function () {
    function IgrGridAnimationPhaseSettings() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    IgrGridAnimationPhaseSettings.prototype.createImplementation = function () {
        return new GridAnimationPhaseSettings_internal();
    };
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: true,
        configurable: true
    });
    IgrGridAnimationPhaseSettings.prototype.onImplementationCreated = function () {
    };
    IgrGridAnimationPhaseSettings.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    };
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "durationMilliseconds", {
        /**
         * The duration of an animation, in milliseconds
        */
        get: function () {
            return this.i.l;
        },
        set: function (v) {
            this.i.l = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "holdInitialMilliseconds", {
        /**
         * The amount of time to delay an animation, in milliseconds.
        */
        get: function () {
            return this.i.m;
        },
        set: function (v) {
            this.i.m = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "perItemDelayMilliseconds", {
        /**
         * The amount of time to delay each item animating, in milliseconds
        */
        get: function () {
            return this.i.n;
        },
        set: function (v) {
            this.i.n = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "subItemDurationMilliseconds", {
        /**
         * The amount of time each item animating should take, in milliseconds
        */
        get: function () {
            return this.i.o;
        },
        set: function (v) {
            this.i.o = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "desiredSubItemDurationMilliseconds", {
        /**
         * The desired amount of time each item animating should take, in milliseconds
        */
        get: function () {
            return this.i.k;
        },
        set: function (v) {
            this.i.k = +v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "shouldItemsFinishSimultaneously", {
        /**
         * Configures whether all items animating should finish simultaneously
        */
        get: function () {
            return this.i.e;
        },
        set: function (v) {
            this.i.e = ensureBool(v);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgrGridAnimationPhaseSettings.prototype, "easingFunctionType", {
        /**
         * Configures the Easing function used to animate the individual items
        */
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureEnum(GridEasingFunctionType_$type, v);
        },
        enumerable: true,
        configurable: true
    });
    IgrGridAnimationPhaseSettings.prototype.findByName = function (name) {
        return null;
    };
    return IgrGridAnimationPhaseSettings;
}());
export { IgrGridAnimationPhaseSettings };
