// igniteui-react-grids/esm5/lib/SummaryCell.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { GridCellBase } from "./GridCellBase";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var SummaryCell = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(SummaryCell, _super);
    function SummaryCell(a) {
        var _this = _super.call(this, a) || this;
        _this.b7 = null;
        _this.b8 = null;
        _this.b9 = null;
        _this.cd = -1;
        _this.cc = -1;
        _this.z.setRawStyleProperty("vertical-align", "middle");
        _this.b9 = a.createElement("div");
        _this.b7 = a.createElement("span");
        _this.b7.setRawStyleProperty("display", "inline-block");
        _this.b7.setRawStyleProperty("width", "40%");
        _this.b7.setRawStyleProperty("white-space", "nowrap");
        _this.b7.setRawStyleProperty("text-overflow", "ellipsis");
        _this.b7.setRawStyleProperty("vertical-align", "middle");
        _this.b7.setRawStyleProperty("overflow", "hidden");
        _this.b9.append(_this.b7);
        _this.b8 = a.createElement("span");
        _this.b8.setRawStyleProperty("display", "inline-block");
        _this.b8.setRawStyleProperty("width", "60%");
        _this.b8.setRawStyleProperty("text-align", "right");
        _this.b8.setRawStyleProperty("white-space", "nowrap");
        _this.b8.setRawStyleProperty("text-overflow", "ellipsis");
        _this.b8.setRawStyleProperty("vertical-align", "middle");
        _this.b8.setRawStyleProperty("overflow", "hidden");
        _this.b9.append(_this.b8);
        _this.z.append(_this.b9);
        return _this;
    }
    Object.defineProperty(SummaryCell.prototype, "ca", {
        get: function () {
            return this.b7;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SummaryCell.prototype, "cb", {
        get: function () {
            return this.b8;
        },
        enumerable: true,
        configurable: true
    });
    SummaryCell.prototype.bw = function (a, b) {
        _super.prototype.bw.call(this, a, b);
        var c = Math.max(0, a - b);
        if (this.cd != c) {
            this.cd = c;
        }
    };
    SummaryCell.prototype.bq = function (a, b) {
        _super.prototype.bq.call(this, a, b);
        var c = Math.max(0, a - b);
        if (this.cc != c) {
            this.cc = c;
        }
    };
    SummaryCell.prototype.get_r = function () {
        return false;
    };
    SummaryCell.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        if (a == "Indent" || a == "IsCollapsable") {
            this.b9.setRawStyleProperty("margin-left", this.au + this.aw + this.av + "px");
        }
    };
    SummaryCell.$t = markType(SummaryCell, 'SummaryCell', GridCellBase.$);
    return SummaryCell;
}(GridCellBase));
export { SummaryCell };
