// igniteui-react-grids/esm5/lib/TextCell.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { GridCellBase } from "./GridCellBase";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var TextCell = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(TextCell, _super);
    function TextCell(a) {
        var _this = _super.call(this, a) || this;
        _this.b7 = null;
        _this.ca = -1;
        _this.b9 = -1;
        _this.b7 = a.createElement("span");
        _this.z.setRawStyleProperty("text-align", "left");
        _this.b7.setRawStyleProperty("display", "inline-block");
        _this.b7.setRawStyleProperty("line-height", "normal");
        _this.b7.setRawStyleProperty("vertical-align", "middle");
        _this.b7.setRawStyleProperty("overflow", "hidden");
        _this.b7.setRawStyleProperty("white-space", "nowrap");
        _this.b7.setRawStyleProperty("text-overflow", "ellipsis");
        _this.z.setRawStyleProperty("vertical-align", "middle");
        _this.z.append(_this.b7);
        return _this;
    }
    Object.defineProperty(TextCell.prototype, "b8", {
        get: function () {
            return this.b7;
        },
        enumerable: true,
        configurable: true
    });
    TextCell.prototype.bw = function (a, b) {
        _super.prototype.bw.call(this, a, b);
        var c = Math.max(0, a - b - (this.au + this.aw + this.av));
        if (this.ca != c) {
            this.ca = c;
            this.b7.setRawStyleProperty("max-width", c.toString() + "px");
        }
    };
    TextCell.prototype.bq = function (a, b) {
        _super.prototype.bq.call(this, a, b);
        var c = Math.max(0, a - b);
        if (this.b9 != c) {
            this.b9 = c;
            this.b7.setRawStyleProperty("max-height", c.toString() + "px");
        }
    };
    TextCell.prototype.get_r = function () {
        return true;
    };
    TextCell.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        if (a == "Indent" || a == "IsCollapsable") {
            this.b7.setRawStyleProperty("margin-left", this.au + this.aw + this.av + "px");
        }
    };
    TextCell.$t = markType(TextCell, 'TextCell', GridCellBase.$);
    return TextCell;
}(GridCellBase));
export { TextCell };
